<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:27 BST 2008 -->
<TITLE>
SeqIOTools (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojava.bio.seq.io.SeqIOTools class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="SeqIOTools (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SeqIOTools.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilder.html" title="interface in org.biojava.bio.seq.io"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/seq/io/SeqIOTools.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SeqIOTools.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.biojava.bio.seq.io</FONT>
<BR>
Class SeqIOTools</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../../resources/inherit.gif" ALT="extended by "><B>org.biojava.bio.seq.io.SeqIOTools</B>
</PRE>
<HR>
<B>Deprecated.</B>&nbsp;<I>use org.biojavax.bio.seq.RichSequence.IOTools</I>
<P>
<DL>
<DT><PRE>public final class <B>SeqIOTools</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
A set of convenience methods for handling common file formats.
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.1</DD>
<DT><B>Author:</B></DT>
  <DD>Thomas Down, Mark Schreiber, Nimesh Singh, Matthew Pocock, Keith James</DD>
</DL>
<HR>

<P>

<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#biojavaToFile(int, java.io.OutputStream, java.lang.Object)">biojavaToFile</A></B>(int&nbsp;fileType,
              java.io.OutputStream&nbsp;os,
              java.lang.Object&nbsp;biojava)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Converts a Biojava object to the given filetype.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#biojavaToFile(java.lang.String, java.lang.String, java.io.OutputStream, java.lang.Object)">biojavaToFile</A></B>(java.lang.String&nbsp;formatName,
              java.lang.String&nbsp;alphabetName,
              java.io.OutputStream&nbsp;os,
              java.lang.Object&nbsp;biojava)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a Biojava <code>SequenceIterator</code>,
 <code>SequenceDB</code>, <code>Sequence</code> or <code>Aligment</code>
 to an <code>OutputStream</code></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#fileToBiojava(int, java.io.BufferedReader)">fileToBiojava</A></B>(int&nbsp;fileType,
              java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads a file and returns the corresponding Biojava object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#fileToBiojava(java.lang.String, java.lang.String, java.io.BufferedReader)">fileToBiojava</A></B>(java.lang.String&nbsp;formatName,
              java.lang.String&nbsp;alphabetName,
              java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads a file with the specified format and alphabet</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#formatToFactory(org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.symbol.Alphabet)">formatToFactory</A></B>(<A HREF="../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
                <A HREF="../../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>as this essentially duplicates the operation
 available in the method <code>identifyBuilderFactory</code>.</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getAlphabet(int)">getAlphabet</A></B>(int&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>getAlphabet</code> accepts a value which represents a
 sequence format and returns the relevant
 <code>FiniteAlphabet</code> object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getBuilderFactory(int)">getBuilderFactory</A></B>(int&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>getBuilderFactory</code> accepts a value which represents
 a sequence format and returns the relevant
 <code>SequenceBuilderFactory</code> object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getEmblBuilderFactory()">getEmblBuilderFactory</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Get a default SequenceBuilderFactory for handling EMBL
 files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getFastaBuilderFactory()">getFastaBuilderFactory</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Get a default SequenceBuilderFactory for handling FASTA
 files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getGenbankBuilderFactory()">getGenbankBuilderFactory</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Get a default SequenceBuilderFactory for handling GenBank
 files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getGenpeptBuilderFactory()">getGenpeptBuilderFactory</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Get a default SequenceBuilderFactory for handling Genpept
 files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getSequenceFormat(int)">getSequenceFormat</A></B>(int&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>getSequenceFormat</code> accepts a value which represents
 a sequence format and returns the relevant
 <code>SequenceFormat</code> object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#getSwissprotBuilderFactory()">getSwissprotBuilderFactory</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Get a default SequenceBuilderFactory for handling Swissprot
 files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#guessFileType(java.io.File)">guessFileType</A></B>(java.io.File&nbsp;seqFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>because there is no standard file naming convention
 and guessing by file name is inherantly error prone and bad.</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#identifyFormat(java.lang.String, java.lang.String)">identifyFormat</A></B>(java.lang.String&nbsp;formatName,
               java.lang.String&nbsp;alphabetName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>identifyFormat</code> performs a case-insensitive mapping
 of a pair of common sequence format name (such as 'embl',
 'genbank' or 'fasta') and alphabet name (such as 'dna', 'rna',
 'protein', 'aa') to an integer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readEmbl(java.io.BufferedReader)">readEmbl</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an EMBL-format stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readEmblNucleotide(java.io.BufferedReader)">readEmblNucleotide</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an EMBL-format stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readEmblRNA(java.io.BufferedReader)">readEmblRNA</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an EMBL-format stream, but for RNA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Read a fasta file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SequenceBuilderFactory)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;seqFactory)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Read a fasta file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFasta(java.io.InputStream, org.biojava.bio.symbol.Alphabet)">readFasta</A></B>(java.io.InputStream&nbsp;seqFile,
          <A HREF="../../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Create a sequence database from a fasta file provided as an
 input stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFastaDNA(java.io.BufferedReader)">readFastaDNA</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an FASTA-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFastaProtein(java.io.BufferedReader)">readFastaProtein</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an FASTA-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFastaRNA(java.io.BufferedReader)">readFastaRNA</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an FASTA-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readGenbank(java.io.BufferedReader)">readGenbank</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an Genbank-format stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readGenbankXml(java.io.BufferedReader)">readGenbankXml</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an GenbankXML-format stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readGenpept(java.io.BufferedReader)">readGenpept</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an Genpept-format stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readSwissprot(java.io.BufferedReader)">readSwissprot</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Iterate over the sequences in an Swissprot-format stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeEmbl(java.io.OutputStream, org.biojava.bio.seq.Sequence)">writeEmbl</A></B>(java.io.OutputStream&nbsp;os,
          <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a single Sequence to an OutputStream in EMBL format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeEmbl(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)">writeEmbl</A></B>(java.io.OutputStream&nbsp;os,
          <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a stream of Sequences to an OutputStream in EMBL format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a single Sequence to an OutputStream in Fasta format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.db.SequenceDB)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Write a sequenceDB to an output stream in fasta format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes sequences from a SequenceIterator to an OutputStream in
 Fasta Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeGenbank(java.io.OutputStream, org.biojava.bio.seq.Sequence)">writeGenbank</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a single Sequence to an OutputStream in Genbank format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeGenbank(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)">writeGenbank</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a stream of Sequences to an OutputStream in Genbank
 format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeGenpept(java.io.OutputStream, org.biojava.bio.seq.Sequence)">writeGenpept</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a single Sequence to an OutputStream in Genpept format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeGenpept(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)">writeGenpept</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a stream of Sequences to an OutputStream in Genpept
 format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeSwissprot(java.io.OutputStream, org.biojava.bio.seq.Sequence)">writeSwissprot</A></B>(java.io.OutputStream&nbsp;os,
               <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a single Sequence to an OutputStream in SwissProt format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeSwissprot(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)">writeSwissprot</A></B>(java.io.OutputStream&nbsp;os,
               <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a stream of Sequences to an OutputStream in SwissProt
 format.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getEmblBuilderFactory()"><!-- --></A><H3>
getEmblBuilderFactory</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A> <B>getEmblBuilderFactory</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Get a default SequenceBuilderFactory for handling EMBL
 files.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <CODE>SmartSequenceBuilder.FACTORY</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="readEmbl(java.io.BufferedReader)"><!-- --></A><H3>
readEmbl</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readEmbl</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an EMBL-format stream.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - A reader for the EMBL source or file
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="readEmblRNA(java.io.BufferedReader)"><!-- --></A><H3>
readEmblRNA</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readEmblRNA</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an EMBL-format stream, but for RNA.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - A reader for the EMBL source or file
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="readEmblNucleotide(java.io.BufferedReader)"><!-- --></A><H3>
readEmblNucleotide</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readEmblNucleotide</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an EMBL-format stream.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - A reader for the EMBL source or file
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="getGenbankBuilderFactory()"><!-- --></A><H3>
getGenbankBuilderFactory</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A> <B>getGenbankBuilderFactory</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Get a default SequenceBuilderFactory for handling GenBank
 files.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SmartSequenceBuilder.FACTORY</code></DL>
</DD>
</DL>
<HR>

<A NAME="readGenbank(java.io.BufferedReader)"><!-- --></A><H3>
readGenbank</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readGenbank</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an Genbank-format stream.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - A reader for the Genbank source or file
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="readGenbankXml(java.io.BufferedReader)"><!-- --></A><H3>
readGenbankXml</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readGenbankXml</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an GenbankXML-format stream.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - A reader for the GenbanXML source or file
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="getGenpeptBuilderFactory()"><!-- --></A><H3>
getGenpeptBuilderFactory</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A> <B>getGenpeptBuilderFactory</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Get a default SequenceBuilderFactory for handling Genpept
 files.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SmartSequenceBuilder.FACTORY</code></DL>
</DD>
</DL>
<HR>

<A NAME="readGenpept(java.io.BufferedReader)"><!-- --></A><H3>
readGenpept</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readGenpept</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an Genpept-format stream.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - A reader for the Genpept source or file
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="getSwissprotBuilderFactory()"><!-- --></A><H3>
getSwissprotBuilderFactory</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A> <B>getSwissprotBuilderFactory</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Get a default SequenceBuilderFactory for handling Swissprot
 files.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SmartSequenceBuilder.FACTORY</code></DL>
</DD>
</DL>
<HR>

<A NAME="readSwissprot(java.io.BufferedReader)"><!-- --></A><H3>
readSwissprot</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readSwissprot</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an Swissprot-format stream.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - A reader for the Swissprot source or file
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="getFastaBuilderFactory()"><!-- --></A><H3>
getFastaBuilderFactory</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A> <B>getFastaBuilderFactory</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Get a default SequenceBuilderFactory for handling FASTA
 files.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SmartSequenceBuilder.FACTORY</code></DL>
</DD>
</DL>
<HR>

<A NAME="readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization)"><!-- --></A><H3>
readFasta</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readFasta</B>(java.io.BufferedReader&nbsp;br,
                                         <A HREF="../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Read a fasta file.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the BufferedReader to read data from<DD><CODE>sTok</CODE> - a SymbolTokenization that understands the sequences
<DT><B>Returns:</B><DD>a SequenceIterator over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SequenceBuilderFactory)"><!-- --></A><H3>
readFasta</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readFasta</B>(java.io.BufferedReader&nbsp;br,
                                         <A HREF="../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                         <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;seqFactory)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Read a fasta file using a custom type of SymbolList. For example,
 use SmartSequenceBuilder.FACTORY to emulate readFasta(BufferedReader,
 SymbolTokenization) and SmartSequenceBuilder.BIT_PACKED to force all
 symbols to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the BufferedReader to read data from<DD><CODE>sTok</CODE> - a SymbolTokenization that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a SymbolList
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaDNA(java.io.BufferedReader)"><!-- --></A><H3>
readFastaDNA</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readFastaDNA</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an FASTA-format stream of DNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the BufferedReader to read data from
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaRNA(java.io.BufferedReader)"><!-- --></A><H3>
readFastaRNA</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readFastaRNA</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an FASTA-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the BufferedReader to read data from
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaProtein(java.io.BufferedReader)"><!-- --></A><H3>
readFastaProtein</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A> <B>readFastaProtein</B>(java.io.BufferedReader&nbsp;br)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Iterate over the sequences in an FASTA-format stream of Protein sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the BufferedReader to read data from
<DT><B>Returns:</B><DD>a <CODE>SequenceIterator</CODE> that iterates over each
 <CODE>Sequence</CODE> in the file</DL>
</DD>
</DL>
<HR>

<A NAME="readFasta(java.io.InputStream, org.biojava.bio.symbol.Alphabet)"><!-- --></A><H3>
readFasta</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> <B>readFasta</B>(java.io.InputStream&nbsp;seqFile,
                                   <A HREF="../../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)
                            throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Create a sequence database from a fasta file provided as an
 input stream.  Note this somewhat duplicates functionality in
 the readFastaDNA and readFastaProtein methods but uses a stream
 rather than a reader and returns a SequenceDB rather than a
 SequenceIterator. If the returned DB is likely to be large then
 the above mentioned methods should be used.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>seqFile</CODE> - The file containg the fasta formatted sequences<DD><CODE>alpha</CODE> - The <code>Alphabet</code> of the sequence, ie DNA, RNA etc
<DT><B>Returns:</B><DD>a <code>SequenceDB</code> containing all the <code>Sequences</code>
 in the file.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if problems occur during reading of the
 stream.<DT><B>Since:</B></DT>
  <DD>1.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeFasta(java.io.OutputStream, org.biojava.bio.seq.db.SequenceDB)"><!-- --></A><H3>
writeFasta</H3>
<PRE>
public static void <B>writeFasta</B>(java.io.OutputStream&nbsp;os,
                              <A HREF="../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db)
                       throws java.io.IOException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Write a sequenceDB to an output stream in fasta format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the stream to write the fasta formatted data to.<DD><CODE>db</CODE> - the database of <code>Sequence</code>s to write
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.<DT><B>Since:</B></DT>
  <DD>1.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)"><!-- --></A><H3>
writeFasta</H3>
<PRE>
public static void <B>writeFasta</B>(java.io.OutputStream&nbsp;os,
                              <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)
                       throws java.io.IOException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes sequences from a SequenceIterator to an OutputStream in
 Fasta Format.  This makes for a useful format filter where a
 StreamReader can be sent to the StreamWriter after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input <code>Sequences</code>
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.<DT><B>Since:</B></DT>
  <DD>1.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence)"><!-- --></A><H3>
writeFasta</H3>
<PRE>
public static void <B>writeFasta</B>(java.io.OutputStream&nbsp;os,
                              <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)
                       throws java.io.IOException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a single Sequence to an OutputStream in Fasta format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>seq</CODE> - the Sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeEmbl(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)"><!-- --></A><H3>
writeEmbl</H3>
<PRE>
public static void <B>writeEmbl</B>(java.io.OutputStream&nbsp;os,
                             <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)
                      throws java.io.IOException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a stream of Sequences to an OutputStream in EMBL format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>in</CODE> - a SequenceIterator.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeEmbl(java.io.OutputStream, org.biojava.bio.seq.Sequence)"><!-- --></A><H3>
writeEmbl</H3>
<PRE>
public static void <B>writeEmbl</B>(java.io.OutputStream&nbsp;os,
                             <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)
                      throws java.io.IOException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a single Sequence to an OutputStream in EMBL format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>seq</CODE> - the Sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeSwissprot(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)"><!-- --></A><H3>
writeSwissprot</H3>
<PRE>
public static void <B>writeSwissprot</B>(java.io.OutputStream&nbsp;os,
                                  <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)
                           throws java.io.IOException,
                                  <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a stream of Sequences to an OutputStream in SwissProt
 format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>in</CODE> - a SequenceIterator.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if the <CODE>Sequence</CODE> cannot be converted to SwissProt
 format
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeSwissprot(java.io.OutputStream, org.biojava.bio.seq.Sequence)"><!-- --></A><H3>
writeSwissprot</H3>
<PRE>
public static void <B>writeSwissprot</B>(java.io.OutputStream&nbsp;os,
                                  <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)
                           throws java.io.IOException,
                                  <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a single Sequence to an OutputStream in SwissProt format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>seq</CODE> - the Sequence.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if the <CODE>Sequence</CODE> cannot be written to SwissProt format
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeGenpept(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)"><!-- --></A><H3>
writeGenpept</H3>
<PRE>
public static void <B>writeGenpept</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)
                         throws java.io.IOException,
                                <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a stream of Sequences to an OutputStream in Genpept
 format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>in</CODE> - a SequenceIterator.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if the <CODE>Sequence</CODE> cannot be written to Genpept format
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeGenpept(java.io.OutputStream, org.biojava.bio.seq.Sequence)"><!-- --></A><H3>
writeGenpept</H3>
<PRE>
public static void <B>writeGenpept</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)
                         throws java.io.IOException,
                                <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a single Sequence to an OutputStream in Genpept format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>seq</CODE> - the Sequence.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if the <CODE>Sequence</CODE> cannot be written to Genpept format
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeGenbank(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)"><!-- --></A><H3>
writeGenbank</H3>
<PRE>
public static void <B>writeGenbank</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)
                         throws java.io.IOException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a stream of Sequences to an OutputStream in Genbank
 format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>in</CODE> - a SequenceIterator.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="writeGenbank(java.io.OutputStream, org.biojava.bio.seq.Sequence)"><!-- --></A><H3>
writeGenbank</H3>
<PRE>
public static void <B>writeGenbank</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)
                         throws java.io.IOException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a single Sequence to an OutputStream in Genbank format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the OutputStream.<DD><CODE>seq</CODE> - the Sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there was an error while writing.</DL>
</DD>
</DL>
<HR>

<A NAME="identifyFormat(java.lang.String, java.lang.String)"><!-- --></A><H3>
identifyFormat</H3>
<PRE>
public static int <B>identifyFormat</B>(java.lang.String&nbsp;formatName,
                                 java.lang.String&nbsp;alphabetName)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD><code>identifyFormat</code> performs a case-insensitive mapping
 of a pair of common sequence format name (such as 'embl',
 'genbank' or 'fasta') and alphabet name (such as 'dna', 'rna',
 'protein', 'aa') to an integer. The value returned will be one
 of the public static final fields in
 <code>SeqIOConstants</code>, or a bitwise-or combination of
 them. The method will reject known illegal combinations of
 format and alphabet (such as swissprot + dna) by throwing an
 <code>IllegalArgumentException</code>. It will return the
 <code>SeqIOConstants.UNKNOWN</code> value when either format or
 alphabet are unknown.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>formatName</CODE> - a <code>String</code>.<DD><CODE>alphabetName</CODE> - a <code>String</code>.
<DT><B>Returns:</B><DD>an <code>int</code>.</DL>
</DD>
</DL>
<HR>

<A NAME="getSequenceFormat(int)"><!-- --></A><H3>
getSequenceFormat</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A> <B>getSequenceFormat</B>(int&nbsp;identifier)
                                        throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD><code>getSequenceFormat</code> accepts a value which represents
 a sequence format and returns the relevant
 <code>SequenceFormat</code> object.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>identifier</CODE> - an <code>int</code> which represents a binary
 value with bits set according to the scheme described in
 <code>SeqIOConstants</code>.
<DT><B>Returns:</B><DD>a <code>SequenceFormat</code>.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if an error occurs.</DL>
</DD>
</DL>
<HR>

<A NAME="getBuilderFactory(int)"><!-- --></A><H3>
getBuilderFactory</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A> <B>getBuilderFactory</B>(int&nbsp;identifier)
                                                throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD><code>getBuilderFactory</code> accepts a value which represents
 a sequence format and returns the relevant
 <code>SequenceBuilderFactory</code> object.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>identifier</CODE> - an <code>int</code> which represents a binary
 value with bits set according to the scheme described in
 <code>SeqIOConstants</code>.
<DT><B>Returns:</B><DD>a <code>SequenceBuilderFactory</code>.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if an error occurs.</DL>
</DD>
</DL>
<HR>

<A NAME="getAlphabet(int)"><!-- --></A><H3>
getAlphabet</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A> <B>getAlphabet</B>(int&nbsp;identifier)
                                  throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD><code>getAlphabet</code> accepts a value which represents a
 sequence format and returns the relevant
 <code>FiniteAlphabet</code> object.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>identifier</CODE> - an <code>int</code> which represents a binary
 value with bits set according to the scheme described in
 <code>SeqIOConstants</code>.
<DT><B>Returns:</B><DD>a <code>FiniteAlphabet</code>.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if an error occurs.</DL>
</DD>
</DL>
<HR>

<A NAME="guessFileType(java.io.File)"><!-- --></A><H3>
guessFileType</H3>
<PRE>
public static int <B>guessFileType</B>(java.io.File&nbsp;seqFile)
                         throws java.io.IOException,
                                java.io.FileNotFoundException</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>because there is no standard file naming convention
 and guessing by file name is inherantly error prone and bad.</I>
<P>
<DD>Attempts to guess the filetype of a file given the name.  For
 use with the functions below that take an int fileType as a
 parameter. EMBL and Genbank files are assumed to contain DNA
 sequence.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>seqFile</CODE> - the <CODE>File</CODE> to read from.
<DT><B>Returns:</B><DD>a value that describes the file type.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if <CODE>seqFile</CODE> cannot be read
<DD><CODE>java.io.FileNotFoundException</CODE> - if <CODE>seqFile</CODE> cannot be found</DL>
</DD>
</DL>
<HR>

<A NAME="formatToFactory(org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.symbol.Alphabet)"><!-- --></A><H3>
formatToFactory</H3>
<PRE>
public static <A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A> <B>formatToFactory</B>(<A HREF="../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
                                                     <A HREF="../../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)
                                              throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>as this essentially duplicates the operation
 available in the method <code>identifyBuilderFactory</code>.</I>
<P>
<DD>Attempts to retrieve the most appropriate
 <code>SequenceBuilder</code> object for some combination of
 <code>Alphabet</code> and <code>SequenceFormat</code>
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>format</CODE> - currently supports <code>FastaFormat</code>,
 <code>GenbankFormat</code>, <code>EmblLikeFormat</code><DD><CODE>alpha</CODE> - currently only supports the DNA and Protein
 alphabets
<DT><B>Returns:</B><DD>the <code>SequenceBuilderFactory</code>
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if the combination of alpha and format is
 unrecognized.</DL>
</DD>
</DL>
<HR>

<A NAME="fileToBiojava(java.lang.String, java.lang.String, java.io.BufferedReader)"><!-- --></A><H3>
fileToBiojava</H3>
<PRE>
public static java.lang.Object <B>fileToBiojava</B>(java.lang.String&nbsp;formatName,
                                             java.lang.String&nbsp;alphabetName,
                                             java.io.BufferedReader&nbsp;br)
                                      throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Reads a file with the specified format and alphabet
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>formatName</CODE> - the name of the format eg genbank or
 swissprot (case insensitive)<DD><CODE>alphabetName</CODE> - the name of the alphabet eg dna or rna or
 protein (case insensitive)<DD><CODE>br</CODE> - a BufferedReader for the input
<DT><B>Returns:</B><DD>either an Alignment object or a SequenceIterator
 (depending on the format read)
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if an error occurs while reading or a
 unrecognized format, alphabet combination is used (eg swissprot
 and DNA).<DT><B>Since:</B></DT>
  <DD>1.3</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="fileToBiojava(int, java.io.BufferedReader)"><!-- --></A><H3>
fileToBiojava</H3>
<PRE>
public static java.lang.Object <B>fileToBiojava</B>(int&nbsp;fileType,
                                             java.io.BufferedReader&nbsp;br)
                                      throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Reads a file and returns the corresponding Biojava object. You
 need to cast it as an Alignment or a SequenceIterator as
 appropriate.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>fileType</CODE> - a value that describes the file type<DD><CODE>br</CODE> - the reader for the input
<DT><B>Returns:</B><DD>either a <code>SequenceIterator</code> if the file type is a 
 sequence file, or a <code>Alignment</code> if the file is a sequence
 alignment.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if the file cannot be parsed</DL>
</DD>
</DL>
<HR>

<A NAME="biojavaToFile(java.lang.String, java.lang.String, java.io.OutputStream, java.lang.Object)"><!-- --></A><H3>
biojavaToFile</H3>
<PRE>
public static void <B>biojavaToFile</B>(java.lang.String&nbsp;formatName,
                                 java.lang.String&nbsp;alphabetName,
                                 java.io.OutputStream&nbsp;os,
                                 java.lang.Object&nbsp;biojava)
                          throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A>,
                                 java.io.IOException,
                                 <A HREF="../../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Writes a Biojava <code>SequenceIterator</code>,
 <code>SequenceDB</code>, <code>Sequence</code> or <code>Aligment</code>
 to an <code>OutputStream</code>
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>formatName</CODE> - eg fasta, GenBank (case insensitive)<DD><CODE>alphabetName</CODE> - eg DNA, RNA (case insensititve)<DD><CODE>os</CODE> - where to write to<DD><CODE>biojava</CODE> - the object to write
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - problems getting data from the biojava object.
<DD><CODE>java.io.IOException</CODE> - if there are IO problems
<DD><CODE><A HREF="../../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - a Symbol cannot be parsed</DL>
</DD>
</DL>
<HR>

<A NAME="biojavaToFile(int, java.io.OutputStream, java.lang.Object)"><!-- --></A><H3>
biojavaToFile</H3>
<PRE>
public static void <B>biojavaToFile</B>(int&nbsp;fileType,
                                 java.io.OutputStream&nbsp;os,
                                 java.lang.Object&nbsp;biojava)
                          throws <A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A>,
                                 java.io.IOException,
                                 <A HREF="../../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<DD>Converts a Biojava object to the given filetype.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>fileType</CODE> - a value that describes the type of sequence file<DD><CODE>os</CODE> - the stream to write the formatted results to<DD><CODE>biojava</CODE> - a <code>SequenceIterator</code>, <code>SequenceDB</code>, 
 <code>Sequence</code>, or <code>Alignment</code>
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></CODE> - if <code>biojava</code> cannot be 
 converted to that format.
<DD><CODE>java.io.IOException</CODE> - if the output cannot be written to 
 <code>os</code>
<DD><CODE><A HREF="../../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - if <code>biojava
 </code> contains a <code>Symbol</code> that cannot be understood by the
 parser.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SeqIOTools.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/io/SequenceBuilder.html" title="interface in org.biojava.bio.seq.io"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/seq/io/SeqIOTools.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SeqIOTools.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
