<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:24 BST 2008 -->
<TITLE>
SimilarityPairFeature (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojava.bio.seq.homol.SimilarityPairFeature interface">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="SimilarityPairFeature (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SimilarityPairFeature.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/homol/HomologyFeature.Template.html" title="class in org.biojava.bio.seq.homol"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.EmptyPairwiseAlignment.html" title="class in org.biojava.bio.seq.homol"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/seq/homol/SimilarityPairFeature.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SimilarityPairFeature.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.biojava.bio.seq.homol</FONT>
<BR>
Interface SimilarityPairFeature</H2>
<DL>
<DT><B>All Superinterfaces:</B> <DD><A HREF="../../../../../org/biojava/bio/Annotatable.html" title="interface in org.biojava.bio">Annotatable</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html" title="interface in org.biojava.utils">Changeable</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>, <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html" title="interface in org.biojava.bio.seq">StrandedFeature</A></DD>
</DL>
<DL>
<DT><B>All Known Implementing Classes:</B> <DD><A HREF="../../../../../org/biojava/bio/seq/impl/SimpleSimilarityPairFeature.html" title="class in org.biojava.bio.seq.impl">SimpleSimilarityPairFeature</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public interface <B>SimilarityPairFeature</B><DT>extends <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html" title="interface in org.biojava.bio.seq">StrandedFeature</A></DL>
</PRE>

<P>
<p><code>SimilarityPairFeature</code> describes a pairwise
 similarity between two nucleotide sequences (as it extends
 <code>StrandedFeature</code>). It is analagous to, and based on,
 the BioPerl Bio::SeqFeature::SimilarityPair.</p>

 <p>It is different from <code>HomologyFeature</code> in that it
 expresses a relationship between only two sequence regions (rather
 than >= 2), with one clearly defined as the query sequence and the
 other as the subject (database hit). These are identified by
 constant labels in the
 <code>Alignment</code>. <code>HomologyFeature</code> identifies the
 related sequence regions by means of an <code>Homology</code>
 instance which contains an <code>Alignment</code> which uses the
 <code>HomologyFeature</code>s themselves as labels.</p>

 <p>In cases where there is no alignment available, for example when
 MSPCrunch output or GFF have been used, the
 <code>EmptyPairwiseAlignment</code> in the EMPTY_PAIRWISE field may
 be used. This may also be useful if an implementation elides the
 alignment data for some reason.</p>
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.2</DD>
<DT><B>Author:</B></DT>
  <DD>Keith James</DD>
</DL>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->

<A NAME="nested_class_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Nested Class Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.EmptyPairwiseAlignment.html" title="class in org.biojava.bio.seq.homol">SimilarityPairFeature.EmptyPairwiseAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>EmptyPairwiseAlignment</code> empty pairwise alignment
 which has labels to empty symbol lists.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.Template.html" title="class in org.biojava.bio.seq.homol">SimilarityPairFeature.Template</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>Template</code> for construction of
 <code>SimilarityPairFeature</code>s.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="nested_classes_inherited_from_class_org.biojava.bio.seq.StrandedFeature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Nested classes/interfaces inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html" title="interface in org.biojava.bio.seq">StrandedFeature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.Strand.html" title="class in org.biojava.bio.seq">StrandedFeature.Strand</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<A NAME="nested_classes_inherited_from_class_org.biojava.bio.seq.Feature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Nested classes/interfaces inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/Feature.ByLocationComparator.html" title="class in org.biojava.bio.seq">Feature.ByLocationComparator</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<A NAME="nested_classes_inherited_from_class_org.biojava.bio.seq.FeatureHolder"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Nested classes/interfaces inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.EmptyFeatureHolder.html" title="class in org.biojava.bio.seq">FeatureHolder.EmptyFeatureHolder</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<A NAME="nested_classes_inherited_from_class_org.biojava.bio.Annotatable"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Nested classes/interfaces inherited from interface org.biojava.bio.<A HREF="../../../../../org/biojava/bio/Annotatable.html" title="interface in org.biojava.bio">Annotatable</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/Annotatable.AnnotationForwarder.html" title="class in org.biojava.bio">Annotatable.AnnotationForwarder</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#EMPTY_PAIRWISE">EMPTY_PAIRWISE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constant <code>EMPTY_PAIRWISE</code> is an empty alignment for
 situations where there is no available alignment data or the
 implementation does not want to create one.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#QUERY_LABEL">QUERY_LABEL</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constant <code>QUERY_LABEL</code> is the alignment label used
 for all query sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/utils/ChangeType.html" title="class in org.biojava.utils">ChangeType</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#SIBLING">SIBLING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The sibling of this feature has altered.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#SUBJECT_LABEL">SUBJECT_LABEL</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constant <code>SUBJECT_LABEL</code> is the alignment label used
 for all subject sequences.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.biojava.bio.seq.StrandedFeature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html" title="interface in org.biojava.bio.seq">StrandedFeature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html#NEGATIVE">NEGATIVE</A>, <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html#POSITIVE">POSITIVE</A>, <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html#STRAND">STRAND</A>, <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html#UNKNOWN">UNKNOWN</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.biojava.bio.seq.Feature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/Feature.html#byLocationOrder">byLocationOrder</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#LOCATION">LOCATION</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#PROPERTY_DATA_KEY">PROPERTY_DATA_KEY</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#SOURCE">SOURCE</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#SOURCETERM">SOURCETERM</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#TYPE">TYPE</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#TYPETERM">TYPETERM</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.biojava.bio.seq.FeatureHolder"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#EMPTY_FEATURE_HOLDER">EMPTY_FEATURE_HOLDER</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#FEATURES">FEATURES</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#SCHEMA">SCHEMA</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.biojava.bio.Annotatable"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.biojava.bio.<A HREF="../../../../../org/biojava/bio/Annotatable.html" title="interface in org.biojava.bio">Annotatable</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/Annotatable.html#ANNOTATION">ANNOTATION</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getAlignment</code> returns the <code>Alignment</code> of
 two similar features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#getScore()">getScore</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getScore</code> returns the alignment score.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html" title="interface in org.biojava.bio.seq.homol">SimilarityPairFeature</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#getSibling()">getSibling</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getSibling</code> returns the sibling
 <code>Feature</code>, query for subject and vice versa.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#setSibling(org.biojava.bio.seq.homol.SimilarityPairFeature)">setSibling</A></B>(<A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html" title="interface in org.biojava.bio.seq.homol">SimilarityPairFeature</A>&nbsp;sibling)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>setSibling</code> sets the sibling feature of the
 pair.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.biojava.bio.seq.StrandedFeature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html" title="interface in org.biojava.bio.seq">StrandedFeature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html#getStrand()">getStrand</A>, <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html#getSymbols()">getSymbols</A>, <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.html#setStrand(org.biojava.bio.seq.StrandedFeature.Strand)">setStrand</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.biojava.bio.seq.Feature"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/Feature.html#features()">features</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#getLocation()">getLocation</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#getParent()">getParent</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#getSequence()">getSequence</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#getSource()">getSource</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#getSourceTerm()">getSourceTerm</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#getType()">getType</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#getTypeTerm()">getTypeTerm</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#makeTemplate()">makeTemplate</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#setLocation(org.biojava.bio.symbol.Location)">setLocation</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#setSource(java.lang.String)">setSource</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#setSourceTerm(org.biojava.ontology.Term)">setSourceTerm</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#setType(java.lang.String)">setType</A>, <A HREF="../../../../../org/biojava/bio/seq/Feature.html#setTypeTerm(org.biojava.ontology.Term)">setTypeTerm</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.biojava.bio.seq.FeatureHolder"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface org.biojava.bio.seq.<A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#containsFeature(org.biojava.bio.seq.Feature)">containsFeature</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#countFeatures()">countFeatures</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#filter(org.biojava.bio.seq.FeatureFilter)">filter</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#filter(org.biojava.bio.seq.FeatureFilter, boolean)">filter</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#getSchema()">getSchema</A>, <A HREF="../../../../../org/biojava/bio/seq/FeatureHolder.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.biojava.utils.Changeable"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface org.biojava.utils.<A HREF="../../../../../org/biojava/utils/Changeable.html" title="interface in org.biojava.utils">Changeable</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/utils/Changeable.html#addChangeListener(org.biojava.utils.ChangeListener)">addChangeListener</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#addChangeListener(org.biojava.utils.ChangeListener, org.biojava.utils.ChangeType)">addChangeListener</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#isUnchanging(org.biojava.utils.ChangeType)">isUnchanging</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#removeChangeListener(org.biojava.utils.ChangeListener)">removeChangeListener</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#removeChangeListener(org.biojava.utils.ChangeListener, org.biojava.utils.ChangeType)">removeChangeListener</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.biojava.bio.Annotatable"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface org.biojava.bio.<A HREF="../../../../../org/biojava/bio/Annotatable.html" title="interface in org.biojava.bio">Annotatable</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/bio/Annotatable.html#getAnnotation()">getAnnotation</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.biojava.utils.Changeable"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface org.biojava.utils.<A HREF="../../../../../org/biojava/utils/Changeable.html" title="interface in org.biojava.utils">Changeable</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../../org/biojava/utils/Changeable.html#addChangeListener(org.biojava.utils.ChangeListener)">addChangeListener</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#addChangeListener(org.biojava.utils.ChangeListener, org.biojava.utils.ChangeType)">addChangeListener</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#isUnchanging(org.biojava.utils.ChangeType)">isUnchanging</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#removeChangeListener(org.biojava.utils.ChangeListener)">removeChangeListener</A>, <A HREF="../../../../../org/biojava/utils/Changeable.html#removeChangeListener(org.biojava.utils.ChangeListener, org.biojava.utils.ChangeType)">removeChangeListener</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="SIBLING"><!-- --></A><H3>
SIBLING</H3>
<PRE>
static final <A HREF="../../../../../org/biojava/utils/ChangeType.html" title="class in org.biojava.utils">ChangeType</A> <B>SIBLING</B></PRE>
<DL>
<DD>The sibling of this feature has altered.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="QUERY_LABEL"><!-- --></A><H3>
QUERY_LABEL</H3>
<PRE>
static final java.lang.String <B>QUERY_LABEL</B></PRE>
<DL>
<DD>Constant <code>QUERY_LABEL</code> is the alignment label used
 for all query sequences.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../../constant-values.html#org.biojava.bio.seq.homol.SimilarityPairFeature.QUERY_LABEL">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="SUBJECT_LABEL"><!-- --></A><H3>
SUBJECT_LABEL</H3>
<PRE>
static final java.lang.String <B>SUBJECT_LABEL</B></PRE>
<DL>
<DD>Constant <code>SUBJECT_LABEL</code> is the alignment label used
 for all subject sequences.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../../constant-values.html#org.biojava.bio.seq.homol.SimilarityPairFeature.SUBJECT_LABEL">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="EMPTY_PAIRWISE"><!-- --></A><H3>
EMPTY_PAIRWISE</H3>
<PRE>
static final <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> <B>EMPTY_PAIRWISE</B></PRE>
<DL>
<DD>Constant <code>EMPTY_PAIRWISE</code> is an empty alignment for
 situations where there is no available alignment data or the
 implementation does not want to create one.
<P>
<DL>
</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getSibling()"><!-- --></A><H3>
getSibling</H3>
<PRE>
<A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html" title="interface in org.biojava.bio.seq.homol">SimilarityPairFeature</A> <B>getSibling</B>()</PRE>
<DL>
<DD><code>getSibling</code> returns the sibling
 <code>Feature</code>, query for subject and vice versa.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>Feature</code>.</DL>
</DD>
</DL>
<HR>

<A NAME="setSibling(org.biojava.bio.seq.homol.SimilarityPairFeature)"><!-- --></A><H3>
setSibling</H3>
<PRE>
void <B>setSibling</B>(<A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html" title="interface in org.biojava.bio.seq.homol">SimilarityPairFeature</A>&nbsp;sibling)
                throws <A HREF="../../../../../org/biojava/utils/ChangeVetoException.html" title="class in org.biojava.utils">ChangeVetoException</A></PRE>
<DL>
<DD><code>setSibling</code> sets the sibling feature of the
 pair. This is used to set the reciprocal
 <code>SimilarityPairFeature</code> as both cannot be set using
 the <code>Template</code>.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../../org/biojava/utils/ChangeVetoException.html" title="class in org.biojava.utils">ChangeVetoException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getAlignment()"><!-- --></A><H3>
getAlignment</H3>
<PRE>
<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> <B>getAlignment</B>()</PRE>
<DL>
<DD><code>getAlignment</code> returns the <code>Alignment</code> of
 two similar features.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>an <code>Alignment</code> value.</DL>
</DD>
</DL>
<HR>

<A NAME="getScore()"><!-- --></A><H3>
getScore</H3>
<PRE>
double <B>getScore</B>()</PRE>
<DL>
<DD><code>getScore</code> returns the alignment score.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>double</code>.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SimilarityPairFeature.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/homol/HomologyFeature.Template.html" title="class in org.biojava.bio.seq.homol"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.EmptyPairwiseAlignment.html" title="class in org.biojava.bio.seq.homol"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/seq/homol/SimilarityPairFeature.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SimilarityPairFeature.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
