<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:31 BST 2008 -->
<TITLE>
ChainImpl (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojava.bio.structure.ChainImpl class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="ChainImpl (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/ChainImpl.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/structure/Compound.html" title="class in org.biojava.bio.structure"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/structure/ChainImpl.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="ChainImpl.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.biojava.bio.structure</FONT>
<BR>
Class ChainImpl</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>org.biojava.bio.structure.ChainImpl</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>ChainImpl</B><DT>extends java.lang.Object<DT>implements <A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></DL>
</PRE>

<P>
A Chain in a PDB file. It contains several groups which can be of
 type "amino", "hetatm", "nucleotide".
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.4</DD>
<DT><B>Author:</B></DT>
  <DD>Andreas Prlic, Jules Jacobsen</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#DEFAULT_CHAIN_ID">DEFAULT_CHAIN_ID</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The default chain identifier is an empty space.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#ChainImpl()">ChainImpl</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a ChainImpl object.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#addGroup(org.biojava.bio.structure.Group)">addGroup</A></B>(<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&nbsp;group)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;add a group to the list of ATOM record group of this chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#clone()">clone</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an identical copy of this Chain .</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getAnnotation()">getAnnotation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get/set the Annotation of a Chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getAtomGroup(int)">getAtomGroup</A></B>(int&nbsp;position)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return the group at position .</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getAtomGroups()">getAtomGroups</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return all groups that have been specified in the ATOM section of this chain .</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getAtomGroups(java.lang.String)">getAtomGroups</A></B>(java.lang.String&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the Groups of a certain type, that are listed int the ATOM records of the PDB file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getAtomLength()">getAtomLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the number of Groups in the ATOM records of the chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getAtomSequence()">getAtomSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the sequence of amino acids as it has been provided in the ATOM records.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getBJSequence()">getBJSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convert the SEQRES groups of a Chain to a Biojava Sequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getGroup(int)">getGroup</A></B>(int&nbsp;position)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use getAtomGroup or getSeqResGroup instead</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getGroupByPDB(java.lang.String)">getGroupByPDB</A></B>(java.lang.String&nbsp;pdbresnum)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get a group by its PDB residue numbering.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getGroups()">getGroups</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use getAtomGroups instead</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getGroups(java.lang.String)">getGroups</A></B>(java.lang.String&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use getAtomGroups instead</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getGroupsByPDB(java.lang.String, java.lang.String)">getGroupsByPDB</A></B>(java.lang.String&nbsp;pdbresnumStart,
               java.lang.String&nbsp;pdbresnumEnd)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get all groups that are located between two PDB residue numbers.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getGroupsByPDB(java.lang.String, java.lang.String, boolean)">getGroupsByPDB</A></B>(java.lang.String&nbsp;pdbresnumStart,
               java.lang.String&nbsp;pdbresnumEnd,
               boolean&nbsp;ignoreMissing)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get all groups that are located between two PDB residue numbers.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Compound.html" title="class in org.biojava.bio.structure">Compound</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getHeader()">getHeader</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the Header ( a Compound object) for this chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getId()">getId</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the ID used by Hibernate.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getLength()">getLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use getAtomLength instead</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getLengthAminos()">getLengthAminos</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;returns the length of the AminoAcids in the ATOM records of this chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getName()">getName</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get and set the name of this chain (Chain id in PDB file ).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Structure.html" title="interface in org.biojava.bio.structure">Structure</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getParent()">getParent</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the parent Structure of this chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSeqResGroup(int)">getSeqResGroup</A></B>(int&nbsp;position)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return the Group at position X.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSeqResGroups()">getSeqResGroups</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return all groups of this chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSeqResGroups(java.lang.String)">getSeqResGroups</A></B>(java.lang.String&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return a List of all groups of a special type (e.g.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSeqResLength()">getSeqResLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of groups in the SEQRES records of the chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSeqResSequence()">getSeqResSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the sequence for all amino acids as it is specified in the SEQRES residues.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSequence()">getSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use getAtomSequence instead</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSwissprotId()">getSwissprotId</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get the Swissprot id of this chains .</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setAnnotation(org.biojava.bio.Annotation)">setAnnotation</A></B>(<A HREF="../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A>&nbsp;anno)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get/set the Annotation of a Chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setAtomGroups(java.util.List)">setAtomGroups</A></B>(java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;&nbsp;groups)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set all groups that have been specified in the ATOM section of this chain .</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setHeader(org.biojava.bio.structure.Compound)">setHeader</A></B>(<A HREF="../../../../org/biojava/bio/structure/Compound.html" title="class in org.biojava.bio.structure">Compound</A>&nbsp;mol)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set the Header from the PDB file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setId(java.lang.Long)">setId</A></B>(java.lang.Long&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set the ID used by Hibernate.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;nam)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get and set the name of this chain (Chain id in PDB file ).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setParent(org.biojava.bio.structure.Structure)">setParent</A></B>(<A HREF="../../../../org/biojava/bio/structure/Structure.html" title="interface in org.biojava.bio.structure">Structure</A>&nbsp;parent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set the back-reference to its parent Structure.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setSeqResGroups(java.util.List)">setSeqResGroups</A></B>(java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;&nbsp;groups)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set the list of SeqResGroups for this chain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setSwissprotId(java.lang.String)">setSwissprotId</A></B>(java.lang.String&nbsp;sp_id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;set the Swissprot id of this chains .</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;String representation.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="DEFAULT_CHAIN_ID"><!-- --></A><H3>
DEFAULT_CHAIN_ID</H3>
<PRE>
public static java.lang.String <B>DEFAULT_CHAIN_ID</B></PRE>
<DL>
<DD>The default chain identifier is an empty space.
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="ChainImpl()"><!-- --></A><H3>
ChainImpl</H3>
<PRE>
public <B>ChainImpl</B>()</PRE>
<DL>
<DD>Constructs a ChainImpl object.
<P>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getId()"><!-- --></A><H3>
getId</H3>
<PRE>
public java.lang.Long <B>getId</B>()</PRE>
<DL>
<DD>Get the ID used by Hibernate.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getId()">getId</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the ID used by Hibernate<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setId(java.lang.Long)"><CODE>Chain.setId(Long)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setId(java.lang.Long)"><!-- --></A><H3>
setId</H3>
<PRE>
public void <B>setId</B>(java.lang.Long&nbsp;id)</PRE>
<DL>
<DD>Set the ID used by Hibernate.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setId(java.lang.Long)">setId</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>id</CODE> - assigned by Hibernate<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getId()"><CODE>Chain.getId()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setParent(org.biojava.bio.structure.Structure)"><!-- --></A><H3>
setParent</H3>
<PRE>
public void <B>setParent</B>(<A HREF="../../../../org/biojava/bio/structure/Structure.html" title="interface in org.biojava.bio.structure">Structure</A>&nbsp;parent)</PRE>
<DL>
<DD>Set the back-reference to its parent Structure.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setParent(org.biojava.bio.structure.Structure)">setParent</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>parent</CODE> - the parent Structure object for this Chain<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getParent()"><CODE>Chain.getParent()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getParent()"><!-- --></A><H3>
getParent</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Structure.html" title="interface in org.biojava.bio.structure">Structure</A> <B>getParent</B>()</PRE>
<DL>
<DD>Returns the parent Structure of this chain.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getParent()">getParent</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the parent Structure object<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setParent(org.biojava.bio.structure.Structure)"><CODE>Chain.setParent(Structure)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="clone()"><!-- --></A><H3>
clone</H3>
<PRE>
public java.lang.Object <B>clone</B>()</PRE>
<DL>
<DD>Returns an identical copy of this Chain .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#clone()">clone</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE><DT><B>Overrides:</B><DD><CODE>clone</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>an identical copy of this Chain</DL>
</DD>
</DL>
<HR>

<A NAME="setAnnotation(org.biojava.bio.Annotation)"><!-- --></A><H3>
setAnnotation</H3>
<PRE>
public void <B>setAnnotation</B>(<A HREF="../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A>&nbsp;anno)</PRE>
<DL>
<DD>get/set the Annotation of a Chain.
  allows to annotate a protein chain, e.g. molecule description "AZURIN" for pdb 1a4a.A
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setAnnotation(org.biojava.bio.Annotation)">setAnnotation</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>anno</CODE> - the Annotation to be provided.<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAnnotation()"><CODE>Chain.getAnnotation()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getAnnotation()"><!-- --></A><H3>
getAnnotation</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A> <B>getAnnotation</B>()</PRE>
<DL>
<DD>get/set the Annotation of a Chain.
  allows to annotate a protein chain, e.g. molecule description "AZURIN" for pdb 1a4a.A
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAnnotation()">getAnnotation</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the Annotation of this chain<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setAnnotation(org.biojava.bio.Annotation)"><CODE>Chain.setAnnotation(Annotation)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setHeader(org.biojava.bio.structure.Compound)"><!-- --></A><H3>
setHeader</H3>
<PRE>
public void <B>setHeader</B>(<A HREF="../../../../org/biojava/bio/structure/Compound.html" title="class in org.biojava.bio.structure">Compound</A>&nbsp;mol)</PRE>
<DL>
<DD>Set the Header from the PDB file.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setHeader(org.biojava.bio.structure.Compound)">setHeader</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>mol</CODE> - the Compound that contains the header information for this chain<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getHeader()"><CODE>Chain.getHeader()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getHeader()"><!-- --></A><H3>
getHeader</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Compound.html" title="class in org.biojava.bio.structure">Compound</A> <B>getHeader</B>()</PRE>
<DL>
<DD>Returns the Header ( a Compound object) for this chain.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getHeader()">getHeader</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the Compound object that contains the Header information<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setHeader(org.biojava.bio.structure.Compound)"><CODE>Chain.setHeader(Compound)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setSwissprotId(java.lang.String)"><!-- --></A><H3>
setSwissprotId</H3>
<PRE>
public void <B>setSwissprotId</B>(java.lang.String&nbsp;sp_id)</PRE>
<DL>
<DD>set the Swissprot id of this chains .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setSwissprotId(java.lang.String)">setSwissprotId</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>sp_id</CODE> - a String specifying the swissprot id value<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSwissprotId()"><CODE>getSwissprotId()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getSwissprotId()"><!-- --></A><H3>
getSwissprotId</H3>
<PRE>
public java.lang.String <B>getSwissprotId</B>()</PRE>
<DL>
<DD>get the Swissprot id of this chains .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSwissprotId()">getSwissprotId</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a String representing the swissprot id value<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setSwissprotId(java.lang.String)"><CODE>setSwissprotId(java.lang.String)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="addGroup(org.biojava.bio.structure.Group)"><!-- --></A><H3>
addGroup</H3>
<PRE>
public void <B>addGroup</B>(<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&nbsp;group)</PRE>
<DL>
<DD>add a group to the list of ATOM record group of this chain.
 To add SEQRES records a more complex alignment between ATOM and SEQRES residues
 is required, please see SeqRes2AtomAligner for more details on that.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#addGroup(org.biojava.bio.structure.Group)">addGroup</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>group</CODE> - a Group object</DL>
</DD>
</DL>
<HR>

<A NAME="getGroup(int)"><!-- --></A><H3>
getGroup</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A> <B>getGroup</B>(int&nbsp;position)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>use getAtomGroup or getSeqResGroup instead</I>
<P>
<DD>return the group at position .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getGroup(int)">getGroup</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>position</CODE> - an int
<DT><B>Returns:</B><DD>a Group object<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomGroup(int)"><CODE>Chain.getAtomGroup(int)</CODE></A>, 
<A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResGroup(int)"><CODE>Chain.getSeqResGroup(int)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getAtomGroup(int)"><!-- --></A><H3>
getAtomGroup</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A> <B>getAtomGroup</B>(int&nbsp;position)</PRE>
<DL>
<DD>return the group at position .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomGroup(int)">getAtomGroup</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>position</CODE> - an int
<DT><B>Returns:</B><DD>a Group object</DL>
</DD>
</DL>
<HR>

<A NAME="getGroups(java.lang.String)"><!-- --></A><H3>
getGroups</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt; <B>getGroups</B>(java.lang.String&nbsp;type)</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>use getAtomGroups instead</I>
<P>
<DD>return an array of all groups of a special type (e.g. amino,
 hetatm, nucleotide).
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getGroups(java.lang.String)">getGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>type</CODE> - a String
<DT><B>Returns:</B><DD>an List object containing the groups of type...</DL>
</DD>
</DL>
<HR>

<A NAME="getAtomGroups(java.lang.String)"><!-- --></A><H3>
getAtomGroups</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt; <B>getAtomGroups</B>(java.lang.String&nbsp;type)</PRE>
<DL>
<DD>Get the Groups of a certain type, that are listed int the ATOM records of the PDB file.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomGroups(java.lang.String)">getAtomGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>type</CODE> - the type of the groups to return.
<DT><B>Returns:</B><DD>a list of group objects<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setAtomGroups(java.util.List)"><CODE>Chain.setAtomGroups(List)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getGroups()"><!-- --></A><H3>
getGroups</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt; <B>getGroups</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>use getAtomGroups instead</I>
<P>
<DD>return all groups of this chain .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getGroups()">getGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>an ArrayList object representing the Groups of this Chain.</DL>
</DD>
</DL>
<HR>

<A NAME="getAtomGroups()"><!-- --></A><H3>
getAtomGroups</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt; <B>getAtomGroups</B>()</PRE>
<DL>
<DD>Return all groups that have been specified in the ATOM section of this chain .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomGroups()">getAtomGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a List object representing the Groups of this Chain.<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setAtomGroups(java.util.List)"><CODE>Chain.setAtomGroups(List)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setAtomGroups(java.util.List)"><!-- --></A><H3>
setAtomGroups</H3>
<PRE>
public void <B>setAtomGroups</B>(java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;&nbsp;groups)</PRE>
<DL>
<DD>Set all groups that have been specified in the ATOM section of this chain .
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setAtomGroups(java.util.List)">setAtomGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>groups</CODE> - a List object representing the Groups of this Chain.<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomGroups()"><CODE>Chain.getAtomGroups()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getGroupsByPDB(java.lang.String, java.lang.String, boolean)"><!-- --></A><H3>
getGroupsByPDB</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>[] <B>getGroupsByPDB</B>(java.lang.String&nbsp;pdbresnumStart,
                              java.lang.String&nbsp;pdbresnumEnd,
                              boolean&nbsp;ignoreMissing)
                       throws <A HREF="../../../../org/biojava/bio/structure/StructureException.html" title="class in org.biojava.bio.structure">StructureException</A></PRE>
<DL>
<DD>Get all groups that are located between two PDB residue numbers. In contrast to getGroupsByPDB
 this method call ignores if the exact outer groups are not found. This is useful e.g. when requesting the range
 of groups as specified by the DBREF records - these frequently are rather inaccurate.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getGroupsByPDB(java.lang.String, java.lang.String, boolean)">getGroupsByPDB</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>pdbresnumStart</CODE> - PDB residue number of start<DD><CODE>pdbresnumEnd</CODE> - PDB residue number of end<DD><CODE>ignoreMissing</CODE> - ignore missing groups in this range.
<DT><B>Returns:</B><DD>Groups in between. or throws a StructureException if either start or end can not be found,
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/structure/StructureException.html" title="class in org.biojava.bio.structure">StructureException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getGroupByPDB(java.lang.String)"><!-- --></A><H3>
getGroupByPDB</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A> <B>getGroupByPDB</B>(java.lang.String&nbsp;pdbresnum)
                    throws <A HREF="../../../../org/biojava/bio/structure/StructureException.html" title="class in org.biojava.bio.structure">StructureException</A></PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getGroupByPDB(java.lang.String)">Chain</A></CODE></B></DD>
<DD>get a group by its PDB residue numbering. if the PDB residue number is not know,
 throws a StructureException.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getGroupByPDB(java.lang.String)">getGroupByPDB</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>pdbresnum</CODE> - the PDB residue number of the group
<DT><B>Returns:</B><DD>the matching group
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/structure/StructureException.html" title="class in org.biojava.bio.structure">StructureException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getGroupsByPDB(java.lang.String, java.lang.String)"><!-- --></A><H3>
getGroupsByPDB</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>[] <B>getGroupsByPDB</B>(java.lang.String&nbsp;pdbresnumStart,
                              java.lang.String&nbsp;pdbresnumEnd)
                       throws <A HREF="../../../../org/biojava/bio/structure/StructureException.html" title="class in org.biojava.bio.structure">StructureException</A></PRE>
<DL>
<DD>Get all groups that are located between two PDB residue numbers.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getGroupsByPDB(java.lang.String, java.lang.String)">getGroupsByPDB</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>pdbresnumStart</CODE> - PDB residue number of start<DD><CODE>pdbresnumEnd</CODE> - PDB residue number of end
<DT><B>Returns:</B><DD>Groups in between. or throws a StructureException if either start or end can not be found,
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/structure/StructureException.html" title="class in org.biojava.bio.structure">StructureException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getLength()"><!-- --></A><H3>
getLength</H3>
<PRE>
public int <B>getLength</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>use getAtomLength instead</I>
<P>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getLength()">Chain</A></CODE></B></DD>
<DD>get total length of chain, including HETATMs..
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getLength()">getLength</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>an int representing the length of the whole chain including HETATMs<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomLength()"><CODE>Chain.getAtomLength()</CODE></A>, 
<A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResLength()"><CODE>Chain.getSeqResLength()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getLengthAminos()"><!-- --></A><H3>
getLengthAminos</H3>
<PRE>
public int <B>getLengthAminos</B>()</PRE>
<DL>
<DD>returns the length of the AminoAcids in the ATOM records of this chain.
 note: not all amino acids need to have 3D coords, in fact in could be that none
 has!
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getLengthAminos()">getLengthAminos</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>an int representing the length of the AminoAcids in the ATOM records of the chain.</DL>
</DD>
</DL>
<HR>

<A NAME="getSeqResLength()"><!-- --></A><H3>
getSeqResLength</H3>
<PRE>
public int <B>getSeqResLength</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResLength()">Chain</A></CODE></B></DD>
<DD>Returns the number of groups in the SEQRES records of the chain.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResLength()">getSeqResLength</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the length</DL>
</DD>
</DL>
<HR>

<A NAME="setName(java.lang.String)"><!-- --></A><H3>
setName</H3>
<PRE>
public void <B>setName</B>(java.lang.String&nbsp;nam)</PRE>
<DL>
<DD>get and set the name of this chain (Chain id in PDB file ).
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setName(java.lang.String)">setName</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nam</CODE> - a String specifying the name value<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getName()"><CODE>getName()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getName()"><!-- --></A><H3>
getName</H3>
<PRE>
public java.lang.String <B>getName</B>()</PRE>
<DL>
<DD>get and set the name of this chain (Chain id in PDB file ).
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getName()">getName</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a String representing the name value<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#setName(java.lang.String)"><CODE>setName(java.lang.String)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD>String representation.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#toString()">toString</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE><DT><B>Overrides:</B><DD><CODE>toString</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>String representation of the Chain</DL>
</DD>
</DL>
<HR>

<A NAME="getBJSequence()"><!-- --></A><H3>
getBJSequence</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A> <B>getBJSequence</B>()
                       throws <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD>Convert the SEQRES groups of a Chain to a Biojava Sequence object.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getBJSequence()">getBJSequence</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the SEQRES groups of the Chain as a Sequence object.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="getSequence()"><!-- --></A><H3>
getSequence</H3>
<PRE>
public java.lang.String <B>getSequence</B>()</PRE>
<DL>
<DD><B>Deprecated.</B>&nbsp;<I>use getAtomSequence instead</I>
<P>
<DD>get amino acid sequence of the chain. for backwards compatibility this returns
 the Atom sequence of the chain.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSequence()">getSequence</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a String representing the sequence.<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getAtomSequence()"><CODE>getAtomSequence()</CODE></A>, 
<A HREF="../../../../org/biojava/bio/structure/ChainImpl.html#getSeqResSequence()"><CODE>getSeqResSequence()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getAtomSequence()"><!-- --></A><H3>
getAtomSequence</H3>
<PRE>
public java.lang.String <B>getAtomSequence</B>()</PRE>
<DL>
<DD>Return the sequence of amino acids as it has been provided in the ATOM records.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomSequence()">getAtomSequence</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>amino acid sequence as string<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResSequence()"><CODE>Chain.getSeqResSequence()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getSeqResSequence()"><!-- --></A><H3>
getSeqResSequence</H3>
<PRE>
public java.lang.String <B>getSeqResSequence</B>()</PRE>
<DL>
<DD>Get the sequence for all amino acids as it is specified in the SEQRES residues.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResSequence()">getSeqResSequence</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the amino acid sequence of the SEQRES sequence as a string<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomSequence()"><CODE>Chain.getAtomSequence()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getSeqResGroup(int)"><!-- --></A><H3>
getSeqResGroup</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A> <B>getSeqResGroup</B>(int&nbsp;position)</PRE>
<DL>
<DD>return the Group at position X.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResGroup(int)">getSeqResGroup</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>position</CODE> - an int
<DT><B>Returns:</B><DD>a Group object</DL>
</DD>
</DL>
<HR>

<A NAME="getSeqResGroups(java.lang.String)"><!-- --></A><H3>
getSeqResGroups</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt; <B>getSeqResGroups</B>(java.lang.String&nbsp;type)</PRE>
<DL>
<DD>Return a List of all groups of a special type (e.g. amino,
 hetatm, nucleotide).
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResGroups(java.lang.String)">getSeqResGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>type</CODE> - a String
<DT><B>Returns:</B><DD>an List object<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setSeqResGroups(java.util.List)"><CODE>Chain.setSeqResGroups(List)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getSeqResGroups()"><!-- --></A><H3>
getSeqResGroups</H3>
<PRE>
public java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt; <B>getSeqResGroups</B>()</PRE>
<DL>
<DD>Return all groups of this chain.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResGroups()">getSeqResGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a List of all Group objects of this chain<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#setSeqResGroups(java.util.List)"><CODE>Chain.setSeqResGroups(List)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setSeqResGroups(java.util.List)"><!-- --></A><H3>
setSeqResGroups</H3>
<PRE>
public void <B>setSeqResGroups</B>(java.util.List&lt;<A HREF="../../../../org/biojava/bio/structure/Group.html" title="interface in org.biojava.bio.structure">Group</A>&gt;&nbsp;groups)</PRE>
<DL>
<DD>Set the list of SeqResGroups for this chain.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#setSeqResGroups(java.util.List)">setSeqResGroups</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>groups</CODE> - a List of Group objects that from the SEQRES groups of this chain.<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/structure/Chain.html#getSeqResGroups()"><CODE>Chain.getSeqResGroups()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getAtomLength()"><!-- --></A><H3>
getAtomLength</H3>
<PRE>
public int <B>getAtomLength</B>()</PRE>
<DL>
<DD>Return the number of Groups in the ATOM records of the chain.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html#getAtomLength()">getAtomLength</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure">Chain</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the length</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/ChainImpl.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/structure/Chain.html" title="interface in org.biojava.bio.structure"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/structure/Compound.html" title="class in org.biojava.bio.structure"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/structure/ChainImpl.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="ChainImpl.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
