<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:42 BST 2008 -->
<TITLE>
RichSequence.IOTools (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojavax.bio.seq.RichSequence.IOTools class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="RichSequence.IOTools (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/RichSequence.IOTools.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.SingleRichSeqIterator.html" title="class in org.biojavax.bio.seq"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojavax/bio/seq/RichSequence.IOTools.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="RichSequence.IOTools.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.biojavax.bio.seq</FONT>
<BR>
Class RichSequence.IOTools</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>org.biojavax.bio.seq.RichSequence.IOTools</B>
</PRE>
<DL>
<DT><B>Enclosing interface:</B><DD><A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public static final class <B>RichSequence.IOTools</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
A set of convenience methods for handling common file formats.
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.5</DD>
<DT><B>Author:</B></DT>
  <DD>Mark Schreiber, Richard Holland</DD>
</DL>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->

<A NAME="nested_class_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Nested Class Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.SingleRichSeqIterator.html" title="class in org.biojavax.bio.seq">RichSequence.IOTools.SingleRichSeqIterator</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Used to iterate over a single rich sequence</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getDNAParser()">getDNAParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a DNA symbol tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getNucleotideParser()">getNucleotideParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a nucleotide symbol tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getProteinParser()">getProteinParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a protein symbol tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getRNAParser()">getRNAParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a RNA symbol tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBL(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readEMBL</A></B>(java.io.BufferedReader&nbsp;br,
         <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
         <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
         <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a EMBL file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLDNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLDNA</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBL-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLProtein(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLProtein</A></B>(java.io.BufferedReader&nbsp;br,
                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBL-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLRNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLRNA</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBL-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxml(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readEMBLxml</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a EMBLxml file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxmlDNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLxmlDNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBLxml-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxmlProtein(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLxmlProtein</A></B>(java.io.BufferedReader&nbsp;br,
                   <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBLxml-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxmlRNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLxmlRNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBLxml-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.Namespace)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a fasta file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
          <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a fasta file building a custom type of <code>RichSequence</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFastaDNA(java.io.BufferedReader, org.biojavax.Namespace)">readFastaDNA</A></B>(java.io.BufferedReader&nbsp;br,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an FASTA-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFastaProtein(java.io.BufferedReader, org.biojavax.Namespace)">readFastaProtein</A></B>(java.io.BufferedReader&nbsp;br,
                 <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an FASTA-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFastaRNA(java.io.BufferedReader, org.biojavax.Namespace)">readFastaRNA</A></B>(java.io.BufferedReader&nbsp;br,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an FASTA-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFile(java.io.File, org.biojavax.Namespace)">readFile</A></B>(java.io.File&nbsp;file,
         <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a file is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFile(java.io.File, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readFile</A></B>(java.io.File&nbsp;file,
         <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
         <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a file is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbank(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readGenbank</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a GenBank file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbankDNA(java.io.BufferedReader, org.biojavax.Namespace)">readGenbankDNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an GenBank-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbankProtein(java.io.BufferedReader, org.biojavax.Namespace)">readGenbankProtein</A></B>(java.io.BufferedReader&nbsp;br,
                   <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an GenBank-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbankRNA(java.io.BufferedReader, org.biojavax.Namespace)">readGenbankRNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an GenBank-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseq(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readINSDseq</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a INSDseq file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseqDNA(java.io.BufferedReader, org.biojavax.Namespace)">readINSDseqDNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an INSDseq-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseqProtein(java.io.BufferedReader, org.biojavax.Namespace)">readINSDseqProtein</A></B>(java.io.BufferedReader&nbsp;br,
                   <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an INSDseq-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseqRNA(java.io.BufferedReader, org.biojavax.Namespace)">readINSDseqRNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an INSDseq-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readStream(java.io.BufferedInputStream, org.biojavax.Namespace)">readStream</A></B>(java.io.BufferedInputStream&nbsp;stream,
           <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a stream is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readStream(java.io.BufferedInputStream, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readStream</A></B>(java.io.BufferedInputStream&nbsp;stream,
           <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
           <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a stream is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProt(java.io.BufferedReader, org.biojavax.Namespace)">readUniProt</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an UniProt-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProt(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readUniProt</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a UniProt file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProtXML(java.io.BufferedReader, org.biojavax.Namespace)">readUniProtXML</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an UniProt XML-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProtXML(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readUniProtXML</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
               <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a UniProt XML file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#registerFormat(java.lang.Class)">registerFormat</A></B>(java.lang.Class&nbsp;formatClass)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Register a new format with IOTools for auto-guessing.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBL(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeEMBL</A></B>(java.io.OutputStream&nbsp;os,
          <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
          <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 EMBL Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBL(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeEMBL</A></B>(java.io.OutputStream&nbsp;os,
          <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
          <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in EMBL format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBLxml(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeEMBLxml</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 EMBLxml Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBLxml(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeEMBLxml</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in EMBLxml format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
           <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes <CODE>Sequence</CODE>s from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 Fasta Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace, org.biojavax.bio.seq.io.FastaHeader)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
           <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
           <A HREF="../../../../org/biojavax/bio/seq/io/FastaHeader.html" title="class in org.biojavax.bio.seq.io">FastaHeader</A>&nbsp;header)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes <CODE>Sequence</CODE>s from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 Fasta Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
           <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in Fasta format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace, org.biojavax.bio.seq.io.FastaHeader)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
           <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
           <A HREF="../../../../org/biojavax/bio/seq/io/FastaHeader.html" title="class in org.biojavax.bio.seq.io">FastaHeader</A>&nbsp;header)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in Fasta format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeGenbank(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeGenbank</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 GenBank Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeGenbank(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeGenbank</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in GenBank format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeINSDseq(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeINSDseq</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 INSDseq Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeINSDseq(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeINSDseq</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in INSDseq format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProt(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeUniProt</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 UniProt Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProt(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeUniProt</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in UniProt format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProtXML(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeUniProtXML</A></B>(java.io.OutputStream&nbsp;os,
                <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 UniProt XML Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProtXML(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeUniProtXML</A></B>(java.io.OutputStream&nbsp;os,
                <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in UniProt XML format.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="registerFormat(java.lang.Class)"><!-- --></A><H3>
registerFormat</H3>
<PRE>
public static void <B>registerFormat</B>(java.lang.Class&nbsp;formatClass)</PRE>
<DL>
<DD>Register a new format with IOTools for auto-guessing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>formatClass</CODE> - the <code>RichSequenceFormat</code> object to register.</DL>
</DD>
</DL>
<HR>

<A NAME="readStream(java.io.BufferedInputStream, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readStream</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readStream</B>(java.io.BufferedInputStream&nbsp;stream,
                                              <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                              <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                                       throws java.io.IOException</PRE>
<DL>
<DD>Guess which format a stream is then attempt to read it.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>stream</CODE> - the <code>BufferedInputStream</code> to attempt to read.<DD><CODE>seqFactory</CODE> - a factory used to build a <code>RichSequence</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then
              <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code>
              over each sequence in the file
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - in case the stream is unrecognisable or problems occur in reading it.</DL>
</DD>
</DL>
<HR>

<A NAME="readStream(java.io.BufferedInputStream, org.biojavax.Namespace)"><!-- --></A><H3>
readStream</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readStream</B>(java.io.BufferedInputStream&nbsp;stream,
                                              <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                                       throws java.io.IOException</PRE>
<DL>
<DD>Guess which format a stream is then attempt to read it.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>stream</CODE> - the <code>BufferedInputStream</code> to attempt to read.<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then
              <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code>
              over each sequence in the file
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - If the file cannot be read.</DL>
</DD>
</DL>
<HR>

<A NAME="readFile(java.io.File, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readFile</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readFile</B>(java.io.File&nbsp;file,
                                            <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                                     throws java.io.IOException</PRE>
<DL>
<DD>Guess which format a file is then attempt to read it.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - the <code>File</code> to attempt to read.<DD><CODE>seqFactory</CODE> - a factory used to build a <code>RichSequence</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then
              <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code>
              over each sequence in the file
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - in case the file is unrecognisable or problems occur in reading it.</DL>
</DD>
</DL>
<HR>

<A NAME="readFile(java.io.File, org.biojavax.Namespace)"><!-- --></A><H3>
readFile</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readFile</B>(java.io.File&nbsp;file,
                                            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                                     throws java.io.IOException</PRE>
<DL>
<DD>Guess which format a file is then attempt to read it.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - the <code>File</code> to attempt to read.<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then
              <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code>
              over each sequence in the file
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - If the file cannot be read.</DL>
</DD>
</DL>
<HR>

<A NAME="readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.Namespace)"><!-- --></A><H3>
readFasta</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readFasta</B>(java.io.BufferedReader&nbsp;br,
                                             <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a fasta file.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader<code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then
              <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code>
              over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readFasta</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readFasta</B>(java.io.BufferedReader&nbsp;br,
                                             <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                             <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a fasta file building a custom type of <code>RichSequence</code>.
 For example, use <code>RichSequenceBuilderFactory.FACTORY</code>
 to emulate <code>readFasta(BufferedReader, SymbolTokenization)</code>
 and <code>RichSequenceBuilderFactory.PACKED</code> to force all symbols
 to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a <code>RichSequence</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaDNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readFastaDNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readFastaDNA</B>(java.io.BufferedReader&nbsp;br,
                                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an FASTA-format stream of DNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaRNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readFastaRNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readFastaRNA</B>(java.io.BufferedReader&nbsp;br,
                                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an FASTA-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readFastaProtein(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readFastaProtein</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readFastaProtein</B>(java.io.BufferedReader&nbsp;br,
                                                    <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an FASTA-format stream of Protein sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readGenbank(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readGenbank</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readGenbank</B>(java.io.BufferedReader&nbsp;br,
                                               <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                               <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a GenBank file using a custom type of SymbolList. For example,
 use RichSequenceBuilderFactory.FACTORY to emulate readFasta(BufferedReader,
 SymbolTokenization) and RichSequenceBuilderFactory.PACKED to force all
 symbols to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a <code>SymbolList</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readGenbankDNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readGenbankDNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readGenbankDNA</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an GenBank-format stream of DNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readGenbankRNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readGenbankRNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readGenbankRNA</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an GenBank-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readGenbankProtein(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readGenbankProtein</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readGenbankProtein</B>(java.io.BufferedReader&nbsp;br,
                                                      <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an GenBank-format stream of Protein sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readINSDseq(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readINSDseq</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readINSDseq</B>(java.io.BufferedReader&nbsp;br,
                                               <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                               <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a INSDseq file using a custom type of SymbolList. For example,
 use RichSequenceBuilderFactory.FACTORY to emulate readFasta(BufferedReader,
 SymbolTokenization) and RichSequenceBuilderFactory.PACKED to force all
 symbols to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a <code>SymbolList</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readINSDseqDNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readINSDseqDNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readINSDseqDNA</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an INSDseq-format stream of DNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readINSDseqRNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readINSDseqRNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readINSDseqRNA</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an INSDseq-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readINSDseqProtein(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readINSDseqProtein</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readINSDseqProtein</B>(java.io.BufferedReader&nbsp;br,
                                                      <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an INSDseq-format stream of Protein sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBLxml(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBLxml</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBLxml</B>(java.io.BufferedReader&nbsp;br,
                                               <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                               <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a EMBLxml file using a custom type of SymbolList. For example,
 use RichSequenceBuilderFactory.FACTORY to emulate readFasta(BufferedReader,
 SymbolTokenization) and RichSequenceBuilderFactory.PACKED to force all
 symbols to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a <code>SymbolList</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBLxmlDNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBLxmlDNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBLxmlDNA</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an EMBLxml-format stream of DNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBLxmlRNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBLxmlRNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBLxmlRNA</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an EMBLxml-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBLxmlProtein(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBLxmlProtein</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBLxmlProtein</B>(java.io.BufferedReader&nbsp;br,
                                                      <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an EMBLxml-format stream of Protein sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBL(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBL</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBL</B>(java.io.BufferedReader&nbsp;br,
                                            <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                            <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                            <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a EMBL file using a custom type of SymbolList. For example,
 use RichSequenceBuilderFactory.FACTORY to emulate readFasta(BufferedReader,
 SymbolTokenization) and RichSequenceBuilderFactory.PACKED to force all
 symbols to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a <code>SymbolList</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBLDNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBLDNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBLDNA</B>(java.io.BufferedReader&nbsp;br,
                                               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an EMBL-format stream of DNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBLRNA(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBLRNA</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBLRNA</B>(java.io.BufferedReader&nbsp;br,
                                               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an EMBL-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readEMBLProtein(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readEMBLProtein</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readEMBLProtein</B>(java.io.BufferedReader&nbsp;br,
                                                   <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an EMBL-format stream of Protein sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readUniProt(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readUniProt</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readUniProt</B>(java.io.BufferedReader&nbsp;br,
                                               <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                               <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a UniProt file using a custom type of SymbolList. For example,
 use RichSequenceBuilderFactory.FACTORY to emulate readFasta(BufferedReader,
 SymbolTokenization) and RichSequenceBuilderFactory.PACKED to force all
 symbols to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a <code>SymbolList</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readUniProt(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readUniProt</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readUniProt</B>(java.io.BufferedReader&nbsp;br,
                                               <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an UniProt-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readUniProtXML(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)"><!-- --></A><H3>
readUniProtXML</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readUniProtXML</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
                                                  <A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Read a UniProt XML file using a custom type of SymbolList. For example,
 use RichSequenceBuilderFactory.FACTORY to emulate readFasta(BufferedReader,
 SymbolTokenization) and RichSequenceBuilderFactory.PACKED to force all
 symbols to be encoded using bit-packing.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>sTok</CODE> - a <code>SymbolTokenization</code> that understands the sequences<DD><CODE>seqFactory</CODE> - a factory used to build a <code>SymbolList</code><DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="readUniProtXML(java.io.BufferedReader, org.biojavax.Namespace)"><!-- --></A><H3>
readUniProtXML</H3>
<PRE>
public static <A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A> <B>readUniProtXML</B>(java.io.BufferedReader&nbsp;br,
                                                  <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</PRE>
<DL>
<DD>Iterate over the sequences in an UniProt XML-format stream of RNA sequences.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>br</CODE> - the <code>BufferedReader</code> to read data from<DD><CODE>ns</CODE> - a <code>Namespace</code> to load the sequences into. Null implies that it should
              use the namespace specified in the file. If no namespace is
              specified in the file, then <code>RichObjectFactory.getDefaultNamespace()</code>
              is used.
<DT><B>Returns:</B><DD>a <code>RichSequenceIterator</code> over each sequence in the fasta file</DL>
</DD>
</DL>
<HR>

<A NAME="writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace, org.biojavax.bio.seq.io.FastaHeader)"><!-- --></A><H3>
writeFasta</H3>
<PRE>
public static void <B>writeFasta</B>(java.io.OutputStream&nbsp;os,
                              <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                              <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
                              <A HREF="../../../../org/biojavax/bio/seq/io/FastaHeader.html" title="class in org.biojavax.bio.seq.io">FastaHeader</A>&nbsp;header)
                       throws java.io.IOException</PRE>
<DL>
<DD>Writes <CODE>Sequence</CODE>s from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 Fasta Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input <CODE>RichSequence</CODE>s<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the <CODE>RichSequence</CODE>s to. <CODE>Null</CODE> implies that it should
 use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)"><!-- --></A><H3>
writeFasta</H3>
<PRE>
public static void <B>writeFasta</B>(java.io.OutputStream&nbsp;os,
                              <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                              <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                       throws java.io.IOException</PRE>
<DL>
<DD>Writes <CODE>Sequence</CODE>s from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 Fasta Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input <CODE>RichSequence</CODE>s<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the <CODE>RichSequence</CODE>s to. <CODE>Null</CODE> implies that it should
 use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><!-- --></A><H3>
writeFasta</H3>
<PRE>
public static void <B>writeFasta</B>(java.io.OutputStream&nbsp;os,
                              <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                              <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                       throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in Fasta format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace, org.biojavax.bio.seq.io.FastaHeader)"><!-- --></A><H3>
writeFasta</H3>
<PRE>
public static void <B>writeFasta</B>(java.io.OutputStream&nbsp;os,
                              <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                              <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
                              <A HREF="../../../../org/biojavax/bio/seq/io/FastaHeader.html" title="class in org.biojavax.bio.seq.io">FastaHeader</A>&nbsp;header)
                       throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in Fasta format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.<DD><CODE>header</CODE> - a <code>FastaHeader</code> that controls the fields in the header.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeGenbank(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)"><!-- --></A><H3>
writeGenbank</H3>
<PRE>
public static void <B>writeGenbank</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 GenBank Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input Sequences<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeGenbank(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><!-- --></A><H3>
writeGenbank</H3>
<PRE>
public static void <B>writeGenbank</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in GenBank format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeINSDseq(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)"><!-- --></A><H3>
writeINSDseq</H3>
<PRE>
public static void <B>writeINSDseq</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 INSDseq Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input Sequences<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeINSDseq(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><!-- --></A><H3>
writeINSDseq</H3>
<PRE>
public static void <B>writeINSDseq</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in INSDseq format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeEMBLxml(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)"><!-- --></A><H3>
writeEMBLxml</H3>
<PRE>
public static void <B>writeEMBLxml</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 EMBLxml Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input Sequences<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeEMBLxml(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><!-- --></A><H3>
writeEMBLxml</H3>
<PRE>
public static void <B>writeEMBLxml</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in EMBLxml format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeEMBL(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)"><!-- --></A><H3>
writeEMBL</H3>
<PRE>
public static void <B>writeEMBL</B>(java.io.OutputStream&nbsp;os,
                             <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                      throws java.io.IOException</PRE>
<DL>
<DD>Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 EMBL Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input Sequences<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeEMBL(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><!-- --></A><H3>
writeEMBL</H3>
<PRE>
public static void <B>writeEMBL</B>(java.io.OutputStream&nbsp;os,
                             <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                             <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                      throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in EMBL format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeUniProt(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)"><!-- --></A><H3>
writeUniProt</H3>
<PRE>
public static void <B>writeUniProt</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 UniProt Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input Sequences<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeUniProt(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><!-- --></A><H3>
writeUniProt</H3>
<PRE>
public static void <B>writeUniProt</B>(java.io.OutputStream&nbsp;os,
                                <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                         throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in UniProt format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeUniProtXML(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)"><!-- --></A><H3>
writeUniProtXML</H3>
<PRE>
public static void <B>writeUniProtXML</B>(java.io.OutputStream&nbsp;os,
                                   <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                                   <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                            throws java.io.IOException</PRE>
<DL>
<DD>Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 UniProt XML Format.  This makes for a useful format filter where a
 <code>StreamReader</code> can be sent to the <code>RichStreamWriter</code> after formatting.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - The stream to write fasta formatted data to<DD><CODE>in</CODE> - The source of input Sequences<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="writeUniProtXML(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><!-- --></A><H3>
writeUniProtXML</H3>
<PRE>
public static void <B>writeUniProtXML</B>(java.io.OutputStream&nbsp;os,
                                   <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                                   <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)
                            throws java.io.IOException</PRE>
<DL>
<DD>Writes a single <code>Sequence</code> to an <code>OutputStream</code> in UniProt XML format.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>os</CODE> - the <code>OutputStream</code>.<DD><CODE>seq</CODE> - the <code>Sequence</code>.<DD><CODE>ns</CODE> - a <code>Namespace</code> to write the sequences to. Null implies that it should
              use the namespace specified in the individual sequence.
<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE> - if there is an IO problem</DL>
</DD>
</DL>
<HR>

<A NAME="getDNAParser()"><!-- --></A><H3>
getDNAParser</H3>
<PRE>
public static <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> <B>getDNAParser</B>()</PRE>
<DL>
<DD>Creates a DNA symbol tokenizer.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SymbolTokenization</code> for parsing DNA.</DL>
</DD>
</DL>
<HR>

<A NAME="getRNAParser()"><!-- --></A><H3>
getRNAParser</H3>
<PRE>
public static <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> <B>getRNAParser</B>()</PRE>
<DL>
<DD>Creates a RNA symbol tokenizer.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SymbolTokenization</code> for parsing RNA.</DL>
</DD>
</DL>
<HR>

<A NAME="getNucleotideParser()"><!-- --></A><H3>
getNucleotideParser</H3>
<PRE>
public static <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> <B>getNucleotideParser</B>()</PRE>
<DL>
<DD>Creates a nucleotide symbol tokenizer.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SymbolTokenization</code> for parsing nucleotides.</DL>
</DD>
</DL>
<HR>

<A NAME="getProteinParser()"><!-- --></A><H3>
getProteinParser</H3>
<PRE>
public static <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> <B>getProteinParser</B>()</PRE>
<DL>
<DD>Creates a protein symbol tokenizer.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a <code>SymbolTokenization</code> for parsing protein.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/RichSequence.IOTools.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.IOTools.SingleRichSeqIterator.html" title="class in org.biojavax.bio.seq"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojavax/bio/seq/RichSequence.IOTools.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="RichSequence.IOTools.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
