<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:51:03 BST 2008 -->
<TITLE>
Uses of Interface org.biojavax.Namespace (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Interface org.biojavax.Namespace (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/biojavax//class-useNamespace.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Namespace.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Interface<br>org.biojavax.Namespace</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax"><B>org.biojavax</B></A></TD>
<TD>The Biojava extensions packages, classes that extend the core biojava 
     functionality&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio"><B>org.biojavax.bio</B></A></TD>
<TD>Classes to represent biological entities and their relationships.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.ncbi"><B>org.biojavax.bio.db.ncbi</B></A></TD>
<TD>Interfaces to NCBI data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq"><B>org.biojavax.bio.seq</B></A></TD>
<TD>Rich implementations of Sequences, Locations and Features.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq.io"><B>org.biojavax.bio.seq.io</B></A></TD>
<TD>Classes to support the I/O of RichSequence and 
     Bioentry objects.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A> in <A HREF="../../../org/biojavax/package-summary.html">org.biojavax</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../org/biojavax/package-summary.html">org.biojavax</A> that implement <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/biojavax/SimpleNamespace.html" title="class in org.biojavax">SimpleNamespace</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A basic Namespace implemenation.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../org/biojavax/package-summary.html">org.biojavax</A> that return <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></CODE></FONT></TD>
<TD><CODE><B>RichObjectFactory.</B><B><A HREF="../../../org/biojavax/RichObjectFactory.html#getDefaultNamespace()">getDefaultNamespace</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the default namespace object.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A> in <A HREF="../../../org/biojavax/bio/package-summary.html">org.biojavax.bio</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../org/biojavax/bio/package-summary.html">org.biojavax.bio</A> that return <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></CODE></FONT></TD>
<TD><CODE><B>BioEntry.</B><B><A HREF="../../../org/biojavax/bio/BioEntry.html#getNamespace()">getNamespace</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the namespace of this bioentry.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></CODE></FONT></TD>
<TD><CODE><B>SimpleBioEntry.</B><B><A HREF="../../../org/biojavax/bio/SimpleBioEntry.html#getNamespace()">getNamespace</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the namespace of this bioentry.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../org/biojavax/bio/package-summary.html">org.biojavax.bio</A> with parameters of type <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/biojavax/bio/SimpleBioEntry.html#SimpleBioEntry(org.biojavax.Namespace, java.lang.String, java.lang.String, int)">SimpleBioEntry</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
               java.lang.String&nbsp;name,
               java.lang.String&nbsp;accession,
               int&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new bioentry representing the sequence in the given namespace
 with the given name, accession and version.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.ncbi"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A> in <A HREF="../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A> that return <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></CODE></FONT></TD>
<TD><CODE><B>GenbankRichSequenceDB.</B><B><A HREF="../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html#getNamespace()">getNamespace</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Getter for property namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></CODE></FONT></TD>
<TD><CODE><B>GenpeptRichSequenceDB.</B><B><A HREF="../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html#getNamespace()">getNamespace</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Getter for property namespace.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A> with parameters of type <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>GenbankRichSequenceDB.</B><B><A HREF="../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html#getRichSequence(java.lang.String, org.biojavax.Namespace)">getRichSequence</A></B>(java.lang.String&nbsp;id,
                <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;nsp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given the appropriate Genbank ID, return the matching RichSequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>GenpeptRichSequenceDB.</B><B><A HREF="../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html#getRichSequence(java.lang.String, org.biojavax.Namespace)">getRichSequence</A></B>(java.lang.String&nbsp;id,
                <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;nsp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given the appropriate Genbank ID, return the matching RichSequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankRichSequenceDB.</B><B><A HREF="../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Setter for property namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenpeptRichSequenceDB.</B><B><A HREF="../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Setter for property namespace.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A> in <A HREF="../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A> with parameters of type <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.Tools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.Tools.html#createRichSequence(org.biojavax.Namespace, java.lang.String, java.lang.String, org.biojava.bio.symbol.Alphabet)">createRichSequence</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
                   java.lang.String&nbsp;name,
                   java.lang.String&nbsp;seqString,
                   <A HREF="../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new RichSequence in the specified namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.Tools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.Tools.html#createRichSequence(org.biojavax.Namespace, java.lang.String, org.biojava.bio.symbol.SymbolList)">createRichSequence</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
                   java.lang.String&nbsp;name,
                   <A HREF="../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;syms)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new RichSequence in the specified namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBL(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readEMBL</A></B>(java.io.BufferedReader&nbsp;br,
         <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
         <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
         <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a EMBL file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLDNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLDNA</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBL-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLProtein(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLProtein</A></B>(java.io.BufferedReader&nbsp;br,
                <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBL-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLRNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLRNA</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBL-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxml(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readEMBLxml</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a EMBLxml file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxmlDNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLxmlDNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBLxml-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxmlProtein(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLxmlProtein</A></B>(java.io.BufferedReader&nbsp;br,
                   <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBLxml-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxmlRNA(java.io.BufferedReader, org.biojavax.Namespace)">readEMBLxmlRNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an EMBLxml-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.Namespace)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a fasta file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
          <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a fasta file building a custom type of <code>RichSequence</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFastaDNA(java.io.BufferedReader, org.biojavax.Namespace)">readFastaDNA</A></B>(java.io.BufferedReader&nbsp;br,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an FASTA-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFastaProtein(java.io.BufferedReader, org.biojavax.Namespace)">readFastaProtein</A></B>(java.io.BufferedReader&nbsp;br,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an FASTA-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFastaRNA(java.io.BufferedReader, org.biojavax.Namespace)">readFastaRNA</A></B>(java.io.BufferedReader&nbsp;br,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an FASTA-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFile(java.io.File, org.biojavax.Namespace)">readFile</A></B>(java.io.File&nbsp;file,
         <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a file is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFile(java.io.File, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readFile</A></B>(java.io.File&nbsp;file,
         <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
         <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a file is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbank(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readGenbank</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a GenBank file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbankDNA(java.io.BufferedReader, org.biojavax.Namespace)">readGenbankDNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an GenBank-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbankProtein(java.io.BufferedReader, org.biojavax.Namespace)">readGenbankProtein</A></B>(java.io.BufferedReader&nbsp;br,
                   <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an GenBank-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbankRNA(java.io.BufferedReader, org.biojavax.Namespace)">readGenbankRNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an GenBank-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseq(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readINSDseq</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a INSDseq file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseqDNA(java.io.BufferedReader, org.biojavax.Namespace)">readINSDseqDNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an INSDseq-format stream of DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseqProtein(java.io.BufferedReader, org.biojavax.Namespace)">readINSDseqProtein</A></B>(java.io.BufferedReader&nbsp;br,
                   <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an INSDseq-format stream of Protein sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseqRNA(java.io.BufferedReader, org.biojavax.Namespace)">readINSDseqRNA</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an INSDseq-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readStream(java.io.BufferedInputStream, org.biojavax.Namespace)">readStream</A></B>(java.io.BufferedInputStream&nbsp;stream,
           <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a stream is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readStream(java.io.BufferedInputStream, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readStream</A></B>(java.io.BufferedInputStream&nbsp;stream,
           <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
           <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Guess which format a stream is then attempt to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProt(java.io.BufferedReader, org.biojavax.Namespace)">readUniProt</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an UniProt-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProt(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readUniProt</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a UniProt file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProtXML(java.io.BufferedReader, org.biojavax.Namespace)">readUniProtXML</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterate over the sequences in an UniProt XML-format stream of RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProtXML(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readUniProtXML</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
               <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
               <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a UniProt XML file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.Tools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.Tools.html#subSequence(org.biojavax.bio.seq.RichSequence, int, int, org.biojavax.Namespace, java.lang.String, java.lang.String, java.lang.String, int, java.lang.Double)">subSequence</A></B>(<A HREF="../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A>&nbsp;s,
            int&nbsp;from,
            int&nbsp;to,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;newNamespace,
            java.lang.String&nbsp;newName,
            java.lang.String&nbsp;newAccession,
            java.lang.String&nbsp;newIdentifier,
            int&nbsp;newVersion,
            java.lang.Double&nbsp;seqVersion)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new sequence from a subregion of another sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBL(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeEMBL</A></B>(java.io.OutputStream&nbsp;os,
          <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
          <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 EMBL Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBL(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeEMBL</A></B>(java.io.OutputStream&nbsp;os,
          <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
          <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in EMBL format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBLxml(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeEMBLxml</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 EMBLxml Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeEMBLxml(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeEMBLxml</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in EMBLxml format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
           <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes <CODE>Sequence</CODE>s from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 Fasta Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace, org.biojavax.bio.seq.io.FastaHeader)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
           <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
           <A HREF="../../../org/biojavax/bio/seq/io/FastaHeader.html" title="class in org.biojavax.bio.seq.io">FastaHeader</A>&nbsp;header)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes <CODE>Sequence</CODE>s from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 Fasta Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
           <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in Fasta format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace, org.biojavax.bio.seq.io.FastaHeader)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
           <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
           <A HREF="../../../org/biojavax/bio/seq/io/FastaHeader.html" title="class in org.biojavax.bio.seq.io">FastaHeader</A>&nbsp;header)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in Fasta format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeGenbank(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeGenbank</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 GenBank Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeGenbank(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeGenbank</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in GenBank format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeINSDseq(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeINSDseq</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 INSDseq Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeINSDseq(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeINSDseq</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in INSDseq format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProt(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeUniProt</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 UniProt Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProt(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeUniProt</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
             <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in UniProt format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProtXML(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeUniProtXML</A></B>(java.io.OutputStream&nbsp;os,
                <A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in,
                <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes sequences from a <code>SequenceIterator</code> to an <code>OutputStream </code>in
 UniProt XML Format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../org/biojavax/bio/seq/RichSequence.IOTools.html#writeUniProtXML(java.io.OutputStream, org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeUniProtXML</A></B>(java.io.OutputStream&nbsp;os,
                <A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a single <code>Sequence</code> to an <code>OutputStream</code> in UniProt XML format.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A> with parameters of type <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/biojavax/bio/seq/SimpleRichSequence.html#SimpleRichSequence(org.biojavax.Namespace, java.lang.String, java.lang.String, int, org.biojava.bio.symbol.SymbolList, java.lang.Double)">SimpleRichSequence</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
                   java.lang.String&nbsp;name,
                   java.lang.String&nbsp;accession,
                   int&nbsp;version,
                   <A HREF="../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;symList,
                   java.lang.Double&nbsp;seqversion)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new instance of SimpleRichSequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/biojavax/bio/seq/ThinRichSequence.html#ThinRichSequence(org.biojavax.Namespace, java.lang.String, java.lang.String, int, org.biojava.bio.symbol.Alphabet, java.lang.Double)">ThinRichSequence</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
                 java.lang.String&nbsp;name,
                 java.lang.String&nbsp;accession,
                 int&nbsp;version,
                 <A HREF="../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha,
                 java.lang.Double&nbsp;seqversion)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new instance of ThinRichSequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A> in <A HREF="../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A> with parameters of type <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/biojavax/bio/seq/RichLocation.html" title="interface in org.biojavax.bio.seq">RichLocation</A></CODE></FONT></TD>
<TD><CODE><B>GenbankLocationParser.</B><B><A HREF="../../../org/biojavax/bio/seq/io/GenbankLocationParser.html#parseLocation(org.biojavax.Namespace, java.lang.String, java.lang.String)">parseLocation</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;featureNS,
              java.lang.String&nbsp;featureAccession,
              java.lang.String&nbsp;locationString)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parses a location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.
 If namespace is null, then the namespace of the sequence in the fasta is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/EMBLFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>RichSequenceFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/RichSequenceFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;listener,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/INSDseqFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/GenbankFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/UniProtFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/FastaFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rsiol,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.
 If namespace is null, then the namespace of the sequence in the fasta is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the namespace of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the namespace of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.
 If namespace is null, then the sequence's own namespace is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/EMBLFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.
 Namespace is ignored as EMBL has no concept of it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequenceFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/RichSequenceFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/INSDseqFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.
 Namespace is ignored as INSDseq has no concept of it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/GenbankFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.
 Namespace is ignored as Genbank has no concept of it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/UniProtFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.
 Namespace is ignored as UniProt has no concept of it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.
 Namespace is ignored as EMBLxml has no concept of it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/FastaFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes a sequence out to the outputstream given by beginWriting() using the default format of the
 implementing class.
 If namespace is null, then the sequence's own namespace is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../org/biojavax/bio/seq/io/EMBLFormat.html#writeSequence(org.biojava.bio.seq.Sequence, java.lang.String, org.biojavax.Namespace)">writeSequence</A></B>(<A HREF="../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
              java.lang.String&nbsp;format,
              <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;As per <A HREF="../../../org/biojavax/bio/seq/io/EMBLFormat.html#writeSequence(org.biojava.bio.seq.Sequence, org.biojavax.Namespace)"><CODE>EMBLFormat.writeSequence(Sequence, Namespace)</CODE></A>, except
 that it also takes a format parameter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichStreamWriter.</B><B><A HREF="../../../org/biojavax/bio/seq/io/RichStreamWriter.html#writeStream(org.biojava.bio.seq.SequenceIterator, org.biojavax.Namespace)">writeStream</A></B>(<A HREF="../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;ss,
            <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Write each of the sequences in ss to the stream in the given format.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A> with parameters of type <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/biojavax/bio/seq/io/RichStreamReader.html#RichStreamReader(java.io.BufferedReader, org.biojavax.bio.seq.io.RichSequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">RichStreamReader</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceFormat.html" title="interface in org.biojavax.bio.seq.io">RichSequenceFormat</A>&nbsp;format,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;sf,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new stream reader on the given reader, which will attempt to read
 sequences in the given format, having symbols from the given tokenization, and
 pass them to the given factory to be transformed into RichSequence objects in
 the given namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/biojavax/bio/seq/io/RichStreamReader.html#RichStreamReader(java.io.InputStream, org.biojavax.bio.seq.io.RichSequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">RichStreamReader</A></B>(java.io.InputStream&nbsp;is,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceFormat.html" title="interface in org.biojavax.bio.seq.io">RichSequenceFormat</A>&nbsp;format,
                 <A HREF="../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;sf,
                 <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new stream reader on the given input stream, which will attempt to read
 sequences in the given format, having symbols from the given tokenization, and
 pass them to the given factory to be transformed into RichSequence objects in
 the given namespace.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../org/biojavax/Namespace.html" title="interface in org.biojavax"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/biojavax//class-useNamespace.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Namespace.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
