<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:56 BST 2008 -->
<TITLE>
Uses of Class org.biojava.bio.seq.io.ParseException (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Class org.biojava.bio.seq.io.ParseException (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?org/biojava/bio/seq/io//class-useParseException.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="ParseException.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Class<br>org.biojava.bio.seq.io.ParseException</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.xff"><B>org.biojava.bio.program.xff</B></A></TD>
<TD>Event-driven parsing system for the Extensible Feature Format (XFF).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io.agave"><B>org.biojava.bio.seq.io.agave</B></A></TD>
<TD>Classes for converting between AGAVE XML and BioJava objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io.game"><B>org.biojava.bio.seq.io.game</B></A></TD>
<TD>Event-driven parsing system for the Gene Annotation Markup Elements (GAME).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.ontology.io"><B>org.biojava.ontology.io</B></A></TD>
<TD>
Tools for loading and saving ontologies.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.phylo.io.nexus"><B>org.biojavax.bio.phylo.io.nexus</B></A></TD>
<TD>Classes to support the I/O of Nexus files.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.phylo.io.phylip"><B>org.biojavax.bio.phylo.io.phylip</B></A></TD>
<TD>Classes to support the reading and writing of PHYLIP format.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq.io"><B>org.biojavax.bio.seq.io</B></A></TD>
<TD>Classes to support the I/O of RichSequence and 
     Bioentry objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.taxa.io"><B>org.biojavax.bio.taxa.io</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>PhredFormat.</B><B><A HREF="../../../../../../org/biojava/bio/program/phred/PhredFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;siol)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.xff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojava/bio/program/xff/package-summary.html">org.biojava.bio.program.xff</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/xff/package-summary.html">org.biojava.bio.program.xff</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureHandler.</B><B><A HREF="../../../../../../org/biojava/bio/program/xff/FeatureHandler.html#fireEndFeature()">fireEndFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fire the endFeature event.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureHandler.</B><B><A HREF="../../../../../../org/biojava/bio/program/xff/FeatureHandler.html#fireStartFeature()">fireStartFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fire the startFeature event.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureHandler.</B><B><A HREF="../../../../../../org/biojava/bio/program/xff/FeatureHandler.html#setFeatureProperty(java.lang.Object, java.lang.Object)">setFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set a property.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Null implementation</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener of a feature property.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderBase.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderBase.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add an annotation-bundle entry to the feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Notify the listener of a sequence-wide property.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotProcessor.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotProcessor.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>OrganismParser.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/OrganismParser.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;sciNameKey,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener of a sequence-wide property.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblProcessor.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblProcessor.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankProcessor.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankProcessor.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblFileFormer.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaDescriptionLineParser.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/FastaDescriptionLineParser.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Null implementation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Mark the end of data associated with one specific feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblFileFormer.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Notify the listener that processing of the sequence is complete.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotProcessor.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotProcessor.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>OrganismParser.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/OrganismParser.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener that processing of the sequence is complete.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblProcessor.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblProcessor.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankProcessor.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankProcessor.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblFileFormer.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblLikeFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblLikeFormat.html#processSequenceLine(java.lang.String, org.biojava.bio.seq.io.StreamParser)">processSequenceLine</A></B>(java.lang.String&nbsp;line,
                    <A HREF="../../../../../../org/biojava/bio/seq/io/StreamParser.html" title="interface in org.biojava.bio.seq.io">StreamParser</A>&nbsp;parser)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Dispatch symbol data from SQ-block line of an EMBL-like file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EmblLikeFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblLikeFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/FastaFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;siol)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads information from a flatfile to a <code>SeqIOListener</code>
 using a <code>SymbolTokenizer</code> to convert sequence strings
 to <code>Symbol</code> objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads a sequence from the specified reader using the Symbol
 parser and Sequence Factory provided.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankXmlFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankXmlFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;theName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;The name is printed out as part of the identifier line.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener that the current sequence is generally known
 by a particular name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblFileFormer.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;theURI)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Null implementation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener of a URI identifying the current sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblFileFormer.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Null implementation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener that a new feature object is starting.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProteinRefSeqFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/ProteinRefSeqFileFormer.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblFileFormer.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SwissprotFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SwissprotFileFormer.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Start the processing of a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOFilter.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOListener.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Start the processing of a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOAdapter.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOAdapter.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GenbankFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFileFormer.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblFileFormer.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblFileFormer.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io.agave"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojava/bio/seq/io/agave/package-summary.html">org.biojava.bio.seq.io.agave</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/io/agave/package-summary.html">org.biojava.bio.seq.io.agave</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.util.ListIterator</CODE></FONT></TD>
<TD><CODE><B>StAXPropertyHandler.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/agave/StAXPropertyHandler.html#getHandlerStackIterator()">getHandlerStackIterator</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get iterator for current stack starting at the position
 below mine.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io.game"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojava/bio/seq/io/game/package-summary.html">org.biojava.bio.seq.io.game</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/io/game/package-summary.html">org.biojava.bio.seq.io.game</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.util.ListIterator</CODE></FONT></TD>
<TD><CODE><B>StAXPropertyHandler.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/game/StAXPropertyHandler.html#getHandlerStackIterator()">getHandlerStackIterator</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;get iterator for current stack starting at the position
 below mine.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.ontology.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojava/ontology/io/package-summary.html">org.biojava.ontology.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/ontology/io/package-summary.html">org.biojava.ontology.io</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/ontology/Ontology.html" title="interface in org.biojava.ontology">Ontology</A></CODE></FONT></TD>
<TD><CODE><B>GOParser.</B><B><A HREF="../../../../../../org/biojava/ontology/io/GOParser.html#parseGO(java.io.BufferedReader, java.lang.String, java.lang.String, org.biojava.ontology.OntologyFactory)">parseGO</A></B>(java.io.BufferedReader&nbsp;goFile,
        java.lang.String&nbsp;ontoName,
        java.lang.String&nbsp;ontoDescription,
        <A HREF="../../../../../../org/biojava/ontology/OntologyFactory.html" title="interface in org.biojava.ontology">OntologyFactory</A>&nbsp;factory)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.phylo.io.nexus"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/package-summary.html">org.biojavax.bio.phylo.io.nexus</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/package-summary.html">org.biojavax.bio.phylo.io.nexus</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DistancesBlockListener.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/DistancesBlockListener.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a TAXLABEL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CharactersBlock.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/CharactersBlock.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a TAXLABEL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DistancesBlock.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/DistancesBlock.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a TAXLABEL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TaxaBlockBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/TaxaBlockBuilder.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TaxaBlockListener.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/TaxaBlockListener.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add another value after the TAXLABEL tag.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TaxaBlock.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/TaxaBlock.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a TAXLABEL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CharactersBlockBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/CharactersBlockBuilder.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CharactersBlockListener.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/CharactersBlockListener.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a TAXLABEL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DistancesBlockBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/DistancesBlockBuilder.html#addTaxLabel(java.lang.String)">addTaxLabel</A></B>(java.lang.String&nbsp;taxLabel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusBlockParser.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusBlockParser.html#commentText(java.lang.String)">commentText</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Receiving free text inside a comment tag.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusBlockParser.Abstract.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusBlockParser.Abstract.html#commentText(java.lang.String)">commentText</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusBlockListener.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusBlockListener.html#commentText(java.lang.String)">commentText</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Receiving free text inside a comment tag.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TreesBlockParser.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/TreesBlockParser.html#commentText(java.lang.String)">commentText</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusFileListener.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileListener.html#commentText(java.lang.String)">commentText</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Receiving free text inside a comment tag.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusFileListener.Abstract.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileListener.Abstract.html#commentText(java.lang.String)">commentText</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusFileFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileFormat.html#parseFile(org.biojavax.bio.phylo.io.nexus.NexusFileListener, java.io.File)">parseFile</A></B>(<A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileListener.html" title="interface in org.biojavax.bio.phylo.io.nexus">NexusFileListener</A>&nbsp;listener,
          java.io.File&nbsp;inputFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parse a file and send events to the given listener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusFileFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileFormat.html#parseInputStream(org.biojavax.bio.phylo.io.nexus.NexusFileListener, java.io.InputStream)">parseInputStream</A></B>(<A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileListener.html" title="interface in org.biojavax.bio.phylo.io.nexus">NexusFileListener</A>&nbsp;listener,
                 java.io.InputStream&nbsp;inputStream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parse a stream and send events to the given listener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusFileFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileFormat.html#parseReader(org.biojavax.bio.phylo.io.nexus.NexusFileListener, java.io.Reader)">parseReader</A></B>(<A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileListener.html" title="interface in org.biojavax.bio.phylo.io.nexus">NexusFileListener</A>&nbsp;listener,
            java.io.Reader&nbsp;inputReader)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parse a reader and send events to the given listener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusBlockParser.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusBlockParser.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notifies the parser of the next token.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>abstract &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusBlockParser.Abstract.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusBlockParser.Abstract.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TaxaBlockParser.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/TaxaBlockParser.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TreesBlockParser.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/TreesBlockParser.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusFileListener.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileListener.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Encountered a token.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NexusFileListener.Abstract.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/NexusFileListener.Abstract.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CharactersBlockParser.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/CharactersBlockParser.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DistancesBlockParser.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/nexus/DistancesBlockParser.html#parseToken(java.lang.String)">parseToken</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.phylo.io.phylip"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/package-summary.html">org.biojavax.bio.phylo.io.phylip</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/package-summary.html">org.biojavax.bio.phylo.io.phylip</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileBuilder.html#endFile()">endFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileListener.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileListener.html#endFile()">endFile</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Finished reading a file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileFormat.html#parse(org.biojavax.bio.phylo.io.phylip.PHYLIPFileListener, java.io.BufferedReader)">parse</A></B>(<A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileListener.html" title="interface in org.biojavax.bio.phylo.io.phylip">PHYLIPFileListener</A>&nbsp;listener,
      java.io.BufferedReader&nbsp;reader)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileFormat.html#parseFile(org.biojavax.bio.phylo.io.phylip.PHYLIPFileListener, java.io.File)">parseFile</A></B>(<A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileListener.html" title="interface in org.biojavax.bio.phylo.io.phylip">PHYLIPFileListener</A>&nbsp;listener,
          java.io.File&nbsp;inputFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileFormat.html#parseInputStream(org.biojavax.bio.phylo.io.phylip.PHYLIPFileListener, java.io.InputStream)">parseInputStream</A></B>(<A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileListener.html" title="interface in org.biojavax.bio.phylo.io.phylip">PHYLIPFileListener</A>&nbsp;listener,
                 java.io.InputStream&nbsp;inputStream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileFormat.html#parseReader(org.biojavax.bio.phylo.io.phylip.PHYLIPFileListener, java.io.Reader)">parseReader</A></B>(<A HREF="../../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileListener.html" title="interface in org.biojavax.bio.phylo.io.phylip">PHYLIPFileListener</A>&nbsp;listener,
            java.io.Reader&nbsp;inputReader)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#addFeatureProperty(java.lang.Object, java.lang.Object)">addFeatureProperty</A></B>(java.lang.Object&nbsp;key,
                   java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener of a feature property.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#addSequenceProperty(java.lang.Object, java.lang.Object)">addSequenceProperty</A></B>(java.lang.Object&nbsp;key,
                    java.lang.Object&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener of a sequence-wide property.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Mark the end of data associated with one specific feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#endSequence()">endSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener that processing of the sequence is complete.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B>UniProtCommentParser.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtCommentParser.html#generate()">generate</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Generates a comment string based on the current values of the
 internal fields.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichFeature.html" title="interface in org.biojavax.bio.seq">RichFeature</A></CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#getCurrentFeature()">getCurrentFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the feature currently being created.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichFeature.html" title="interface in org.biojavax.bio.seq">RichFeature</A></CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#getCurrentFeature()">getCurrentFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichFeature.html" title="interface in org.biojavax.bio.seq">RichFeature</A></CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#getCurrentFeature()">getCurrentFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichFeature.html" title="interface in org.biojavax.bio.seq">RichFeature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#getCurrentFeature()">getCurrentFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the feature currently being created.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>UniProtCommentParser.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtCommentParser.html#parseComment(org.biojavax.Comment)">parseComment</A></B>(<A HREF="../../../../../../org/biojavax/Comment.html" title="interface in org.biojavax">Comment</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parses the comment string from the given comment and populates
 the internal fields appropriately.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>UniProtCommentParser.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtCommentParser.html#parseComment(java.lang.String)">parseComment</A></B>(java.lang.String&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parses the comment string from the given comment and populates
 the internal fields appropriately.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichLocation.html" title="interface in org.biojavax.bio.seq">RichLocation</A></CODE></FONT></TD>
<TD><CODE><B>GenbankLocationParser.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/GenbankLocationParser.html#parseLocation(org.biojavax.Namespace, java.lang.String, java.lang.String)">parseLocation</A></B>(<A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;featureNS,
              java.lang.String&nbsp;featureAccession,
              java.lang.String&nbsp;locationString)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parses a location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichLocation.html" title="interface in org.biojavax.bio.seq">RichLocation</A></CODE></FONT></TD>
<TD><CODE><B>UniProtLocationParser.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtLocationParser.html#parseLocation(java.lang.String)">parseLocation</A></B>(java.lang.String&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Parses a location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.
 If namespace is null, then the namespace of the sequence in the fasta is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/INSDseqFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/GenbankFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/FastaFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rsiol,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.
 If namespace is null, then the namespace of the sequence in the fasta is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/INSDseqFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/GenbankFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/FastaFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setAccession(java.lang.String)">setAccession</A></B>(java.lang.String&nbsp;accession)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the accession of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setAccession(java.lang.String)">setAccession</A></B>(java.lang.String&nbsp;accession)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setAccession(java.lang.String)">setAccession</A></B>(java.lang.String&nbsp;accession)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setAccession(java.lang.String)">setAccession</A></B>(java.lang.String&nbsp;accession)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the accession of the record being read.
 The last accession passed to this routine will always be the one used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setCircular(boolean)">setCircular</A></B>(boolean&nbsp;circular)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Defined this sequence being parser as circular.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setCircular(boolean)">setCircular</A></B>(boolean&nbsp;circular)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setCircular(boolean)">setCircular</A></B>(boolean&nbsp;circular)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setCircular(boolean)">setCircular</A></B>(boolean&nbsp;circular)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Defined this sequence being parser as circular.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setComment(java.lang.String)">setComment</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 about a comment in the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setComment(java.lang.String)">setComment</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setComment(java.lang.String)">setComment</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setComment(java.lang.String)">setComment</A></B>(java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 about a comment in the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setDescription(java.lang.String)">setDescription</A></B>(java.lang.String&nbsp;description)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the description of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setDescription(java.lang.String)">setDescription</A></B>(java.lang.String&nbsp;description)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setDescription(java.lang.String)">setDescription</A></B>(java.lang.String&nbsp;description)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setDescription(java.lang.String)">setDescription</A></B>(java.lang.String&nbsp;description)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the description of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setDivision(java.lang.String)">setDivision</A></B>(java.lang.String&nbsp;division)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the division of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setDivision(java.lang.String)">setDivision</A></B>(java.lang.String&nbsp;division)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setDivision(java.lang.String)">setDivision</A></B>(java.lang.String&nbsp;division)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setDivision(java.lang.String)">setDivision</A></B>(java.lang.String&nbsp;division)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the division of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setIdentifier(java.lang.String)">setIdentifier</A></B>(java.lang.String&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the identifier of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setIdentifier(java.lang.String)">setIdentifier</A></B>(java.lang.String&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setIdentifier(java.lang.String)">setIdentifier</A></B>(java.lang.String&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setIdentifier(java.lang.String)">setIdentifier</A></B>(java.lang.String&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the identifier of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setName(java.lang.String)">setName</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener that the current sequence is generally known
 by a particular name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the namespace of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setNamespace(org.biojavax.Namespace)">setNamespace</A></B>(<A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the namespace of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setRankedCrossRef(org.biojavax.RankedCrossRef)">setRankedCrossRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedCrossRef.html" title="interface in org.biojavax">RankedCrossRef</A>&nbsp;crossRef)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener about a cross reference.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setRankedCrossRef(org.biojavax.RankedCrossRef)">setRankedCrossRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedCrossRef.html" title="interface in org.biojavax">RankedCrossRef</A>&nbsp;crossRef)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setRankedCrossRef(org.biojavax.RankedCrossRef)">setRankedCrossRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedCrossRef.html" title="interface in org.biojavax">RankedCrossRef</A>&nbsp;crossRef)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setRankedCrossRef(org.biojavax.RankedCrossRef)">setRankedCrossRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedCrossRef.html" title="interface in org.biojavax">RankedCrossRef</A>&nbsp;ref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener about a cross reference.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setRankedDocRef(org.biojavax.RankedDocRef)">setRankedDocRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedDocRef.html" title="interface in org.biojavax">RankedDocRef</A>&nbsp;ref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 about a literature reference in the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setRankedDocRef(org.biojavax.RankedDocRef)">setRankedDocRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedDocRef.html" title="interface in org.biojavax">RankedDocRef</A>&nbsp;ref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setRankedDocRef(org.biojavax.RankedDocRef)">setRankedDocRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedDocRef.html" title="interface in org.biojavax">RankedDocRef</A>&nbsp;ref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setRankedDocRef(org.biojavax.RankedDocRef)">setRankedDocRef</A></B>(<A HREF="../../../../../../org/biojavax/RankedDocRef.html" title="interface in org.biojavax">RankedDocRef</A>&nbsp;ref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 about a literature reference in the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setRelationship(org.biojavax.bio.BioEntryRelationship)">setRelationship</A></B>(<A HREF="../../../../../../org/biojavax/bio/BioEntryRelationship.html" title="interface in org.biojavax.bio">BioEntryRelationship</A>&nbsp;relationship)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 about a relationship between the bioentry or sequence in the
 record being read and another bioentry.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setRelationship(org.biojavax.bio.BioEntryRelationship)">setRelationship</A></B>(<A HREF="../../../../../../org/biojavax/bio/BioEntryRelationship.html" title="interface in org.biojavax.bio">BioEntryRelationship</A>&nbsp;relationship)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setRelationship(org.biojavax.bio.BioEntryRelationship)">setRelationship</A></B>(<A HREF="../../../../../../org/biojavax/bio/BioEntryRelationship.html" title="interface in org.biojavax.bio">BioEntryRelationship</A>&nbsp;relationship)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setRelationship(org.biojavax.bio.BioEntryRelationship)">setRelationship</A></B>(<A HREF="../../../../../../org/biojavax/bio/BioEntryRelationship.html" title="interface in org.biojavax.bio">BioEntryRelationship</A>&nbsp;relationship)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 about a relationship between the bioentry or sequence in the
 record being read and another bioentry.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setSeqVersion(java.lang.String)">setSeqVersion</A></B>(java.lang.String&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the version of the sequence of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setSeqVersion(java.lang.String)">setSeqVersion</A></B>(java.lang.String&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setSeqVersion(java.lang.String)">setSeqVersion</A></B>(java.lang.String&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setSeqVersion(java.lang.String)">setSeqVersion</A></B>(java.lang.String&nbsp;seqVersion)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the version of the sequence of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setTaxon(org.biojavax.bio.taxa.NCBITaxon)">setTaxon</A></B>(<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A>&nbsp;taxon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the Taxon of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setTaxon(org.biojavax.bio.taxa.NCBITaxon)">setTaxon</A></B>(<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A>&nbsp;taxon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setTaxon(org.biojavax.bio.taxa.NCBITaxon)">setTaxon</A></B>(<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A>&nbsp;taxon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setTaxon(org.biojavax.bio.taxa.NCBITaxon)">setTaxon</A></B>(<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A>&nbsp;taxon)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the Taxon of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>There is no clear mapping between URI and BioSQL. This
  method is no longer used or supported in biojavax. Don't use it. Calling
  it may result in exceptions. Use instead setName(String name),
  setAccession(String accession), setVersion(int version) etc as
  appropriate.</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setURI(java.lang.String)">setURI</A></B>(java.lang.String&nbsp;uri)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener of a URI identifying the current sequence.
 NOT IMPLEMENTED</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html#setVersion(int)">setVersion</A></B>(int&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the version of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#setVersion(int)">setVersion</A></B>(int&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#setVersion(int)">setVersion</A></B>(int&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#setVersion(int)">setVersion</A></B>(int&nbsp;version)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Call back method so the event emitter can tell the listener
 the version of the record being read.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#startFeature(org.biojava.bio.seq.Feature.Template)">startFeature</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Notify the listener that a new feature object is starting.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSeqIOAdapter.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOAdapter.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DebuggingRichSeqIOListener.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/DebuggingRichSeqIOListener.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#startSequence()">startSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Start the processing of a sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.taxa.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A> in <A HREF="../../../../../../org/biojavax/bio/taxa/io/package-summary.html">org.biojavax.bio.taxa.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/taxa/io/package-summary.html">org.biojavax.bio.taxa.io</A> that throw <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A></CODE></FONT></TD>
<TD><CODE><B>SimpleNCBITaxonomyLoader.</B><B><A HREF="../../../../../../org/biojavax/bio/taxa/io/SimpleNCBITaxonomyLoader.html#readName(java.io.BufferedReader)">readName</A></B>(java.io.BufferedReader&nbsp;names)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads the next entry from the names.dmp file and returns the corresponding 
 NCBITaxon object with the name added in already.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A></CODE></FONT></TD>
<TD><CODE><B>NCBITaxonomyLoader.</B><B><A HREF="../../../../../../org/biojavax/bio/taxa/io/NCBITaxonomyLoader.html#readName(java.io.BufferedReader)">readName</A></B>(java.io.BufferedReader&nbsp;names)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads the next entry from the names.dmp file and returns the corresponding 
 NCBITaxon object with the name added in already.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A></CODE></FONT></TD>
<TD><CODE><B>SimpleNCBITaxonomyLoader.</B><B><A HREF="../../../../../../org/biojavax/bio/taxa/io/SimpleNCBITaxonomyLoader.html#readNode(java.io.BufferedReader)">readNode</A></B>(java.io.BufferedReader&nbsp;nodes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads the next entry from the nodes.dmp file and returns the corresponding 
 NCBITaxon object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojavax/bio/taxa/NCBITaxon.html" title="interface in org.biojavax.bio.taxa">NCBITaxon</A></CODE></FONT></TD>
<TD><CODE><B>NCBITaxonomyLoader.</B><B><A HREF="../../../../../../org/biojavax/bio/taxa/io/NCBITaxonomyLoader.html#readNode(java.io.BufferedReader)">readNode</A></B>(java.io.BufferedReader&nbsp;nodes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads the next entry from the nodes.dmp file and returns the corresponding 
 NCBITaxon object.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?org/biojava/bio/seq/io//class-useParseException.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="ParseException.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
