<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:56 BST 2008 -->
<TITLE>
Uses of Interface org.biojava.bio.seq.io.SymbolTokenization (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Interface org.biojava.bio.seq.io.SymbolTokenization (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?org/biojava/bio/seq/io//class-useSymbolTokenization.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SymbolTokenization.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Interface<br>org.biojava.bio.seq.io.SymbolTokenization</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program"><B>org.biojava.bio.program</B></A></TD>
<TD>
Java wrappers for interacting with external bioinformatics tools.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssaha"><B>org.biojava.bio.program.ssaha</B></A></TD>
<TD>SSAHA sequence searching API.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db"><B>org.biojava.bio.seq.db</B></A></TD>
<TD>
Collections of biological sequence data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.symbol"><B>org.biojava.bio.symbol</B></A></TD>
<TD>Representation of the Symbols that make up a sequence, and locations within
them.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq"><B>org.biojavax.bio.seq</B></A></TD>
<TD>Rich implementations of Sequences, Locations and Features.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq.io"><B>org.biojavax.bio.seq.io</B></A></TD>
<TD>Classes to support the I/O of RichSequence and 
     Bioentry objects.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojava/bio/program/package-summary.html">org.biojava.bio.program</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/program/package-summary.html">org.biojava.bio.program</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/program/Meme.html#Meme(java.io.InputStream, org.biojava.bio.seq.io.SymbolTokenization)">Meme</A></B>(java.io.InputStream&nbsp;is,
     <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>PhredFormat.</B><B><A HREF="../../../../../../org/biojava/bio/program/phred/PhredFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;siol)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssaha"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/program/ssaha/SequenceStreamer.FileStreamer.html#SequenceStreamer.FileStreamer(org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, java.io.File)">SequenceStreamer.FileStreamer</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
                              <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;toke,
                              java.io.File&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/program/ssaha/SequenceStreamer.FileStreamer.html#SequenceStreamer.FileStreamer(org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, java.util.List)">SequenceStreamer.FileStreamer</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
                              <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;toke,
                              java.util.List&nbsp;files)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> that return <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>TabIndexStore.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/TabIndexStore.html#getSymbolParser()">getSymbolParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>EmblCDROMIndexStore.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/EmblCDROMIndexStore.html#getSymbolParser()">getSymbolParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>BioIndex.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/BioIndex.html#getSymbolParser()">getSymbolParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>IndexStore.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/IndexStore.html#getSymbolParser()">getSymbolParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve the symbol parser used to turn the sequence characters
 into Symobl objects.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/EmblCDROMIndexStore.html#EmblCDROMIndexStore(java.io.File, java.io.File, java.io.File, org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SequenceBuilderFactory, org.biojava.bio.seq.io.SymbolTokenization)">EmblCDROMIndexStore</A></B>(java.io.File&nbsp;pathPrefix,
                    java.io.File&nbsp;divisionLkp,
                    java.io.File&nbsp;entryNamIdx,
                    <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
                    <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;factory,
                    <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;parser)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new <code>EmblCDROMIndexStore</code> backed by a
 random access binary index.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/EmblCDROMIndexStore.html#EmblCDROMIndexStore(java.io.File, java.io.File, org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SequenceBuilderFactory, org.biojava.bio.seq.io.SymbolTokenization)">EmblCDROMIndexStore</A></B>(java.io.File&nbsp;divisionLkp,
                    java.io.File&nbsp;entryNamIdx,
                    <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
                    <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;factory,
                    <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;parser)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new <code>EmblCDROMIndexStore</code> backed by a
 random access binary index.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/TabIndexStore.html#TabIndexStore(java.io.File, java.io.File, java.lang.String, org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SequenceBuilderFactory, org.biojava.bio.seq.io.SymbolTokenization)">TabIndexStore</A></B>(java.io.File&nbsp;storeFile,
              java.io.File&nbsp;indexFile,
              java.lang.String&nbsp;name,
              <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
              <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;sbFactory,
              <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symbolParser)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new TabIndexStore.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/AlternateTokenization.html" title="class in org.biojava.bio.seq.io">AlternateTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Implementation of SymbolTokenization which binds symbols to
 strings of characters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/CharacterTokenization.html" title="class in org.biojava.bio.seq.io">CharacterTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Implementation of SymbolTokenization which binds symbols to
 single unicode characters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/CrossProductTokenization.html" title="class in org.biojava.bio.seq.io">CrossProductTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Tokenization for cross-product alphabets.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/DoubleTokenization.html" title="class in org.biojava.bio.seq.io">DoubleTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/IntegerTokenization.html" title="class in org.biojava.bio.seq.io">IntegerTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/NameTokenization.html" title="class in org.biojava.bio.seq.io">NameTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Simple implementation of SymbolTokenization which uses the `name'
 field of the symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/SubIntegerTokenization.html" title="class in org.biojava.bio.seq.io">SubIntegerTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/WordTokenization.html" title="class in org.biojava.bio.seq.io">WordTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Base class for tokenizations which accept whitespace-separated
 `words'.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Read a fasta file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SequenceBuilderFactory)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;seqFactory)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Read a fasta file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EmblLikeFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/EmblLikeFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GAMEFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GAMEFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this version only reads annotations (no symbols)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>SequenceFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/FastaFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;siol)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads information from a flatfile to a <code>SeqIOListener</code>
 using a <code>SymbolTokenizer</code> to convert sequence strings
 to <code>Symbol</code> objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads a sequence from the specified reader using the Symbol
 parser and Sequence Factory provided.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankXmlFormat.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/GenbankXmlFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/StreamReader.html#StreamReader(java.io.BufferedReader, org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SequenceBuilderFactory)">StreamReader</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;sf)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/StreamReader.html#StreamReader(java.io.InputStream, org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SequenceBuilderFactory)">StreamReader</A></B>(java.io.InputStream&nbsp;is,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;sf)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.symbol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/symbol/SoftMaskedAlphabet.CaseSensitiveTokenization.html" title="class in org.biojava.bio.symbol">SoftMaskedAlphabet.CaseSensitiveTokenization</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This <code>SymbolTokenizer</code> works with a delegate to softmask
 symbol tokenization as appropriate.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> that return <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>IntegerAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/IntegerAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new parser (Mark Schreiber 3 May 2001).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>IntegerAlphabet.SubIntegerAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/IntegerAlphabet.SubIntegerAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>DoubleAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/DoubleAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>DoubleAlphabet.SubDoubleAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/DoubleAlphabet.SubDoubleAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>SingletonAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/SingletonAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>Alphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/Alphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Get a SymbolTokenization by name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>AbstractAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/AbstractAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>SoftMaskedAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/SoftMaskedAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractAlphabet.</B><B><A HREF="../../../../../../org/biojava/bio/symbol/AbstractAlphabet.html#putTokenization(java.lang.String, org.biojava.bio.seq.io.SymbolTokenization)">putTokenization</A></B>(java.lang.String&nbsp;name,
                <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;parser)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Assigns a symbol parser to a String object.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/symbol/SimpleSymbolList.html#SimpleSymbolList(org.biojava.bio.seq.io.SymbolTokenization, java.lang.String)">SimpleSymbolList</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;parser,
                 java.lang.String&nbsp;seqString)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Construct a SymbolList from a string.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A> that return <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getDNAParser()">getDNAParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a DNA symbol tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getNucleotideParser()">getNucleotideParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a nucleotide symbol tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getProteinParser()">getProteinParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a protein symbol tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#getRNAParser()">getRNAParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a RNA symbol tokenizer.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBL(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readEMBL</A></B>(java.io.BufferedReader&nbsp;br,
         <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
         <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
         <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a EMBL file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readEMBLxml(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readEMBLxml</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a EMBLxml file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.Namespace)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a fasta file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readFasta(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readFasta</A></B>(java.io.BufferedReader&nbsp;br,
          <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
          <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
          <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a fasta file building a custom type of <code>RichSequence</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readGenbank(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readGenbank</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a GenBank file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readINSDseq(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readINSDseq</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a INSDseq file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProt(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readUniProt</A></B>(java.io.BufferedReader&nbsp;br,
            <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
            <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
            <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a UniProt file using a custom type of SymbolList.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojavax/bio/seq/RichSequenceIterator.html" title="interface in org.biojavax.bio.seq">RichSequenceIterator</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.IOTools.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/RichSequence.IOTools.html#readUniProtXML(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">readUniProtXML</A></B>(java.io.BufferedReader&nbsp;br,
               <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;sTok,
               <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;seqFactory,
               <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a UniProt XML file using a custom type of SymbolList.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A> in <A HREF="../../../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A> that return <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 Always returns a protein tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 Always returns a DNA tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/INSDseqFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 Always returns a DNA tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/GenbankFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 Returns an dna parser if the letters DNA or RNA appear in the first line of the stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 Always returns a protein tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 Always returns a DNA tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/FastaFormat.html#guessSymbolTokenization(java.io.BufferedInputStream)">guessSymbolTokenization</A></B>(java.io.BufferedInputStream&nbsp;stream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the stream is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 Returns an protein parser if the first line of sequence contains any of F/L/I/P/Q/E, 
 otherwise returns a DNA tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 For formats that only accept one tokenization, just return it without
 checking the file. For formats that accept multiple tokenizations, its
 up to you how you do it.
 Always returns a protein tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 For formats that only accept one tokenization, just return it without
 checking the file. For formats that accept multiple tokenizations, its
 up to you how you do it.
 Always returns a DNA tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceFormat.BasicFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceFormat.BasicFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/INSDseqFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 For formats that only accept one tokenization, just return it without
 checking the file. For formats that accept multiple tokenizations, its
 up to you how you do it.
 Always returns a DNA tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/GenbankFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 For formats that only accept one tokenization, just return it without
 checking the file. For formats that accept multiple tokenizations, its
 up to you how you do it.
 Returns an dna parser if the letters DNA or RNA appear in the first line of the file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 For formats that only accept one tokenization, just return it without
 checking the file. For formats that accept multiple tokenizations, its
 up to you how you do it.
 Always returns a protein tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 For formats that only accept one tokenization, just return it without
 checking the file. For formats that accept multiple tokenizations, its
 up to you how you do it.
 Always returns a DNA tokenizer.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/FastaFormat.html#guessSymbolTokenization(java.io.File)">guessSymbolTokenization</A></B>(java.io.File&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;On the assumption that the file is readable by this format (not checked),
 attempt to guess which symbol tokenization we should use to read it.
 For formats that only accept one tokenization, just return it without
 checking the file. For formats that accept multiple tokenizations, its
 up to you how you do it.
 Returns an protein parser if the first line of sequence contains any of F/L/I/P/Q/E, 
 otherwise returns a DNA tokenizer.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.
 If namespace is null, then the namespace of the sequence in the fasta is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>RichSequenceFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;listener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/INSDseqFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/GenbankFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rlistener,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/FastaFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;rsiol,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.
 If namespace is null, then the namespace of the sequence in the fasta is used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtXMLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtXMLFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>INSDseqFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/INSDseqFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GenbankFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/GenbankFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>UniProtFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/UniProtFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>EMBLxmlFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/EMBLxmlFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>FastaFormat.</B><B><A HREF="../../../../../../org/biojavax/bio/seq/io/FastaFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichStreamReader.html#RichStreamReader(java.io.BufferedReader, org.biojavax.bio.seq.io.RichSequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">RichStreamReader</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceFormat.html" title="interface in org.biojavax.bio.seq.io">RichSequenceFormat</A>&nbsp;format,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;sf,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new stream reader on the given reader, which will attempt to read
 sequences in the given format, having symbols from the given tokenization, and
 pass them to the given factory to be transformed into RichSequence objects in
 the given namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/seq/io/RichStreamReader.html#RichStreamReader(java.io.InputStream, org.biojavax.bio.seq.io.RichSequenceFormat, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSequenceBuilderFactory, org.biojavax.Namespace)">RichStreamReader</A></B>(java.io.InputStream&nbsp;is,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceFormat.html" title="interface in org.biojavax.bio.seq.io">RichSequenceFormat</A>&nbsp;format,
                 <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../../../org/biojavax/bio/seq/io/RichSequenceBuilderFactory.html" title="interface in org.biojavax.bio.seq.io">RichSequenceBuilderFactory</A>&nbsp;sf,
                 <A HREF="../../../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new stream reader on the given input stream, which will attempt to read
 sequences in the given format, having symbols from the given tokenization, and
 pass them to the given factory to be transformed into RichSequence objects in
 the given namespace.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?org/biojava/bio/seq/io//class-useSymbolTokenization.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SymbolTokenization.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
