<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:06 BST 2008 -->
<TITLE>
AbstractChromatogram (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojava.bio.chromatogram.AbstractChromatogram class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="AbstractChromatogram (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AbstractChromatogram.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/chromatogram/AbstractChromatogram.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AbstractChromatogram.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.biojava.bio.chromatogram</FONT>
<BR>
Class AbstractChromatogram</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>org.biojava.bio.chromatogram.AbstractChromatogram</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></DD>
</DL>
<DL>
<DT><B>Direct Known Subclasses:</B> <DD><A HREF="../../../../org/biojava/bio/program/abi/ABIFChromatogram.html" title="class in org.biojava.bio.program.abi">ABIFChromatogram</A>, <A HREF="../../../../org/biojava/bio/program/scf/SCF.html" title="class in org.biojava.bio.program.scf">SCF</A>, <A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html" title="class in org.biojava.bio.chromatogram">SimpleChromatogram</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public abstract class <B>AbstractChromatogram</B><DT>extends java.lang.Object<DT>implements <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></DL>
</PRE>

<P>
A basic, abstract implementation of <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram"><CODE>Chromatogram</CODE></A>.  Provides
 protected setters so that subclasses may set the value of the various
 properties of a chromatogram.

 Chromatograms should be created using <A HREF="../../../../org/biojava/bio/chromatogram/ChromatogramFactory.html" title="class in org.biojava.bio.chromatogram"><CODE>ChromatogramFactory</CODE></A> or a
 parser for a particular file format.
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.3</DD>
<DT><B>Author:</B></DT>
  <DD>Rhett Sutphin (<a href="http://genome.uiowa.edu/">UI CBCB</a>), Matthew Pocock</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.biojava.bio.chromatogram.Chromatogram"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.biojava.bio.chromatogram.<A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#DNA">DNA</A>, <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#OFFSETS">OFFSETS</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#AbstractChromatogram()">AbstractChromatogram</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new AbstractChromatogram.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#clearTraces()">clearTraces</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the trace data structures to null.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#createImmutableAlignment(java.util.Map)">createImmutableAlignment</A></B>(java.util.Map&nbsp;labelsToSymLists)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A factory method for creating new immutable alignments, particularly
 for use as base call alignments.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#createImmutableSymbolList(org.biojava.bio.symbol.Alphabet, java.util.List)">createImmutableSymbolList</A></B>(<A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha,
                          java.util.List&nbsp;syms)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A factory method for creating new symbol lists with a given alphabet.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getBaseCalls()">getBaseCalls</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the total number of base calls.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getMax()">getMax</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the max intensity from all the traces.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getMax(org.biojava.bio.symbol.AtomicSymbol)">getMax</A></B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the max intensity on the trace for the specified nucleotide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getSequenceLength()">getSequenceLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the sequence length.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getSignificantBits()">getSignificantBits</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the number of significant bits.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)">getTrace</A></B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an array containing the intensities of the sampled waveform
  representing the chromatogram trace for base <code>nucleotide</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getTraceLength()">getTraceLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the length of the trace of the Chromatogram.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;int[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverse(int[])">reverse</A></B>(int[]&nbsp;src)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Utility method for reversing an int[] array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplement()">reverseComplement</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a new <code>Chromatogram</code> representing the reverse
 complement of this one.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementBaseCallList(java.lang.Object)">reverseComplementBaseCallList</A></B>(java.lang.Object&nbsp;label)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return a symbol list containing the reverse complement of the base call
 data for the given label.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementBaseCalls()">reverseComplementBaseCalls</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a new base call alignment that is the reverse complement of
 one in this chromatogram.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected abstract &nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementInstance()">reverseComplementInstance</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a new instance of this AbstractChromatogram subclass for use in
 <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplement()"><CODE>reverseComplement()</CODE></A>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#setBaseCallAlignment(org.biojava.bio.symbol.Alignment)">setBaseCallAlignment</A></B>(<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Provides the list of base calls.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#setBits(int)">setBits</A></B>(int&nbsp;bits)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the number of significant bits in the trace samples.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#setTrace(org.biojava.bio.symbol.AtomicSymbol, int[], int)">setTrace</A></B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nuc,
         int[]&nbsp;trace,
         int&nbsp;maxVal)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Provides the trace samples for a particular nucleotide.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="AbstractChromatogram()"><!-- --></A><H3>
AbstractChromatogram</H3>
<PRE>
public <B>AbstractChromatogram</B>()</PRE>
<DL>
<DD>Create a new AbstractChromatogram.
<P>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getTrace(org.biojava.bio.symbol.AtomicSymbol)"><!-- --></A><H3>
getTrace</H3>
<PRE>
public int[] <B>getTrace</B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)
               throws <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)">Chromatogram</A></CODE></B></DD>
<DD>Returns an array containing the intensities of the sampled waveform
  representing the chromatogram trace for base <code>nucleotide</code>.  
  This may be a reference the actual internal representation of the 
  samples, so callers <b>must not modify it</b>.
  <p>
  The resulting array for each nucleotide must be <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTraceLength()"><CODE>Chromatogram.getTraceLength()</CODE></A>
  <code>int</code>s long.
  </p>
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)">getTrace</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nucleotide</CODE> - the trace to examine.  Must be the symbol for A, C, G, or T
         as provided by <A HREF="../../../../org/biojava/bio/seq/DNATools.html" title="class in org.biojava.bio.seq"><CODE>DNATools</CODE></A>
<DT><B>Returns:</B><DD>an array of integers representing the values of a particular 
          chromatogram trace.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - if <code>nucleotide</code> isn't in the DNA alphabet</DL>
</DD>
</DL>
<HR>

<A NAME="getTraceLength()"><!-- --></A><H3>
getTraceLength</H3>
<PRE>
public int <B>getTraceLength</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTraceLength()">Chromatogram</A></CODE></B></DD>
<DD>Returns the length of the trace of the Chromatogram.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTraceLength()">getTraceLength</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the number of samples in the trace<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)"><CODE>Chromatogram.getTrace(AtomicSymbol)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getMax()"><!-- --></A><H3>
getMax</H3>
<PRE>
public int <B>getMax</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax()">Chromatogram</A></CODE></B></DD>
<DD>Gets the max intensity from all the traces.  Must be equivalent
  to the max of calling <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax(org.biojava.bio.symbol.AtomicSymbol)"><CODE>Chromatogram.getMax(AtomicSymbol)</CODE></A> on each
  of the four non-ambiguous DNA nucleotides.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax()">getMax</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the max intensity</DL>
</DD>
</DL>
<HR>

<A NAME="getMax(org.biojava.bio.symbol.AtomicSymbol)"><!-- --></A><H3>
getMax</H3>
<PRE>
public int <B>getMax</B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)
           throws <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax(org.biojava.bio.symbol.AtomicSymbol)">Chromatogram</A></CODE></B></DD>
<DD>Gets the max intensity on the trace for the specified nucleotide.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax(org.biojava.bio.symbol.AtomicSymbol)">getMax</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nucleotide</CODE> - the trace to examine.  Must be a concrete 
         (non-ambiguous) nucleotide from the DNA alphabet
<DT><B>Returns:</B><DD>the max intensity
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - when the nucleotide isn't from the DNA
          alphabet</DL>
</DD>
</DL>
<HR>

<A NAME="getBaseCalls()"><!-- --></A><H3>
getBaseCalls</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> <B>getBaseCalls</B>()</PRE>
<DL>
<DD>Return the total number of base calls.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getBaseCalls()">getBaseCalls</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the total number of base calls</DL>
</DD>
</DL>
<HR>

<A NAME="getSequenceLength()"><!-- --></A><H3>
getSequenceLength</H3>
<PRE>
public int <B>getSequenceLength</B>()</PRE>
<DL>
<DD>Return the sequence length.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getSequenceLength()">getSequenceLength</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the sequence length</DL>
</DD>
</DL>
<HR>

<A NAME="getSignificantBits()"><!-- --></A><H3>
getSignificantBits</H3>
<PRE>
public int <B>getSignificantBits</B>()</PRE>
<DL>
<DD>Return the number of significant bits.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getSignificantBits()">getSignificantBits</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the significant bits</DL>
</DD>
</DL>
<HR>

<A NAME="setBaseCallAlignment(org.biojava.bio.symbol.Alignment)"><!-- --></A><H3>
setBaseCallAlignment</H3>
<PRE>
protected final void <B>setBaseCallAlignment</B>(<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)
                                   throws <A HREF="../../../../org/biojava/bio/symbol/IllegalAlphabetException.html" title="class in org.biojava.bio.symbol">IllegalAlphabetException</A>,
                                          java.lang.IllegalArgumentException,
                                          java.util.NoSuchElementException</PRE>
<DL>
<DD>Provides the list of base calls.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>align</CODE> - the base call alignment
<DT><B>Throws:</B>
<DD><CODE>java.util.NoSuchElementException</CODE> - when align doesn't contain alignments with
         the required DNA and OFFSETS labels
<DD><CODE>java.lang.IllegalArgumentException</CODE> - the lists in align aren't all the same length.
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalAlphabetException.html" title="class in org.biojava.bio.symbol">IllegalAlphabetException</A></CODE> - if the required lists don't have the
         correct alphabets.  See the documentation of
         <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getBaseCalls()"><CODE>Chromatogram.getBaseCalls()</CODE></A> for details.<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getBaseCalls()"><CODE>Chromatogram.getBaseCalls()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="clearTraces()"><!-- --></A><H3>
clearTraces</H3>
<PRE>
protected final void <B>clearTraces</B>()</PRE>
<DL>
<DD>Sets the trace data structures to null.  If a subclass needs to replace
 the traces with new traces of a different length, this method must be
 called first to avoid provoking an IllegalArgumentException from setTrace
 due to a length mismatch.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setTrace(org.biojava.bio.symbol.AtomicSymbol, int[], int)"><!-- --></A><H3>
setTrace</H3>
<PRE>
protected final void <B>setTrace</B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nuc,
                              int[]&nbsp;trace,
                              int&nbsp;maxVal)
                       throws java.lang.IllegalArgumentException,
                              <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD>Provides the trace samples for a particular nucleotide.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nuc</CODE> - A DNA nucleotide<DD><CODE>trace</CODE> - the trace samples themselves<DD><CODE>maxVal</CODE> - the maximum value in the trace array.  If this value
        is <CODE>Integer.MIN_VALUE</CODE>, this method will do a linear
        search of trace to determine the max.
<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalArgumentException</CODE> - when trace.length is different
         from any of the existing (non-null) traces
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - when nuc is not a concrete DNA nucleotide</DL>
</DD>
</DL>
<HR>

<A NAME="setBits(int)"><!-- --></A><H3>
setBits</H3>
<PRE>
protected final void <B>setBits</B>(int&nbsp;bits)
                      throws java.lang.IllegalArgumentException</PRE>
<DL>
<DD>Sets the number of significant bits in the trace samples.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>bits</CODE> - a non-negative integer indicating the number of
         significant bits in each trace sample
<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalArgumentException</CODE> - when <code>bits</code> is negative</DL>
</DD>
</DL>
<HR>

<A NAME="reverseComplementInstance()"><!-- --></A><H3>
reverseComplementInstance</H3>
<PRE>
protected abstract <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A> <B>reverseComplementInstance</B>()</PRE>
<DL>
<DD>Returns a new instance of this AbstractChromatogram subclass for use in
 <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplement()"><CODE>reverseComplement()</CODE></A>.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a reverse-complemented AbstractChromatogram</DL>
</DD>
</DL>
<HR>

<A NAME="reverseComplement()"><!-- --></A><H3>
reverseComplement</H3>
<PRE>
public <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A> <B>reverseComplement</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#reverseComplement()">Chromatogram</A></CODE></B></DD>
<DD>Returns a new <code>Chromatogram</code> representing the reverse
 complement of this one.
 <p>
 Implementors should copy the metadata about the chromatogram (i.e., base 
 calls) as is appropriate to their formats.
 </p>
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#reverseComplement()">reverseComplement</A></CODE> in interface <CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a new chromatogram that is the reverse complement of this one</DL>
</DD>
</DL>
<HR>

<A NAME="reverseComplementBaseCalls()"><!-- --></A><H3>
reverseComplementBaseCalls</H3>
<PRE>
protected <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> <B>reverseComplementBaseCalls</B>()</PRE>
<DL>
<DD>Returns a new base call alignment that is the reverse complement of
 one in this chromatogram.  This is achieved by calling
 <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementBaseCallList(java.lang.Object)"><CODE>reverseComplementBaseCallList(java.lang.Object)</CODE></A> for each label in the current
 base call alignment.  When that method returns null, no list will
 appear in reverse complement base call alignment with the null-provoking
 label.  For this reason, subclasses are encouraged to override
 <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementBaseCallList(java.lang.Object)"><CODE>reverseComplementBaseCallList(java.lang.Object)</CODE></A> to handle any additional per-base
 metadata that they store.
 <p>
 This implementation should be safely inheritable
 for all chromatogram implementations, unless one just doesn't want
 base calls on its reverse complement output.  If this is the case,
 it should override this method to return null.
 </p>
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a new <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol"><CODE>Alignment</CODE></A> that is the reverse complement of the
          one in the current chromatogram</DL>
</DD>
</DL>
<HR>

<A NAME="reverseComplementBaseCallList(java.lang.Object)"><!-- --></A><H3>
reverseComplementBaseCallList</H3>
<PRE>
protected <A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A> <B>reverseComplementBaseCallList</B>(java.lang.Object&nbsp;label)</PRE>
<DL>
<DD>Return a symbol list containing the reverse complement of the base call
 data for the given label.  The returned list will be stored in the
 reverse complement's base call alignment under the same label.
 <p>
 Implementation note: subclasses which do not use an <A HREF="../../../../org/biojava/bio/symbol/IntegerAlphabet.html" title="class in org.biojava.bio.symbol"><CODE>IntegerAlphabet</CODE></A> for
 their offsets lists must override this method, at least for the case where
 <code>label == <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#OFFSETS"><CODE>Chromatogram.OFFSETS</CODE></A></code>.
 </p>
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>label</CODE> - the label Object
<DT><B>Returns:</B><DD>an appropriately reverse-complemented SymbolList, or null if the
          label is unhandled.</DL>
</DD>
</DL>
<HR>

<A NAME="createImmutableSymbolList(org.biojava.bio.symbol.Alphabet, java.util.List)"><!-- --></A><H3>
createImmutableSymbolList</H3>
<PRE>
protected <A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A> <B>createImmutableSymbolList</B>(<A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha,
                                               java.util.List&nbsp;syms)
                                        throws <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A>,
                                               java.lang.ClassCastException</PRE>
<DL>
<DD>A factory method for creating new symbol lists with a given alphabet.
 The default implementation should be fine for nearly all cases, but the
 option is given in case a developer wishes to use a more memory
 efficient implementation.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>alpha</CODE> - the <A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol"><CODE>Alphabet</CODE></A> for the new list<DD><CODE>syms</CODE> - the symbols to put in the new list
<DT><B>Returns:</B><DD>a new immutable SymbolList containing all the given symbols using
          the given Alphabet
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - when alpha and syms are incompatible
<DD><CODE>java.lang.ClassCastException</CODE> - when any object in syms isn't a Symbol</DL>
</DD>
</DL>
<HR>

<A NAME="createImmutableAlignment(java.util.Map)"><!-- --></A><H3>
createImmutableAlignment</H3>
<PRE>
protected <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> <B>createImmutableAlignment</B>(java.util.Map&nbsp;labelsToSymLists)
                                      throws java.lang.IllegalArgumentException,
                                             java.lang.ClassCastException</PRE>
<DL>
<DD>A factory method for creating new immutable alignments, particularly
 for use as base call alignments.  The default implementation should
 be fine for nearly all cases.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>labelsToSymLists</CODE> - a <CODE>Map</CODE> whose keys are desired labels
        for the alignment and whose values are the SymbolLists.
        All the SymbolLists must be the same length.
<DT><B>Returns:</B><DD>a new Alignment
<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalArgumentException</CODE> - if the lists aren't all the same length
<DD><CODE>java.lang.ClassCastException</CODE> - if any of the values in the map aren't
         SymbolLists</DL>
</DD>
</DL>
<HR>

<A NAME="reverse(int[])"><!-- --></A><H3>
reverse</H3>
<PRE>
protected static final int[] <B>reverse</B>(int[]&nbsp;src)</PRE>
<DL>
<DD>Utility method for reversing an int[] array. Visible for subclass use.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>src</CODE> - the source array
<DT><B>Returns:</B><DD>an array of the same length containing all values in src in
     reverse order</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AbstractChromatogram.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/chromatogram/AbstractChromatogram.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AbstractChromatogram.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
