<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:06 BST 2008 -->
<TITLE>
Chromatogram (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojava.bio.chromatogram.Chromatogram interface">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Chromatogram (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Chromatogram.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/ChromatogramFactory.html" title="class in org.biojava.bio.chromatogram"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/chromatogram/Chromatogram.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromatogram.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.biojava.bio.chromatogram</FONT>
<BR>
Interface Chromatogram</H2>
<DL>
<DT><B>All Known Implementing Classes:</B> <DD><A HREF="../../../../org/biojava/bio/program/abi/ABIFChromatogram.html" title="class in org.biojava.bio.program.abi">ABIFChromatogram</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A>, <A HREF="../../../../org/biojava/bio/program/scf/SCF.html" title="class in org.biojava.bio.program.scf">SCF</A>, <A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html" title="class in org.biojava.bio.chromatogram">SimpleChromatogram</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public interface <B>Chromatogram</B></DL>
</PRE>

<P>
Encapsulates the basic information you would want from a chromatogram.
 Read-only.
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.3</DD>
<DT><B>Author:</B></DT>
  <DD>Rhett Sutphin (<a href="http://genome.uiowa.edu/">UI CBCB</a>), Matthew Pocock</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#DNA">DNA</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The sequence label for the list of called bases.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#OFFSETS">OFFSETS</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The sequence label for the trace offsets of the called bases.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getBaseCalls()">getBaseCalls</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an alignment that describes the base calls for this chromatogram.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax()">getMax</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the max intensity from all the traces.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax(org.biojava.bio.symbol.AtomicSymbol)">getMax</A></B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the max intensity on the trace for the specified nucleotide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getSequenceLength()">getSequenceLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of bases called by whatever base-calling software
 analyzed the chromatogram as loaded.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getSignificantBits()">getSignificantBits</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of bits of the traces which are significant.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)">getTrace</A></B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an array containing the intensities of the sampled waveform
  representing the chromatogram trace for base <code>nucleotide</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTraceLength()">getTraceLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the length of the trace of the Chromatogram.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#reverseComplement()">reverseComplement</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a new <code>Chromatogram</code> representing the reverse
 complement of this one.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="DNA"><!-- --></A><H3>
DNA</H3>
<PRE>
static final java.lang.Object <B>DNA</B></PRE>
<DL>
<DD>The sequence label for the list of called bases.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="OFFSETS"><!-- --></A><H3>
OFFSETS</H3>
<PRE>
static final java.lang.Object <B>OFFSETS</B></PRE>
<DL>
<DD>The sequence label for the trace offsets of the called bases.
<P>
<DL>
</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getMax()"><!-- --></A><H3>
getMax</H3>
<PRE>
int <B>getMax</B>()</PRE>
<DL>
<DD>Gets the max intensity from all the traces.  Must be equivalent
  to the max of calling <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getMax(org.biojava.bio.symbol.AtomicSymbol)"><CODE>getMax(AtomicSymbol)</CODE></A> on each
  of the four non-ambiguous DNA nucleotides.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the max intensity</DL>
</DD>
</DL>
<HR>

<A NAME="getMax(org.biojava.bio.symbol.AtomicSymbol)"><!-- --></A><H3>
getMax</H3>
<PRE>
int <B>getMax</B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)
           throws <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD>Gets the max intensity on the trace for the specified nucleotide.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nucleotide</CODE> - the trace to examine.  Must be a concrete 
         (non-ambiguous) nucleotide from the DNA alphabet
<DT><B>Returns:</B><DD>the max intensity
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - when the nucleotide isn't from the DNA
          alphabet</DL>
</DD>
</DL>
<HR>

<A NAME="getTraceLength()"><!-- --></A><H3>
getTraceLength</H3>
<PRE>
int <B>getTraceLength</B>()</PRE>
<DL>
<DD>Returns the length of the trace of the Chromatogram.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of samples in the trace<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)"><CODE>getTrace(AtomicSymbol)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getTrace(org.biojava.bio.symbol.AtomicSymbol)"><!-- --></A><H3>
getTrace</H3>
<PRE>
int[] <B>getTrace</B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nucleotide)
               throws <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD>Returns an array containing the intensities of the sampled waveform
  representing the chromatogram trace for base <code>nucleotide</code>.  
  This may be a reference the actual internal representation of the 
  samples, so callers <b>must not modify it</b>.
  <p>
  The resulting array for each nucleotide must be <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getTraceLength()"><CODE>getTraceLength()</CODE></A>
  <code>int</code>s long.
  </p>
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nucleotide</CODE> - the trace to examine.  Must be the symbol for A, C, G, or T
         as provided by <A HREF="../../../../org/biojava/bio/seq/DNATools.html" title="class in org.biojava.bio.seq"><CODE>DNATools</CODE></A>
<DT><B>Returns:</B><DD>an array of integers representing the values of a particular 
          chromatogram trace.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - if <code>nucleotide</code> isn't in the DNA alphabet</DL>
</DD>
</DL>
<HR>

<A NAME="getSignificantBits()"><!-- --></A><H3>
getSignificantBits</H3>
<PRE>
int <B>getSignificantBits</B>()</PRE>
<DL>
<DD>Returns the number of bits of the traces which are significant.  For 
 instance, if the chromatogram were originally encoded with a single byte 
 per trace sample, this method must return 8.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of significant bits</DL>
</DD>
</DL>
<HR>

<A NAME="getBaseCalls()"><!-- --></A><H3>
getBaseCalls</H3>
<PRE>
<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> <B>getBaseCalls</B>()</PRE>
<DL>
<DD>Returns an alignment that describes the base calls for this chromatogram.
 All of the <code>SymbolList</code>s in this alignment must be the same
 length and that length must equal <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getSequenceLength()"><CODE>getSequenceLength()</CODE></A>.
 <p>
 The alignment must contain, at the least, two sequences:
 </p>
 <ol>
   <li>A sequence containing the called bases.  The alphabet of this list
       must be <A HREF="../../../../org/biojava/bio/seq/DNATools.html#getDNA()"><CODE>DNATools.getDNA()</CODE></A>.  
       The label for this list in the alignment must be 
       <code>Chromatogram.DNA</code></li>
   <li>A sequence containing the trace offsets at which the called bases 
       were called.  The alphabet of this list must be an 
       <A HREF="../../../../org/biojava/bio/symbol/IntegerAlphabet.html" title="class in org.biojava.bio.symbol"><CODE>IntegerAlphabet</CODE></A> or a
       <A HREF="../../../../org/biojava/bio/symbol/IntegerAlphabet.SubIntegerAlphabet.html" title="class in org.biojava.bio.symbol"><CODE>IntegerAlphabet.SubIntegerAlphabet</CODE></A>.
       The label for this list in the alignment must be 
       <code>Chromatogram.OFFSETS</code>.</li>
 </ol>
 <p>
 Implementors may provide other sequences as they see fit.
 </p>
<P>
<DD><DL>

<DT><B>Returns:</B><DD>an alignment of at least two sequences, as described above.</DL>
</DD>
</DL>
<HR>

<A NAME="getSequenceLength()"><!-- --></A><H3>
getSequenceLength</H3>
<PRE>
int <B>getSequenceLength</B>()</PRE>
<DL>
<DD>Returns the number of bases called by whatever base-calling software
 analyzed the chromatogram as loaded.  Must equal 
 <code><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getBaseCalls()"><CODE>getBaseCalls()</CODE></A>.length()</code>.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of bases</DL>
</DD>
</DL>
<HR>

<A NAME="reverseComplement()"><!-- --></A><H3>
reverseComplement</H3>
<PRE>
<A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A> <B>reverseComplement</B>()</PRE>
<DL>
<DD>Returns a new <code>Chromatogram</code> representing the reverse
 complement of this one.
 <p>
 Implementors should copy the metadata about the chromatogram (i.e., base 
 calls) as is appropriate to their formats.
 </p>
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a new chromatogram that is the reverse complement of this one</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Chromatogram.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/ChromatogramFactory.html" title="class in org.biojava.bio.chromatogram"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/chromatogram/Chromatogram.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromatogram.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
