<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:07 BST 2008 -->
<TITLE>
SimpleChromatogram (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojava.bio.chromatogram.SimpleChromatogram class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="SimpleChromatogram (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SimpleChromatogram.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/ChromatogramTools.html" title="class in org.biojava.bio.chromatogram"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/UnsupportedChromatogramFormatException.html" title="class in org.biojava.bio.chromatogram"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/chromatogram/SimpleChromatogram.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SimpleChromatogram.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.biojava.bio.chromatogram</FONT>
<BR>
Class SimpleChromatogram</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">org.biojava.bio.chromatogram.AbstractChromatogram</A>
      <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>org.biojava.bio.chromatogram.SimpleChromatogram</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>SimpleChromatogram</B><DT>extends <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A></DL>
</PRE>

<P>
A basic chromatogram implementation which provides public mutators 
 for setting the various attributes of the chromatogram.  
 <p>
 In general, <b>new chromatogram implementations</b> should be derived from
 <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram"><CODE>AbstractChromatogram</CODE></A>, <b>not this class</b>, as it is generally 
 undesirable to allow the internal structures of a <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram"><CODE>Chromatogram</CODE></A> to be
 manipulated externally. This class could still be useful, however, for 
 programatically generated "chromatograms".
 </p>
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.3</DD>
<DT><B>Author:</B></DT>
  <DD>Rhett Sutphin (<a href="http://genome.uiowa.edu/">UI CBCB</a>), Matthew Pocock</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.biojava.bio.chromatogram.Chromatogram"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.biojava.bio.chromatogram.<A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html" title="interface in org.biojava.bio.chromatogram">Chromatogram</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#DNA">DNA</A>, <A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#OFFSETS">OFFSETS</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html#SimpleChromatogram()">SimpleChromatogram</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new instance of SimpleChromatogram.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html#clearTraceValues()">clearTraceValues</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets all the traces to null.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html#reverseComplementInstance()">reverseComplementInstance</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a new instance of this AbstractChromatogram subclass for use in
 <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplement()"><CODE>AbstractChromatogram.reverseComplement()</CODE></A>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html#setSignificantBits(int)">setSignificantBits</A></B>(int&nbsp;bits)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the number of significant bits in the data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html#setSymbolLists(org.biojava.bio.symbol.SymbolList, org.biojava.bio.symbol.SymbolList)">setSymbolLists</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;dna,
               <A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;offsets)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set the DNA and OFFSETS symbol lists for the basecall alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html#setTraceValues(org.biojava.bio.symbol.AtomicSymbol, int[], int)">setTraceValues</A></B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nuc,
               int[]&nbsp;trace,
               int&nbsp;maxVal)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the trace array for one of the DNA nucleotides.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.biojava.bio.chromatogram.AbstractChromatogram"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class org.biojava.bio.chromatogram.<A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#clearTraces()">clearTraces</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#createImmutableAlignment(java.util.Map)">createImmutableAlignment</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#createImmutableSymbolList(org.biojava.bio.symbol.Alphabet, java.util.List)">createImmutableSymbolList</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getBaseCalls()">getBaseCalls</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getMax()">getMax</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getMax(org.biojava.bio.symbol.AtomicSymbol)">getMax</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getSequenceLength()">getSequenceLength</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getSignificantBits()">getSignificantBits</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)">getTrace</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getTraceLength()">getTraceLength</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverse(int[])">reverse</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplement()">reverseComplement</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementBaseCallList(java.lang.Object)">reverseComplementBaseCallList</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementBaseCalls()">reverseComplementBaseCalls</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#setBaseCallAlignment(org.biojava.bio.symbol.Alignment)">setBaseCallAlignment</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#setBits(int)">setBits</A>, <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#setTrace(org.biojava.bio.symbol.AtomicSymbol, int[], int)">setTrace</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="SimpleChromatogram()"><!-- --></A><H3>
SimpleChromatogram</H3>
<PRE>
public <B>SimpleChromatogram</B>()</PRE>
<DL>
<DD>Creates a new instance of SimpleChromatogram.
<P>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="setSymbolLists(org.biojava.bio.symbol.SymbolList, org.biojava.bio.symbol.SymbolList)"><!-- --></A><H3>
setSymbolLists</H3>
<PRE>
public void <B>setSymbolLists</B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;dna,
                           <A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;offsets)
                    throws <A HREF="../../../../org/biojava/bio/symbol/IllegalAlphabetException.html" title="class in org.biojava.bio.symbol">IllegalAlphabetException</A>,
                           java.lang.IllegalArgumentException</PRE>
<DL>
<DD>Set the DNA and OFFSETS symbol lists for the basecall alignment.  Beware:
 this method does no consistency checks to be sure that all the offsets
 are valid indices into the trace arrays.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>dna</CODE> - a symbol list in the DNA alphabet that contains the base calls
        for this chromatogram<DD><CODE>offsets</CODE> - a symbol list in an integer or sub-integer alphabet that
        contains the locations in the chromatogram for the bases called 
        in <code>dna</code>
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalAlphabetException.html" title="class in org.biojava.bio.symbol">IllegalAlphabetException</A></CODE> - when the alphabets aren't as specified
<DD><CODE>java.lang.IllegalArgumentException</CODE> - when the lists aren't the same length</DL>
</DD>
</DL>
<HR>

<A NAME="setTraceValues(org.biojava.bio.symbol.AtomicSymbol, int[], int)"><!-- --></A><H3>
setTraceValues</H3>
<PRE>
public void <B>setTraceValues</B>(<A HREF="../../../../org/biojava/bio/symbol/AtomicSymbol.html" title="interface in org.biojava.bio.symbol">AtomicSymbol</A>&nbsp;nuc,
                           int[]&nbsp;trace,
                           int&nbsp;maxVal)
                    throws java.lang.IllegalArgumentException,
                           <A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></PRE>
<DL>
<DD>Sets the trace array for one of the DNA nucleotides.  The provided
 array <i>will not be copied</i>, so any modifications to it will be
 reflected in calls to <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getTrace(org.biojava.bio.symbol.AtomicSymbol)"><CODE>AbstractChromatogram.getTrace(org.biojava.bio.symbol.AtomicSymbol)</CODE></A>.
 <p>
 If you need to set a new set of traces whose length is different
 from the old set, you must call <A HREF="../../../../org/biojava/bio/chromatogram/SimpleChromatogram.html#clearTraceValues()"><CODE>clearTraceValues()</CODE></A> first,
 or you will provoke an <code>IllegalArgumentException</code>.
 </p>
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>nuc</CODE> - the nucleotide for which to set the trace<DD><CODE>trace</CODE> - the sampled intensities along the trace<DD><CODE>maxVal</CODE> - the maximum value on the trace, or <CODE>Integer.MIN_VALUE</CODE>
        to force this method to calculate it
<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalArgumentException</CODE> - when trace.length is different
         from any of the existing (non-null) traces
<DD><CODE><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></CODE> - when nuc is not a concrete DNA nucleotide</DL>
</DD>
</DL>
<HR>

<A NAME="clearTraceValues()"><!-- --></A><H3>
clearTraceValues</H3>
<PRE>
public void <B>clearTraceValues</B>()</PRE>
<DL>
<DD>Sets all the traces to null.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setSignificantBits(int)"><!-- --></A><H3>
setSignificantBits</H3>
<PRE>
public void <B>setSignificantBits</B>(int&nbsp;bits)</PRE>
<DL>
<DD>Sets the number of significant bits in the data.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>bits</CODE> - how many bits of the trace samples are significant<DT><B>See Also:</B><DD><A HREF="../../../../org/biojava/bio/chromatogram/Chromatogram.html#getSignificantBits()"><CODE>Chromatogram.getSignificantBits()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="reverseComplementInstance()"><!-- --></A><H3>
reverseComplementInstance</H3>
<PRE>
protected <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A> <B>reverseComplementInstance</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementInstance()">AbstractChromatogram</A></CODE></B></DD>
<DD>Returns a new instance of this AbstractChromatogram subclass for use in
 <A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplement()"><CODE>AbstractChromatogram.reverseComplement()</CODE></A>.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementInstance()">reverseComplementInstance</A></CODE> in class <CODE><A HREF="../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html" title="class in org.biojava.bio.chromatogram">AbstractChromatogram</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a reverse-complemented AbstractChromatogram</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/SimpleChromatogram.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/ChromatogramTools.html" title="class in org.biojava.bio.chromatogram"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/chromatogram/UnsupportedChromatogramFormatException.html" title="class in org.biojava.bio.chromatogram"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/chromatogram/SimpleChromatogram.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SimpleChromatogram.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
