<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:51:06 BST 2008 -->
<TITLE>
Uses of Package org.biojava.bio.seq (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Package org.biojava.bio.seq (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/seq/package-use.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Package<br>org.biojava.bio.seq</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.alignment"><B>org.biojava.bio.alignment</B></A></TD>
<TD>
Classes to generate and describe sequence alignments.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dist"><B>org.biojava.bio.dist</B></A></TD>
<TD>
Probability distributions over Alphabets.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp"><B>org.biojava.bio.dp</B></A></TD>
<TD>HMM and Dynamic Programming Algorithms.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.gui.sequence"><B>org.biojava.bio.gui.sequence</B></A></TD>
<TD>
Graphical displays of biological sequences and associated annotations.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.gui.sequence.tracklayout"><B>org.biojava.bio.gui.sequence.tracklayout</B></A></TD>
<TD>Classes for the handling of the layout of a WrappedSequencePanel.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.molbio"><B>org.biojava.bio.molbio</B></A></TD>
<TD>The classes and interfaces in this package cover common molecular
biological techniques such as restriction digests and PCR.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.das"><B>org.biojava.bio.program.das</B></A></TD>
<TD>Development client for the Distributed Annotation System.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff"><B>org.biojava.bio.program.gff</B></A></TD>
<TD>GFF manipulation.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff3"><B>org.biojava.bio.program.gff3</B></A></TD>
<TD>
Support for reading and writing GFF3.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssbind"><B>org.biojava.bio.program.ssbind</B></A></TD>
<TD>
Creation of objects from SAX events using the BioJava BlastLikeDataSetCollection DTD.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.unigene"><B>org.biojava.bio.program.unigene</B></A></TD>
<TD>Objects for representing Unigene clusters.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.xff"><B>org.biojava.bio.program.xff</B></A></TD>
<TD>Event-driven parsing system for the Extensible Feature Format (XFF).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.proteomics"><B>org.biojava.bio.proteomics</B></A></TD>
<TD>Utilities to aid in performing various physical analysis of proteins.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.search"><B>org.biojava.bio.search</B></A></TD>
<TD>
Interfaces and classes for representing sequence similarity search results.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq"><B>org.biojava.bio.seq</B></A></TD>
<TD>Classes and interfaces for defining biological sequences and informatics
objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db"><B>org.biojava.bio.seq.db</B></A></TD>
<TD>
Collections of biological sequence data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biofetch"><B>org.biojava.bio.seq.db.biofetch</B></A></TD>
<TD>Client for the OBDA BioFetch protocol.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biosql"><B>org.biojava.bio.seq.db.biosql</B></A></TD>
<TD>General purpose Sequence storage in a relational database.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.flat"><B>org.biojava.bio.seq.db.flat</B></A></TD>
<TD>
Support for OBDA flatfile databases.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.distributed"><B>org.biojava.bio.seq.distributed</B></A></TD>
<TD>Sequences and SequenceDBs which are composed from data taken
from a number of data sources.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.filter"><B>org.biojava.bio.seq.filter</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.homol"><B>org.biojava.bio.seq.homol</B></A></TD>
<TD>The classes and interfaces for defining sequence similarity and
honology.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.impl"><B>org.biojava.bio.seq.impl</B></A></TD>
<TD>Standard in-memory implementations of <code>Sequence</code> and
<code>Feature</code>.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io.agave"><B>org.biojava.bio.seq.io.agave</B></A></TD>
<TD>Classes for converting between AGAVE XML and BioJava objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io.filterxml"><B>org.biojava.bio.seq.io.filterxml</B></A></TD>
<TD>Tools for reading and writing an XML representation of BioJava's FeatureFilter language.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io.game"><B>org.biojava.bio.seq.io.game</B></A></TD>
<TD>Event-driven parsing system for the Gene Annotation Markup Elements (GAME).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.projection"><B>org.biojava.bio.seq.projection</B></A></TD>
<TD>Code for projecting Feature objects and systematically altering their
properties.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.structure"><B>org.biojava.bio.structure</B></A></TD>
<TD>
Interfaces and classes for protein structure (PDB).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db"><B>org.biojavax.bio.db</B></A></TD>
<TD>Interactions between biojavax objects and a DB.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.biosql"><B>org.biojavax.bio.db.biosql</B></A></TD>
<TD>Interface between biojava and biosql databases&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq"><B>org.biojavax.bio.seq</B></A></TD>
<TD>Rich implementations of Sequences, Locations and Features.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq.io"><B>org.biojavax.bio.seq.io</B></A></TD>
<TD>Classes to support the I/O of RichSequence and 
     Bioentry objects.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.alignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojava.bio.alignment"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dist"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/dist/package-summary.html">org.biojava.bio.dist</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.dist"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.dp"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceAnnotator.html#org.biojava.bio.dp"><B>SequenceAnnotator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An object which adds some additional information to a Sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.gui.sequence"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/gui/sequence/package-summary.html">org.biojava.bio.gui.sequence</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.gui.sequence"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.gui.sequence"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.gui.sequence"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/OptimizableFilter.html#org.biojava.bio.gui.sequence"><B>OptimizableFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for filters that can potentialy optimize themselves, and
 compare themselves with other filters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.gui.sequence"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.gui.sequence"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.gui.sequence.tracklayout"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/gui/sequence/tracklayout/package-summary.html">org.biojava.bio.gui.sequence.tracklayout</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.gui.sequence.tracklayout"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.molbio"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/molbio/package-summary.html">org.biojava.bio.molbio</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.molbio"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.molbio"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.molbio"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.molbio"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceAnnotator.html#org.biojava.bio.molbio"><B>SequenceAnnotator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An object which adds some additional information to a Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.html#org.biojava.bio.molbio"><B>StrandedFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds the concept of 'strand' to features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Template.html#org.biojava.bio.molbio"><B>StrandedFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for parameterizing the creation of a new
 <code>StrandedFeature</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.das"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.program.das"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.program.das"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.program.das"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.program.das"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.program.das"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojava.bio.program.das"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.program.gff"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.program.gff"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.program.gff"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceAnnotator.html#org.biojava.bio.program.gff"><B>SequenceAnnotator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An object which adds some additional information to a Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.program.gff"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff3"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/program/gff3/package-summary.html">org.biojava.bio.program.gff3</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.program.gff3"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.program.phred"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RealizingFeatureHolder.html#org.biojava.bio.program.phred"><B>RealizingFeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interface for <code>FeatureHolder</code> objects which know how to
 instantiate new child Features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.program.phred"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssbind"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.program.ssbind"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.unigene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.program.unigene"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.xff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/program/xff/package-summary.html">org.biojava.bio.program.xff</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.program.xff"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.program.xff"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.program.xff"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.program.xff"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Template.html#org.biojava.bio.program.xff"><B>StrandedFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for parameterizing the creation of a new
 <code>StrandedFeature</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.proteomics"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/proteomics/package-summary.html">org.biojava.bio.proteomics</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.proteomics"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.search"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.search"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.search"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/AbstractFeatureHolder.html#org.biojava.bio.seq"><B>AbstractFeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.seq"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.ByLocationComparator.html#org.biojava.bio.seq"><B>Feature.ByLocationComparator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>ByLocationComparator</code> compares
 <code>Feature</code>s by the minimum base position of their
 <code>Location</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.seq"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.ByAnnotationType.html#org.biojava.bio.seq"><B>FeatureFilter.ByAnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter that returns all features that have an annotation bundle that is of a given
 annotation type.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureRealizer.html#org.biojava.bio.seq"><B>FeatureRealizer</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interface for translators which map from Feature.Template
 instances to real Feature objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureTypes.Repository.html#org.biojava.bio.seq"><B>FeatureTypes.Repository</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A named collection of Types.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureTypes.Type.html#org.biojava.bio.seq"><B>FeatureTypes.Type</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A type of feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FilterUtils.FilterTransformer.html#org.biojava.bio.seq"><B>FilterUtils.FilterTransformer</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An object able to transform some FeatureFilter instances sytematically into others.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Frame.html#org.biojava.bio.seq"><B>Frame</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Title:        Frame.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FramedFeature.ReadingFrame.html#org.biojava.bio.seq"><B>FramedFeature.ReadingFrame</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A singleton to hold the frame information</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/GappedSequence.html#org.biojava.bio.seq"><B>GappedSequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Extension of GappedSymbolList which also projects features
 into the gapped coordinate system.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/OptimizableFilter.html#org.biojava.bio.seq"><B>OptimizableFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for filters that can potentialy optimize themselves, and
 compare themselves with other filters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RealizingFeatureHolder.html#org.biojava.bio.seq"><B>RealizingFeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interface for <code>FeatureHolder</code> objects which know how to
 instantiate new child Features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RemoteFeature.html#org.biojava.bio.seq"><B>RemoteFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature that indicates that there is some remote feature that can't be
 represented entirely on a single Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RemoteFeature.Resolver.html#org.biojava.bio.seq"><B>RemoteFeature.Resolver</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that actually can take a RemoteFeature and
 return a Sequence object with the feature resolved into a real feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.html#org.biojava.bio.seq"><B>StrandedFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds the concept of 'strand' to features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.seq"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Template.html#org.biojava.bio.seq"><B>StrandedFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for parameterizing the creation of a new
 <code>StrandedFeature</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq.db"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq.db"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.db"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceAnnotator.html#org.biojava.bio.seq.db"><B>SequenceAnnotator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An object which adds some additional information to a Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojava.bio.seq.db"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biofetch"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/db/biofetch/package-summary.html">org.biojava.bio.seq.db.biofetch</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.db.biofetch"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq.db.biosql"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq.db.biosql"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.db.biosql"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojava.bio.seq.db.biosql"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.flat"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/db/flat/package-summary.html">org.biojava.bio.seq.db.flat</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.db.flat"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.distributed"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq.distributed"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq.distributed"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.distributed"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.filter"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/filter/package-summary.html">org.biojava.bio.seq.filter</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq.filter"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.And.html#org.biojava.bio.seq.filter"><B>FeatureFilter.And</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter that returns all features accepted by both child filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.ByAncestor.html#org.biojava.bio.seq.filter"><B>FeatureFilter.ByAncestor</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filter by applying a nested <code>FeatureFilter</code> to all
 ancestor features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.ByChild.html#org.biojava.bio.seq.filter"><B>FeatureFilter.ByChild</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filter by applying a nested <code>FeatureFilter</code> to the
 child features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.ByDescendant.html#org.biojava.bio.seq.filter"><B>FeatureFilter.ByDescendant</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filter by applying a nested <code>FeatureFilter</code> to all
 descendant features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.ByParent.html#org.biojava.bio.seq.filter"><B>FeatureFilter.ByParent</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Filter by applying a nested <code>FeatureFilter</code> to the
 parent feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.Not.html#org.biojava.bio.seq.filter"><B>FeatureFilter.Not</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter that returns all features not accepted by a child filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.OnlyChildren.html#org.biojava.bio.seq.filter"><B>FeatureFilter.OnlyChildren</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Accepts features where all immediate children meet the supplied filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.OnlyDescendants.html#org.biojava.bio.seq.filter"><B>FeatureFilter.OnlyDescendants</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Accepts features where all descendants meet the supplied filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.Or.html#org.biojava.bio.seq.filter"><B>FeatureFilter.Or</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter that returns all features accepted by at least one child filter.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.homol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.seq.homol"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.seq.homol"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq.homol"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.html#org.biojava.bio.seq.homol"><B>StrandedFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds the concept of 'strand' to features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Template.html#org.biojava.bio.seq.homol"><B>StrandedFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for parameterizing the creation of a new
 <code>StrandedFeature</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/ComponentFeature.html#org.biojava.bio.seq.impl"><B>ComponentFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Feature which represents a component in an assembly (contig).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.seq.impl"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.seq.impl"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq.impl"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq.impl"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureRealizer.html#org.biojava.bio.seq.impl"><B>FeatureRealizer</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interface for translators which map from Feature.Template
 instances to real Feature objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FilterUtils.FilterTransformer.html#org.biojava.bio.seq.impl"><B>FilterUtils.FilterTransformer</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An object able to transform some FeatureFilter instances sytematically into others.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Frame.html#org.biojava.bio.seq.impl"><B>Frame</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Title:        Frame.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FramedFeature.html#org.biojava.bio.seq.impl"><B>FramedFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Title:         FramedFeature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FramedFeature.ReadingFrame.html#org.biojava.bio.seq.impl"><B>FramedFeature.ReadingFrame</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A singleton to hold the frame information</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FramedFeature.Template.html#org.biojava.bio.seq.impl"><B>FramedFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/GappedSequence.html#org.biojava.bio.seq.impl"><B>GappedSequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Extension of GappedSymbolList which also projects features
 into the gapped coordinate system.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RealizingFeatureHolder.html#org.biojava.bio.seq.impl"><B>RealizingFeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interface for <code>FeatureHolder</code> objects which know how to
 instantiate new child Features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RemoteFeature.html#org.biojava.bio.seq.impl"><B>RemoteFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature that indicates that there is some remote feature that can't be
 represented entirely on a single Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RemoteFeature.Resolver.html#org.biojava.bio.seq.impl"><B>RemoteFeature.Resolver</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that actually can take a RemoteFeature and
 return a Sequence object with the feature resolved into a real feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/RemoteFeature.Template.html#org.biojava.bio.seq.impl"><B>RemoteFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.impl"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceFactory.html#org.biojava.bio.seq.impl"><B>SequenceFactory</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use org.biojavax.bio.seq.io.RichSequenceBuilder or 
 use org.biojavax.bio.seq.io.SequenceBuilder</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SimpleFeatureHolder.html#org.biojava.bio.seq.impl"><B>SimpleFeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A no-frills implementation of FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.html#org.biojava.bio.seq.impl"><B>StrandedFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds the concept of 'strand' to features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.seq.impl"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Template.html#org.biojava.bio.seq.impl"><B>StrandedFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for parameterizing the creation of a new
 <code>StrandedFeature</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/ComponentFeature.html#org.biojava.bio.seq.io"><B>ComponentFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Feature which represents a component in an assembly (contig).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/ComponentFeature.Template.html#org.biojava.bio.seq.io"><B>ComponentFeature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template for constructing a new ComponentFeature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.seq.io"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.seq.io"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.io"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojava.bio.seq.io"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.seq.io"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io.agave"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/io/agave/package-summary.html">org.biojava.bio.seq.io.agave</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.seq.io.agave"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.seq.io.agave"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq.io.agave"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.io.agave"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.seq.io.agave"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io.filterxml"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/io/filterxml/package-summary.html">org.biojava.bio.seq.io.filterxml</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq.io.filterxml"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io.game"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/io/game/package-summary.html">org.biojava.bio.seq.io.game</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.seq.io.game"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.seq.io.game"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.projection"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/seq/projection/package-summary.html">org.biojava.bio.seq.projection</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/AbstractFeatureHolder.html#org.biojava.bio.seq.projection"><B>AbstractFeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojava.bio.seq.projection"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojava.bio.seq.projection"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojava.bio.seq.projection"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojava.bio.seq.projection"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FilterUtils.FilterTransformer.html#org.biojava.bio.seq.projection"><B>FilterUtils.FilterTransformer</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An object able to transform some FeatureFilter instances sytematically into others.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.seq.projection"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojava.bio.seq.projection"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.structure"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojava/bio/structure/package-summary.html">org.biojava.bio.structure</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojava.bio.structure"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojavax.bio.db"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojavax.bio.db"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojavax.bio.db"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojavax.bio.db"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojavax.bio.db.biosql"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojavax.bio.db.biosql"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojavax.bio.db.biosql"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.html#org.biojavax.bio.seq"><B>Feature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojavax.bio.seq"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureFilter.html#org.biojavax.bio.seq"><B>FeatureFilter</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/FeatureHolder.html#org.biojavax.bio.seq"><B>FeatureHolder</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojavax.bio.seq"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojavax.bio.seq"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.html#org.biojavax.bio.seq"><B>StrandedFeature</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds the concept of 'strand' to features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/StrandedFeature.Strand.html#org.biojavax.bio.seq"><B>StrandedFeature.Strand</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Class to represent the 'strandedness' of a feature.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> used by <A HREF="../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Feature.Template.html#org.biojavax.bio.seq.io"><B>Feature.Template</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/Sequence.html#org.biojavax.bio.seq.io"><B>Sequence</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../org/biojava/bio/seq/class-use/SequenceIterator.html#org.biojavax.bio.seq.io"><B>SequenceIterator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An iterator over a bag of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/seq/package-use.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
