<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:47 BST 2008 -->
<TITLE>
org.biojava.bio.seq (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="org.biojava.bio.seq package">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="org.biojava.bio.seq (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Package</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-use.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/search/package-summary.html"><B>PREV PACKAGE</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/package-summary.html"><B>NEXT PACKAGE</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/seq/package-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<H2>
Package org.biojava.bio.seq
</H2>
Classes and interfaces for defining biological sequences and informatics
objects.
<P>
<B>See:</B>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="#package_description"><B>Description</B></A>
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Interface Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/ComponentFeature.html" title="interface in org.biojava.bio.seq">ComponentFeature</A></B></TD>
<TD>Feature which represents a component in an assembly (contig).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></B></TD>
<TD>A feature within a sequence, or nested within another feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A></B></TD>
<TD>A filter for accepting or rejecting a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></B></TD>
<TD>The interface for objects that contain features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureRealizer.html" title="interface in org.biojava.bio.seq">FeatureRealizer</A></B></TD>
<TD>Interface for translators which map from Feature.Template
 instances to real Feature objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureTypes.Repository.html" title="interface in org.biojava.bio.seq">FeatureTypes.Repository</A></B></TD>
<TD>A named collection of Types.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureTypes.Type.html" title="interface in org.biojava.bio.seq">FeatureTypes.Type</A></B></TD>
<TD>A type of feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FilterUtils.FilterTransformer.html" title="interface in org.biojava.bio.seq">FilterUtils.FilterTransformer</A></B></TD>
<TD>An object able to transform some FeatureFilter instances sytematically into others.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/Frame.html" title="interface in org.biojava.bio.seq">Frame</A></B></TD>
<TD>Title:        Frame.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FramedFeature.html" title="interface in org.biojava.bio.seq">FramedFeature</A></B></TD>
<TD>Title:         FramedFeature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/GappedSequence.html" title="interface in org.biojava.bio.seq">GappedSequence</A></B></TD>
<TD>Extension of GappedSymbolList which also projects features
 into the gapped coordinate system.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/OptimizableFilter.html" title="interface in org.biojava.bio.seq">OptimizableFilter</A></B></TD>
<TD>The interface for filters that can potentialy optimize themselves, and
 compare themselves with other filters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/RealizingFeatureHolder.html" title="interface in org.biojava.bio.seq">RealizingFeatureHolder</A></B></TD>
<TD>Interface for <code>FeatureHolder</code> objects which know how to
 instantiate new child Features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/RemoteFeature.html" title="interface in org.biojava.bio.seq">RemoteFeature</A></B></TD>
<TD>A feature that indicates that there is some remote feature that can't be
 represented entirely on a single Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/RemoteFeature.Resolver.html" title="interface in org.biojava.bio.seq">RemoteFeature.Resolver</A></B></TD>
<TD>The interface for objects that actually can take a RemoteFeature and
 return a Sequence object with the feature resolved into a real feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></B></TD>
<TD>
 A biological sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/SequenceAnnotator.html" title="interface in org.biojava.bio.seq">SequenceAnnotator</A></B></TD>
<TD>An object which adds some additional information to a Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/SequenceFactory.html" title="interface in org.biojava.bio.seq">SequenceFactory</A></B></TD>
<TD><B>Deprecated.</B>&nbsp;<I>use org.biojavax.bio.seq.io.RichSequenceBuilder or 
 use org.biojavax.bio.seq.io.SequenceBuilder</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A></B></TD>
<TD>An iterator over a bag of sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/StrandedFeature.html" title="interface in org.biojava.bio.seq">StrandedFeature</A></B></TD>
<TD>Adds the concept of 'strand' to features.</TD>
</TR>
</TABLE>
&nbsp;

<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Class Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/AbstractFeatureHolder.html" title="class in org.biojava.bio.seq">AbstractFeatureHolder</A></B></TD>
<TD>An abstract implementation of FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/ByLocationMinMaxComparator.html" title="class in org.biojava.bio.seq">ByLocationMinMaxComparator</A></B></TD>
<TD>A Comparator similar to Feature.ByLocationComparator except that the min and max positions of
 the location are both compared</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/ByLocationMinMaxFeatureComparator.html" title="class in org.biojava.bio.seq">ByLocationMinMaxFeatureComparator</A></B></TD>
<TD>Comparator that compares the min and max positions of Features
 
 Required by org.biojava.bio.gui.sequence.AbstractPeptideDigestRenderer instances.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/CircularView.html" title="class in org.biojava.bio.seq">CircularView</A></B></TD>
<TD>
 A circular view onto another Sequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/ComponentFeature.Template.html" title="class in org.biojava.bio.seq">ComponentFeature.Template</A></B></TD>
<TD>Template for constructing a new ComponentFeature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/DNATools.html" title="class in org.biojava.bio.seq">DNATools</A></B></TD>
<TD>Useful functionality for processing DNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/Feature.ByLocationComparator.html" title="class in org.biojava.bio.seq">Feature.ByLocationComparator</A></B></TD>
<TD><code>ByLocationComparator</code> compares
 <code>Feature</code>s by the minimum base position of their
 <code>Location</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A></B></TD>
<TD>Template class for a plain feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.And.html" title="class in org.biojava.bio.seq">FeatureFilter.And</A></B></TD>
<TD>A filter that returns all features accepted by both child filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.AnnotationContains.html" title="class in org.biojava.bio.seq">FeatureFilter.AnnotationContains</A></B></TD>
<TD>Retrieve features that contain a given annotation, and that the set of values
 contains the value given.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByAncestor.html" title="class in org.biojava.bio.seq">FeatureFilter.ByAncestor</A></B></TD>
<TD>Filter by applying a nested <code>FeatureFilter</code> to all
 ancestor features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByAnnotation.html" title="class in org.biojava.bio.seq">FeatureFilter.ByAnnotation</A></B></TD>
<TD>Retrieve features that contain a given annotation with a given value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByAnnotationType.html" title="class in org.biojava.bio.seq">FeatureFilter.ByAnnotationType</A></B></TD>
<TD>A filter that returns all features that have an annotation bundle that is of a given
 annotation type.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByChild.html" title="class in org.biojava.bio.seq">FeatureFilter.ByChild</A></B></TD>
<TD>Filter by applying a nested <code>FeatureFilter</code> to the
 child features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByClass.html" title="class in org.biojava.bio.seq">FeatureFilter.ByClass</A></B></TD>
<TD>Filter which accepts only those filters which are an instance
 of a specific Java class</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByComponentName.html" title="class in org.biojava.bio.seq">FeatureFilter.ByComponentName</A></B></TD>
<TD>Accepts features which are ComponentFeatures and have a <code>componentSequenceName</code>
 property of the specified value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByDescendant.html" title="class in org.biojava.bio.seq">FeatureFilter.ByDescendant</A></B></TD>
<TD>Filter by applying a nested <code>FeatureFilter</code> to all
 descendant features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByFeature.html" title="class in org.biojava.bio.seq">FeatureFilter.ByFeature</A></B></TD>
<TD>Accept only features which are equal to the specified feature</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByPairwiseScore.html" title="class in org.biojava.bio.seq">FeatureFilter.ByPairwiseScore</A></B></TD>
<TD><code>ByPairwiseScore</code> is used to filter
 <code>SimilarityPairFeature</code>s by their score.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByParent.html" title="class in org.biojava.bio.seq">FeatureFilter.ByParent</A></B></TD>
<TD>Filter by applying a nested <code>FeatureFilter</code> to the
 parent feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.BySequenceName.html" title="class in org.biojava.bio.seq">FeatureFilter.BySequenceName</A></B></TD>
<TD>Accept features that reside on a sequence with a particular name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.BySource.html" title="class in org.biojava.bio.seq">FeatureFilter.BySource</A></B></TD>
<TD>Construct one of these to filter features by source.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ByType.html" title="class in org.biojava.bio.seq">FeatureFilter.ByType</A></B></TD>
<TD>Construct one of these to filter features by type.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ContainedByLocation.html" title="class in org.biojava.bio.seq">FeatureFilter.ContainedByLocation</A></B></TD>
<TD>A filter that returns all features contained within a location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.FrameFilter.html" title="class in org.biojava.bio.seq">FeatureFilter.FrameFilter</A></B></TD>
<TD>Accept features with a given reading frame.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.HasAnnotation.html" title="class in org.biojava.bio.seq">FeatureFilter.HasAnnotation</A></B></TD>
<TD>Retrieve features that contain a given annotation with any value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.Not.html" title="class in org.biojava.bio.seq">FeatureFilter.Not</A></B></TD>
<TD>A filter that returns all features not accepted by a child filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.OnlyChildren.html" title="class in org.biojava.bio.seq">FeatureFilter.OnlyChildren</A></B></TD>
<TD>Accepts features where all immediate children meet the supplied filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.OnlyDescendants.html" title="class in org.biojava.bio.seq">FeatureFilter.OnlyDescendants</A></B></TD>
<TD>Accepts features where all descendants meet the supplied filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.Or.html" title="class in org.biojava.bio.seq">FeatureFilter.Or</A></B></TD>
<TD>A filter that returns all features accepted by at least one child filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.OverlapsLocation.html" title="class in org.biojava.bio.seq">FeatureFilter.OverlapsLocation</A></B></TD>
<TD>A filter that returns all features overlapping a location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ShadowContainedByLocation.html" title="class in org.biojava.bio.seq">FeatureFilter.ShadowContainedByLocation</A></B></TD>
<TD>A filter that accepts all features whose shadow is contained by a specified
 <code>Location</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.ShadowOverlapsLocation.html" title="class in org.biojava.bio.seq">FeatureFilter.ShadowOverlapsLocation</A></B></TD>
<TD>A filter that accepts all features whose shadow overlaps a specified
 <code>Location</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureFilter.StrandFilter.html" title="class in org.biojava.bio.seq">FeatureFilter.StrandFilter</A></B></TD>
<TD>Accept features with a given strandedness.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureHolder.EmptyFeatureHolder.html" title="class in org.biojava.bio.seq">FeatureHolder.EmptyFeatureHolder</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureHolderUtils.html" title="class in org.biojava.bio.seq">FeatureHolderUtils</A></B></TD>
<TD>This class intendes to provide some FeatureHolder utilities.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureTypes.html" title="class in org.biojava.bio.seq">FeatureTypes</A></B></TD>
<TD>Registry of known types of features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FeatureTypes.RepositoryImpl.html" title="class in org.biojava.bio.seq">FeatureTypes.RepositoryImpl</A></B></TD>
<TD>A simple implementation of a Repository.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FilterUtils.html" title="class in org.biojava.bio.seq">FilterUtils</A></B></TD>
<TD>A set of FeatureFilter algebraic operations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FramedFeature.ReadingFrame.html" title="class in org.biojava.bio.seq">FramedFeature.ReadingFrame</A></B></TD>
<TD>A singleton to hold the frame information</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/FramedFeature.Template.html" title="class in org.biojava.bio.seq">FramedFeature.Template</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/GeneticCodes.html" title="class in org.biojava.bio.seq">GeneticCodes</A></B></TD>
<TD>Collects the references to translation methods in one place.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/LazyFeatureHolder.html" title="class in org.biojava.bio.seq">LazyFeatureHolder</A></B></TD>
<TD>Wrapper implementation of FeatureHolder which calls a method
 to create a contained FeatureHolder on demand.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/MergeFeatureHolder.html" title="class in org.biojava.bio.seq">MergeFeatureHolder</A></B></TD>
<TD>FeatureHolder which exposes all the features in a set
 of sub-FeatureHolders.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/NewSimpleAssembly.html" title="class in org.biojava.bio.seq">NewSimpleAssembly</A></B></TD>
<TD>A Sequence which is assembled from other sequences contained
 in a set of ComponentFeature objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/NucleotideTools.html" title="class in org.biojava.bio.seq">NucleotideTools</A></B></TD>
<TD>Useful functionality for processing nucleotide sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/ProteinTools.html" title="class in org.biojava.bio.seq">ProteinTools</A></B></TD>
<TD>The central port-of-call for all information and functionality specific to
 SymbolLists over the protein alphabet.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/RemoteFeature.Region.html" title="class in org.biojava.bio.seq">RemoteFeature.Region</A></B></TD>
<TD>A tuple of Location and sequence ID.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/RemoteFeature.Template.html" title="class in org.biojava.bio.seq">RemoteFeature.Template</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/RNATools.html" title="class in org.biojava.bio.seq">RNATools</A></B></TD>
<TD>Useful functionality for processing DNA and RNA sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/SequenceTools.html" title="class in org.biojava.bio.seq">SequenceTools</A></B></TD>
<TD>Methods for manipulating sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/SimpleAssembly.html" title="class in org.biojava.bio.seq">SimpleAssembly</A></B></TD>
<TD>A Sequence which is assembled from other sequences contained
 in a set of ComponentFeature objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/SimpleFeatureHolder.html" title="class in org.biojava.bio.seq">SimpleFeatureHolder</A></B></TD>
<TD>A no-frills implementation of FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/SimpleFeatureRealizer.html" title="class in org.biojava.bio.seq">SimpleFeatureRealizer</A></B></TD>
<TD>FeatureRealizer which uses a lookup table to map template classes
 to implementations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/StrandedFeature.Strand.html" title="class in org.biojava.bio.seq">StrandedFeature.Strand</A></B></TD>
<TD>Class to represent the 'strandedness' of a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/StrandedFeature.Template.html" title="class in org.biojava.bio.seq">StrandedFeature.Template</A></B></TD>
<TD>Template class for parameterizing the creation of a new
 <code>StrandedFeature</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../org/biojava/bio/seq/StrandParser.html" title="class in org.biojava.bio.seq">StrandParser</A></B></TD>
<TD>Process strings and return strand objects.</TD>
</TR>
</TABLE>
&nbsp;

<P>
<A NAME="package_description"><!-- --></A><H2>
Package org.biojava.bio.seq Description
</H2>

<P>
Classes and interfaces for defining biological sequences and informatics
objects.
<P>
This package contains the core classes and interfaces for defining
sequences and features, sequence database and iterators, and miscellaneous
biological information. All definitions are interfaces. Where ever
possible, at least one pure java implementation of the interface is
supplied.
<P>
The Sequence interface inherits the SymbolList interface, making it a list of
Symbol objects. It also implements Annotatable which allows annotation to be
attached to the entire sequence. Lastly, it implements FeatureHolder so that it
can contain features. A Sequence instance represents a concrete piece of
biological data, rather than a comp-sci abstraction.
<P>
Features are regions of a sequence which have location-specific
annotation or meaning. The location of a feature is a Location object. Features
can contain sub-features, as they are FeatureHolder implementors. They are
also annotatable. Lastly, they contain a couple of fields aimed at enhancing GFF
interoperability.
<P>
Also within this package are the classes that implement these
interfaces. They are usually preceeded with 'Simple' to indicate
that they are vanilla, pure-java implementations. You should
be able to use these objects out-of-the-box.
<P>
The DNATools and ProteinTools classes provide a central point-of-access to the
DNA and Protein alphabets and add on many operators specific to these domains.
For example, DNATools has a method 'complement' that will complement a DNA
SymbolList.
<P>

<P>
<DL>
</DL>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Package</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-use.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/biojava/bio/search/package-summary.html"><B>PREV PACKAGE</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/package-summary.html"><B>NEXT PACKAGE</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio/seq/package-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
