<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:55 BST 2008 -->
<TITLE>
Uses of Interface org.biojava.bio.seq.db.SequenceDB (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Interface org.biojava.bio.seq.db.SequenceDB (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?org/biojava/bio/seq/db//class-useSequenceDB.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SequenceDB.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Interface<br>org.biojava.bio.seq.db.SequenceDB</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.alignment"><B>org.biojava.bio.alignment</B></A></TD>
<TD>
Classes to generate and describe sequence alignments.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp"><B>org.biojava.bio.dp</B></A></TD>
<TD>HMM and Dynamic Programming Algorithms.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.gui"><B>org.biojava.bio.gui</B></A></TD>
<TD>Graphical interfaces for biojava objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.das"><B>org.biojava.bio.program.das</B></A></TD>
<TD>Development client for the Distributed Annotation System.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff"><B>org.biojava.bio.program.gff</B></A></TD>
<TD>GFF manipulation.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssaha"><B>org.biojava.bio.program.ssaha</B></A></TD>
<TD>SSAHA sequence searching API.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssbind"><B>org.biojava.bio.program.ssbind</B></A></TD>
<TD>
Creation of objects from SAX events using the BioJava BlastLikeDataSetCollection DTD.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.unigene"><B>org.biojava.bio.program.unigene</B></A></TD>
<TD>Objects for representing Unigene clusters.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.search"><B>org.biojava.bio.search</B></A></TD>
<TD>
Interfaces and classes for representing sequence similarity search results.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db"><B>org.biojava.bio.seq.db</B></A></TD>
<TD>
Collections of biological sequence data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biosql"><B>org.biojava.bio.seq.db.biosql</B></A></TD>
<TD>General purpose Sequence storage in a relational database.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.distributed"><B>org.biojava.bio.seq.distributed</B></A></TD>
<TD>Sequences and SequenceDBs which are composed from data taken
from a number of data sources.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.impl"><B>org.biojava.bio.seq.impl</B></A></TD>
<TD>Standard in-memory implementations of <code>Sequence</code> and
<code>Feature</code>.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db"><B>org.biojavax.bio.db</B></A></TD>
<TD>Interactions between biojavax objects and a DB.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.biosql"><B>org.biojavax.bio.db.biosql</B></A></TD>
<TD>Interface between biojava and biosql databases&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.ncbi"><B>org.biojavax.bio.db.ncbi</B></A></TD>
<TD>Interfaces to NCBI data.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.alignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>abstract &nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B>SequenceAlignment.</B><B><A HREF="../../../../../../org/biojava/bio/alignment/SequenceAlignment.html#alignAll(org.biojava.bio.seq.SequenceIterator, org.biojava.bio.seq.db.SequenceDB)">alignAll</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;source,
         <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;subjectDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B>NeedlemanWunsch.</B><B><A HREF="../../../../../../org/biojava/bio/alignment/NeedlemanWunsch.html#alignAll(org.biojava.bio.seq.SequenceIterator, org.biojava.bio.seq.db.SequenceDB)">alignAll</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;source,
         <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;subjectDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TrainingAlgorithm.</B><B><A HREF="../../../../../../org/biojava/bio/dp/TrainingAlgorithm.html#train(org.biojava.bio.seq.db.SequenceDB, double, org.biojava.bio.dp.StoppingCriteria)">train</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db,
      double&nbsp;nullWeight,
      <A HREF="../../../../../../org/biojava/bio/dp/StoppingCriteria.html" title="interface in org.biojava.bio.dp">StoppingCriteria</A>&nbsp;stopper)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trains the sequences in db untill stopper says to finnish.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractTrainer.</B><B><A HREF="../../../../../../org/biojava/bio/dp/AbstractTrainer.html#train(org.biojava.bio.seq.db.SequenceDB, double, org.biojava.bio.dp.StoppingCriteria)">train</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db,
      double&nbsp;nullModelWeight,
      <A HREF="../../../../../../org/biojava/bio/dp/StoppingCriteria.html" title="interface in org.biojava.bio.dp">StoppingCriteria</A>&nbsp;stopper)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trains the sequences in db until stopper says to finnish.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.gui"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/gui/package-summary.html">org.biojava.bio.gui</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/gui/package-summary.html">org.biojava.bio.gui</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureTree.</B><B><A HREF="../../../../../../org/biojava/bio/gui/FeatureTree.html#setSequenceDB(org.biojava.bio.seq.db.SequenceDB)">setSequenceDB</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Use this method to provide the sequences for the tree to work with.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.das"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/program/das/DASSequenceDB.html" title="class in org.biojava.bio.program.das">DASSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Collection of sequences retrieved from the DAS network.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>ReferenceServer.</B><B><A HREF="../../../../../../org/biojava/bio/program/das/ReferenceServer.html#getDB()">getDB</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../../../org/biojava/bio/program/gff/GFFTools.html#annotateSequences(org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.program.gff.GFFEntrySet)">annotateSequences</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqs,
                  <A HREF="../../../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A>&nbsp;ents)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Annotates all sequences in a sequence DB with features from a GFF entry set.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../../../org/biojava/bio/program/gff/GFFTools.html#annotateSequences(org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.program.gff.GFFEntrySet)">annotateSequences</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqs,
                  <A HREF="../../../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A>&nbsp;ents)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Annotates all sequences in a sequence DB with features from a GFF entry set.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../../../org/biojava/bio/program/gff/GFFTools.html#gffFromSeqDB(org.biojava.bio.seq.db.SequenceDB)">gffFromSeqDB</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a GFFEntrySet containing one entry for each feature on each
 sequence of a SequenceDB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequencesAsGFF.</B><B><A HREF="../../../../../../org/biojava/bio/program/gff/SequencesAsGFF.html#processDB(org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.program.gff.GFFDocumentHandler)">processDB</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
          <A HREF="../../../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Process all <span class="type">Sequence</span>s within a
 <span class="type">SequenceDB</span>, informing
 <span class="arg">handler</span> of any suitable features.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PhredTools.</B><B><A HREF="../../../../../../org/biojava/bio/program/phred/PhredTools.html#writePhredQuality(java.io.OutputStream, java.io.OutputStream, org.biojava.bio.seq.db.SequenceDB)">writePhredQuality</A></B>(java.io.OutputStream&nbsp;qual,
                  java.io.OutputStream&nbsp;seq,
                  <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes Phred quality data in a Fasta type format.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssaha"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>DataStoreFactory.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssaha/DataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Build a new DataStore.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>CompactedDataStoreFactory.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssaha/CompactedDataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>MappedDataStoreFactory.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssaha/MappedDataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>NIODataStoreFactory.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssaha/NIODataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/program/ssaha/SequenceStreamer.SequenceDBStreamer.html#SequenceStreamer.SequenceDBStreamer(org.biojava.bio.seq.db.SequenceDB)">SequenceStreamer.SequenceDBStreamer</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssbind"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Fields in <A HREF="../../../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A> declared as <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>ViewSequenceFactory.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssbind/ViewSequenceFactory.html#querySeqHolder">querySeqHolder</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>ViewSequenceFactory.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssbind/ViewSequenceFactory.html#getQuerySeqHolder()">getQuerySeqHolder</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getQuerySeqHolder</code> returns the database of query
 sequences used to retrieve sequences for creation of the
 various result objects.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ViewSequenceFactory.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssbind/ViewSequenceFactory.html#setQuerySeqHolder(org.biojava.bio.seq.db.SequenceDB)">setQuerySeqHolder</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;querySeqHolder)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>setQuerySeqHolder</code> sets the query sequence holder
 to a specific database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BlastLikeSearchBuilder.</B><B><A HREF="../../../../../../org/biojava/bio/program/ssbind/BlastLikeSearchBuilder.html#setQuerySeqHolder(org.biojava.bio.seq.db.SequenceDB)">setQuerySeqHolder</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;querySeqHolder)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>setQuerySeqHolder</code> sets the query sequence holder
 to a specific database.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/program/ssbind/BlastLikeSearchBuilder.html#BlastLikeSearchBuilder(java.util.List, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.seq.db.SequenceDBInstallation)">BlastLikeSearchBuilder</A></B>(java.util.List&nbsp;target,
                       <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;querySeqHolder,
                       <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDBInstallation.html" title="interface in org.biojava.bio.seq.db">SequenceDBInstallation</A>&nbsp;subjectDBs)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new <code>BlastLikeSearchBuilder</code> which will
 instantiate results into the <code>List</code> target.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.unigene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneCluster.</B><B><A HREF="../../../../../../org/biojava/bio/program/unigene/UnigeneCluster.html#getAll()">getAll</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;All sequences that map to this cluster.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.search"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>SimpleSeqSimilaritySearchResult.</B><B><A HREF="../../../../../../org/biojava/bio/search/SimpleSeqSimilaritySearchResult.html#getSequenceDB()">getSequenceDB</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>SeqSimilaritySearchResult.</B><B><A HREF="../../../../../../org/biojava/bio/search/SeqSimilaritySearchResult.html#getSequenceDB()">getSequenceDB</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the sequence database against which the search was
 performed.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBSearchResult.</B><B><A HREF="../../../../../../org/biojava/bio/search/SequenceDBSearchResult.html#getSequenceDB()">getSequenceDB</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/search/SeqSimilaritySearchResult.html" title="interface in org.biojava.bio.search">SeqSimilaritySearchResult</A></CODE></FONT></TD>
<TD><CODE><B>SeqSimilaritySearcher.</B><B><A HREF="../../../../../../org/biojava/bio/search/SeqSimilaritySearcher.html#search(org.biojava.bio.symbol.SymbolList, org.biojava.bio.seq.db.SequenceDB, java.util.Map)">search</A></B>(<A HREF="../../../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;querySeq,
       <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db,
       java.util.Map&nbsp;searchParameters)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Using this sequence similarity searcher, search with the given
 sequence against the given sequence database.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/search/SequenceDBSearchResult.html#SequenceDBSearchResult(org.biojava.bio.seq.Sequence, org.biojava.bio.seq.db.SequenceDB, java.util.Map, java.util.List, org.biojava.bio.Annotation)">SequenceDBSearchResult</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;querySequence,
                       <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;sequenceDB,
                       java.util.Map&nbsp;searchParameters,
                       java.util.List&nbsp;hits,
                       <A HREF="../../../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A>&nbsp;annotation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Creates a new <code>SequenceDBSearchResult</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/search/SimpleSeqSimilaritySearchResult.html#SimpleSeqSimilaritySearchResult(org.biojava.bio.seq.Sequence, org.biojava.bio.seq.db.SequenceDB, java.util.Map, java.util.List, org.biojava.bio.Annotation)">SimpleSeqSimilaritySearchResult</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;querySequence,
                                <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;sequenceDB,
                                java.util.Map&nbsp;searchParameters,
                                java.util.List&nbsp;hits,
                                <A HREF="../../../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A>&nbsp;annotation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new <code>SimpleSeqSimilaritySearchResult</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/AbstractSequenceDB.html" title="class in org.biojava.bio.seq.db">AbstractSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of SequenceDB that provides the sequenceIterator
 method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/AnnotatedSequenceDB.html" title="class in org.biojava.bio.seq.db">AnnotatedSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SequenceDB implementation which lazily applies a SequenceAnnotator
 to sequences retrieved from a SequenceDB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/CachingSequenceDB.html" title="class in org.biojava.bio.seq.db">CachingSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SequenceDB implementation that caches the results of another SequenceDB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/DummySequenceDB.html" title="class in org.biojava.bio.seq.db">DummySequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>DummySequenceDB</code> is an implementation which contains
 only a <code>DummySequence</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/HashSequenceDB.html" title="class in org.biojava.bio.seq.db">HashSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An implementation of SequenceDB that uses an underlying HashMap to store the
 sequence objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/IndexedSequenceDB.html" title="class in org.biojava.bio.seq.db">IndexedSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 This class implements SequenceDB on top of a set of sequence files
 and sequence offsets within these files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDBWrapper.html" title="class in org.biojava.bio.seq.db">SequenceDBWrapper</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of SequenceDB that wraps up another database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/SubSequenceDB.html" title="class in org.biojava.bio.seq.db">SubSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/ViewingSequenceDB.html" title="class in org.biojava.bio.seq.db">ViewingSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SequenceDB implementation that returns new SequenceView instances
 wrapping the sequences in an underlying database.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBWrapper.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDBWrapper.html#getParent()">getParent</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the parent SequenceDB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>AnnotatedSequenceDB.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/AnnotatedSequenceDB.html#getParent()">getParent</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the original sequenceDB from this annotated sequenceDB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenbankSequenceDB.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/GenbankSequenceDB.html#getSequences(java.util.Set)">getSequences</A></B>(java.util.Set&nbsp;list)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve sequences from a Genbank</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenbankSequenceDB.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/GenbankSequenceDB.html#getSequences(java.util.Set, org.biojava.bio.seq.db.SequenceDB)">getSequences</A></B>(java.util.Set&nbsp;list,
             <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;database)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve sequences from a Genbank</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenbankSequenceDB.</B><B><A HREF="../../../../../../org/biojava/bio/seq/db/GenbankSequenceDB.html#getSequences(java.util.Set, org.biojava.bio.seq.db.SequenceDB)">getSequences</A></B>(java.util.Set&nbsp;list,
             <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;database)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve sequences from a Genbank</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/AnnotatedSequenceDB.html#AnnotatedSequenceDB(org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.seq.SequenceAnnotator)">AnnotatedSequenceDB</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;parent,
                    <A HREF="../../../../../../org/biojava/bio/seq/SequenceAnnotator.html" title="interface in org.biojava.bio.seq">SequenceAnnotator</A>&nbsp;a)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/CachingSequenceDB.html#CachingSequenceDB(org.biojava.bio.seq.db.SequenceDB)">CachingSequenceDB</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;parent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new CachingSequenceDB that caches the sequences in parent.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDBWrapper.html#SequenceDBWrapper(org.biojava.bio.seq.db.SequenceDB)">SequenceDBWrapper</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;parent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/SubSequenceDB.html#SubSequenceDB(org.biojava.bio.seq.db.SequenceDB, java.util.Set)">SubSequenceDB</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;parent,
              java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/ViewingSequenceDB.html#ViewingSequenceDB(org.biojava.bio.seq.db.SequenceDB)">ViewingSequenceDB</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;parent)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new ViewingSequenceDB that views the sequences in parent.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html" title="class in org.biojava.bio.seq.db.biosql">BioSQLSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I></TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.distributed"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/distributed/DistributedSequenceDB.html" title="class in org.biojava.bio.seq.distributed">DistributedSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sequence database from the meta-DAS system.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/distributed/SequenceDBDataSource.html#SequenceDBDataSource(org.biojava.bio.seq.db.SequenceDB)">SequenceDBDataSource</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>SimpleRemoteFeature.DBResolver.</B><B><A HREF="../../../../../../org/biojava/bio/seq/impl/SimpleRemoteFeature.DBResolver.html#getSeqDB()">getSeqDB</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/impl/SimpleRemoteFeature.DBResolver.html#SimpleRemoteFeature.DBResolver(org.biojava.bio.seq.db.SequenceDB)">SimpleRemoteFeature.DBResolver</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> that return <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFasta(java.io.InputStream, org.biojava.bio.symbol.Alphabet)">readFasta</A></B>(java.io.InputStream&nbsp;seqFile,
          <A HREF="../../../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Create a sequence database from a fasta file provided as an
 input stream.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeFasta(java.io.OutputStream, org.biojava.bio.seq.db.SequenceDB)">writeFasta</A></B>(java.io.OutputStream&nbsp;os,
           <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Write a sequenceDB to an output stream in fasta format.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> with parameters of type <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../org/biojava/bio/seq/io/SequenceDBSequenceBuilder.html#SequenceDBSequenceBuilder(org.biojava.bio.seq.db.SequenceDB, int)">SequenceDBSequenceBuilder</A></B>(<A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db,
                          int&nbsp;mode)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;constructor</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subinterfaces of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;interface</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of RichSequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html" title="class in org.biojavax.bio.db">AbstractRichSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of RichSequenceDB that provides the getRichSequenceIterator
 method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/db/HashRichSequenceDB.html" title="class in org.biojavax.bio.db">HashRichSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An implementation of RichSequenceDB that uses an underlying HashMap to store the
 RichSequence objects.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/db/biosql/BioSQLRichSequenceDB.html" title="class in org.biojavax.bio.db.biosql">BioSQLRichSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.ncbi"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A> in <A HREF="../../../../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A> that implement <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html" title="class in org.biojavax.bio.db.ncbi">GenbankRichSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This class contains functions accessing DNA sequences in Genbank format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html" title="class in org.biojavax.bio.db.ncbi">GenpeptRichSequenceDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This class contains functions accessing Peptide sequences in Genpept format.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?org/biojava/bio/seq/db//class-useSequenceDB.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="SequenceDB.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
