<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:51 BST 2008 -->
<TITLE>
Uses of Class org.biojava.bio.BioException (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Class org.biojava.bio.BioException (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio//class-useBioException.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="BioException.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Class<br>org.biojava.bio.BioException</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.alignment"><B>org.biojava.bio.alignment</B></A></TD>
<TD>
Classes to generate and describe sequence alignments.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.annodb"><B>org.biojava.bio.annodb</B></A></TD>
<TD>
Databases of generic structured data (annotation) objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.chromatogram"><B>org.biojava.bio.chromatogram</B></A></TD>
<TD>
Interfaces and classes for chromatogram data, as produced by DNA sequencing equipment.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dist"><B>org.biojava.bio.dist</B></A></TD>
<TD>
Probability distributions over Alphabets.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp"><B>org.biojava.bio.dp</B></A></TD>
<TD>HMM and Dynamic Programming Algorithms.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp.onehead"><B>org.biojava.bio.dp.onehead</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp.twohead"><B>org.biojava.bio.dp.twohead</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.gui"><B>org.biojava.bio.gui</B></A></TD>
<TD>Graphical interfaces for biojava objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.molbio"><B>org.biojava.bio.molbio</B></A></TD>
<TD>The classes and interfaces in this package cover common molecular
biological techniques such as restriction digests and PCR.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.das"><B>org.biojava.bio.program.das</B></A></TD>
<TD>Development client for the Distributed Annotation System.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.das.dasalignment"><B>org.biojava.bio.program.das.dasalignment</B></A></TD>
<TD>
request alignments via DAS&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.formats"><B>org.biojava.bio.program.formats</B></A></TD>
<TD>
Experimental parsers using the tagvalue framework.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff"><B>org.biojava.bio.program.gff</B></A></TD>
<TD>GFF manipulation.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff3"><B>org.biojava.bio.program.gff3</B></A></TD>
<TD>
Support for reading and writing GFF3.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.homologene"><B>org.biojava.bio.program.homologene</B></A></TD>
<TD>
Support classes for Homologene data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.indexdb"><B>org.biojava.bio.program.indexdb</B></A></TD>
<TD>A flat-file ascii index of ascii flat files as per the OBDA specification.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.sax.blastxml"><B>org.biojava.bio.program.sax.blastxml</B></A></TD>
<TD>
Parsers for the XML output from NCBI blast.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssaha"><B>org.biojava.bio.program.ssaha</B></A></TD>
<TD>SSAHA sequence searching API.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssbind"><B>org.biojava.bio.program.ssbind</B></A></TD>
<TD>
Creation of objects from SAX events using the BioJava BlastLikeDataSetCollection DTD.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.unigene"><B>org.biojava.bio.program.unigene</B></A></TD>
<TD>Objects for representing Unigene clusters.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.xff"><B>org.biojava.bio.program.xff</B></A></TD>
<TD>Event-driven parsing system for the Extensible Feature Format (XFF).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.proteomics"><B>org.biojava.bio.proteomics</B></A></TD>
<TD>Utilities to aid in performing various physical analysis of proteins.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.proteomics.aaindex"><B>org.biojava.bio.proteomics.aaindex</B></A></TD>
<TD>Classes and interfaces to load Amino Acid Index database files.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.search"><B>org.biojava.bio.search</B></A></TD>
<TD>
Interfaces and classes for representing sequence similarity search results.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq"><B>org.biojava.bio.seq</B></A></TD>
<TD>Classes and interfaces for defining biological sequences and informatics
objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db"><B>org.biojava.bio.seq.db</B></A></TD>
<TD>
Collections of biological sequence data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biofetch"><B>org.biojava.bio.seq.db.biofetch</B></A></TD>
<TD>Client for the OBDA BioFetch protocol.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biosql"><B>org.biojava.bio.seq.db.biosql</B></A></TD>
<TD>General purpose Sequence storage in a relational database.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.flat"><B>org.biojava.bio.seq.db.flat</B></A></TD>
<TD>
Support for OBDA flatfile databases.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.distributed"><B>org.biojava.bio.seq.distributed</B></A></TD>
<TD>Sequences and SequenceDBs which are composed from data taken
from a number of data sources.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.homol"><B>org.biojava.bio.seq.homol</B></A></TD>
<TD>The classes and interfaces for defining sequence similarity and
honology.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.impl"><B>org.biojava.bio.seq.impl</B></A></TD>
<TD>Standard in-memory implementations of <code>Sequence</code> and
<code>Feature</code>.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.projection"><B>org.biojava.bio.seq.projection</B></A></TD>
<TD>Code for projecting Feature objects and systematically altering their
properties.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.structure"><B>org.biojava.bio.structure</B></A></TD>
<TD>
Interfaces and classes for protein structure (PDB).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.structure.io"><B>org.biojava.bio.structure.io</B></A></TD>
<TD>Input and Output of Structures&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.symbol"><B>org.biojava.bio.symbol</B></A></TD>
<TD>Representation of the Symbols that make up a sequence, and locations within
them.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.taxa"><B>org.biojava.bio.taxa</B></A></TD>
<TD>Taxonomy object for representing species information.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.directory"><B>org.biojava.directory</B></A></TD>
<TD>Open Bio Sequence Database Access (OBDA) registry support.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.automata"><B>org.biojava.utils.automata</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.math"><B>org.biojava.utils.math</B></A></TD>
<TD>
Mathematical utility classes.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.regex"><B>org.biojava.utils.regex</B></A></TD>
<TD>This package is used to perform regular expression searches of
SymbolLists defined in arbitrary Alphabets.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.walker"><B>org.biojava.utils.walker</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio"><B>org.biojavax.bio</B></A></TD>
<TD>Classes to represent biological entities and their relationships.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db"><B>org.biojavax.bio.db</B></A></TD>
<TD>Interactions between biojavax objects and a DB.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.biosql"><B>org.biojavax.bio.db.biosql</B></A></TD>
<TD>Interface between biojava and biosql databases&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.ncbi"><B>org.biojavax.bio.db.ncbi</B></A></TD>
<TD>Interfaces to NCBI data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq"><B>org.biojavax.bio.seq</B></A></TD>
<TD>Rich implementations of Sequences, Locations and Features.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq.io"><B>org.biojavax.bio.seq.io</B></A></TD>
<TD>Classes to support the I/O of RichSequence and 
     Bioentry objects.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.alignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FlexibleAlignment.</B><B><A HREF="../../../../org/biojava/bio/alignment/FlexibleAlignment.html#addSequence(org.biojava.bio.alignment.AlignmentElement)">addSequence</A></B>(<A HREF="../../../../org/biojava/bio/alignment/AlignmentElement.html" title="interface in org.biojava.bio.alignment">AlignmentElement</A>&nbsp;ae)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;add a new a alignment usings a location to the reference sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ARAlignment.</B><B><A HREF="../../../../org/biojava/bio/alignment/ARAlignment.html#addSequence(org.biojava.bio.alignment.AlignmentElement)">addSequence</A></B>(<A HREF="../../../../org/biojava/bio/alignment/AlignmentElement.html" title="interface in org.biojava.bio.alignment">AlignmentElement</A>&nbsp;ae)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B>NeedlemanWunsch.</B><B><A HREF="../../../../org/biojava/bio/alignment/NeedlemanWunsch.html#alignAll(org.biojava.bio.seq.SequenceIterator, org.biojava.bio.seq.db.SequenceDB)">alignAll</A></B>(<A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;source,
         <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;subjectDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B>NeedlemanWunsch.</B><B><A HREF="../../../../org/biojava/bio/alignment/NeedlemanWunsch.html#getAlignmentString()">getAlignmentString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>SubstitutionMatrix.</B><B><A HREF="../../../../org/biojava/bio/alignment/SubstitutionMatrix.html#getValueAt(org.biojava.bio.symbol.Symbol, org.biojava.bio.symbol.Symbol)">getValueAt</A></B>(<A HREF="../../../../org/biojava/bio/symbol/Symbol.html" title="interface in org.biojava.bio.symbol">Symbol</A>&nbsp;row,
           <A HREF="../../../../org/biojava/bio/symbol/Symbol.html" title="interface in org.biojava.bio.symbol">Symbol</A>&nbsp;col)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;There are some substitution matrices containing more columns than lines.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;int[][]</CODE></FONT></TD>
<TD><CODE><B>SubstitutionMatrix.</B><B><A HREF="../../../../org/biojava/bio/alignment/SubstitutionMatrix.html#parseMatrix(java.lang.String)">parseMatrix</A></B>(java.lang.String&nbsp;matString)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a String representing the contents of a substitution matrix file.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/alignment/FlexibleAlignment.html#FlexibleAlignment(java.util.List)">FlexibleAlignment</A></B>(java.util.List&nbsp;seqList)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;construct this object with the reference sequence which can either be a gappedSymbolList or not
 label in all cases refers to an object that holds the display name (generally just a String).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/alignment/SimpleAlignmentElement.html#SimpleAlignmentElement(java.lang.Object, org.biojava.bio.symbol.SymbolList, org.biojava.bio.symbol.Location)">SimpleAlignmentElement</A></B>(java.lang.Object&nbsp;label,
                       <A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;seq,
                       <A HREF="../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/alignment/SubstitutionMatrix.html#SubstitutionMatrix(org.biojava.bio.symbol.FiniteAlphabet, java.io.File)">SubstitutionMatrix</A></B>(<A HREF="../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A>&nbsp;alpha,
                   java.io.File&nbsp;matrixFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This constructs a <code>SubstitutionMatrix</code>-object that contains
 two <code>Map</code> data structures having biojava symbols as key and
 the value being the index of the matrix containing the substitution score.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/alignment/SubstitutionMatrix.html#SubstitutionMatrix(org.biojava.bio.symbol.FiniteAlphabet, java.lang.String, java.lang.String)">SubstitutionMatrix</A></B>(<A HREF="../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A>&nbsp;alpha,
                   java.lang.String&nbsp;matrixString,
                   java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;With this constructor it is possible to construct a SubstitutionMatrix
 object from a substitution matrix file.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.annodb"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/annodb/package-summary.html">org.biojava.bio.annodb</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/annodb/package-summary.html">org.biojava.bio.annodb</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/annodb/IndexedAnnotationDB.html#IndexedAnnotationDB(java.lang.String, java.io.File, org.biojava.bio.program.tagvalue.Index2Model, java.util.List, int, org.biojava.bio.AnnotationType, org.biojava.bio.annodb.IndexedAnnotationDB.ParserListenerFactory)">IndexedAnnotationDB</A></B>(java.lang.String&nbsp;dbName,
                    java.io.File&nbsp;storeLoc,
                    <A HREF="../../../../org/biojava/bio/program/tagvalue/Index2Model.html" title="class in org.biojava.bio.program.tagvalue">Index2Model</A>&nbsp;model,
                    java.util.List&nbsp;toIndex,
                    int&nbsp;maxKeyLen,
                    <A HREF="../../../../org/biojava/bio/AnnotationType.html" title="interface in org.biojava.bio">AnnotationType</A>&nbsp;schema,
                    <A HREF="../../../../org/biojava/bio/annodb/IndexedAnnotationDB.ParserListenerFactory.html" title="interface in org.biojava.bio.annodb">IndexedAnnotationDB.ParserListenerFactory</A>&nbsp;plFactory)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new IndexedAnnotationDB.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.chromatogram"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/chromatogram/package-summary.html">org.biojava.bio.chromatogram</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/chromatogram/package-summary.html">org.biojava.bio.chromatogram</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/chromatogram/UnsupportedChromatogramFormatException.html" title="class in org.biojava.bio.chromatogram">UnsupportedChromatogramFormatException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Exception indicating that some provided data does not represent
 a supported chromatogram format.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dist"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/dist/package-summary.html">org.biojava.bio.dist</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/dist/package-summary.html">org.biojava.bio.dist</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>DistributionTools.</B><B><A HREF="../../../../org/biojava/bio/dist/DistributionTools.html#areEmissionSpectraEqual(org.biojava.bio.dist.Distribution[], org.biojava.bio.dist.Distribution[])">areEmissionSpectraEqual</A></B>(<A HREF="../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>[]&nbsp;a,
                        <A HREF="../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>[]&nbsp;b)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Compares the emission spectra of two distribution arrays.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>DistributionTools.</B><B><A HREF="../../../../org/biojava/bio/dist/DistributionTools.html#areEmissionSpectraEqual(org.biojava.bio.dist.Distribution, org.biojava.bio.dist.Distribution)">areEmissionSpectraEqual</A></B>(<A HREF="../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>&nbsp;a,
                        <A HREF="../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>&nbsp;b)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Compares the emission spectra of two distributions.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/dp/IllegalTransitionException.html" title="class in org.biojava.bio.dp">IllegalTransitionException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This exception indicates that there is no transition between two states.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>WeightMatrixAnnotator.</B><B><A HREF="../../../../org/biojava/bio/dp/WeightMatrixAnnotator.html#annotate(org.biojava.bio.seq.Sequence)">annotate</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleHMMTrainer.</B><B><A HREF="../../../../org/biojava/bio/dp/SimpleHMMTrainer.html#completeCycle()">completeCycle</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HMMTrainer.</B><B><A HREF="../../../../org/biojava/bio/dp/HMMTrainer.html#completeCycle()">completeCycle</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;indicate that a cycle of training is completed and the
 emission/transition matrices should be updated.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/dp/DP.html" title="class in org.biojava.bio.dp">DP</A></CODE></FONT></TD>
<TD><CODE><B>DPFactory.</B><B><A HREF="../../../../org/biojava/bio/dp/DPFactory.html#createDP(org.biojava.bio.dp.MarkovModel)">createDP</A></B>(<A HREF="../../../../org/biojava/bio/dp/MarkovModel.html" title="interface in org.biojava.bio.dp">MarkovModel</A>&nbsp;model)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/dp/DP.html" title="class in org.biojava.bio.dp">DP</A></CODE></FONT></TD>
<TD><CODE><B>DPFactory.DefaultFactory.</B><B><A HREF="../../../../org/biojava/bio/dp/DPFactory.DefaultFactory.html#createDP(org.biojava.bio.dp.MarkovModel)">createDP</A></B>(<A HREF="../../../../org/biojava/bio/dp/MarkovModel.html" title="interface in org.biojava.bio.dp">MarkovModel</A>&nbsp;model)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/dp/DPMatrix.html" title="interface in org.biojava.bio.dp">DPMatrix</A></CODE></FONT></TD>
<TD><CODE><B>DP.</B><B><A HREF="../../../../org/biojava/bio/dp/DP.html#forwardsBackwards(org.biojava.bio.symbol.SymbolList[], org.biojava.bio.dp.ScoreType)">forwardsBackwards</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>[]&nbsp;symList,
                  <A HREF="../../../../org/biojava/bio/dp/ScoreType.html" title="interface in org.biojava.bio.dp">ScoreType</A>&nbsp;scoreType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/dp/StatePath.html" title="interface in org.biojava.bio.dp">StatePath</A></CODE></FONT></TD>
<TD><CODE><B>DP.</B><B><A HREF="../../../../org/biojava/bio/dp/DP.html#generate(int)">generate</A></B>(int&nbsp;length)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Generates an alignment from a model.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/dp/WeightMatrix.html" title="interface in org.biojava.bio.dp">WeightMatrix</A></CODE></FONT></TD>
<TD><CODE><B>XmlMarkovModel.</B><B><A HREF="../../../../org/biojava/bio/dp/XmlMarkovModel.html#readMatrix(org.w3c.dom.Element)">readMatrix</A></B>(org.w3c.dom.Element&nbsp;root)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/dp/MarkovModel.html" title="interface in org.biojava.bio.dp">MarkovModel</A></CODE></FONT></TD>
<TD><CODE><B>XmlMarkovModel.</B><B><A HREF="../../../../org/biojava/bio/dp/XmlMarkovModel.html#readModel(org.w3c.dom.Element)">readModel</A></B>(org.w3c.dom.Element&nbsp;root)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>WMAsMM.</B><B><A HREF="../../../../org/biojava/bio/dp/WMAsMM.html#registerWithTrainer(org.biojava.bio.dp.ModelTrainer)">registerWithTrainer</A></B>(<A HREF="../../../../org/biojava/bio/dp/ModelTrainer.html" title="interface in org.biojava.bio.dp">ModelTrainer</A>&nbsp;modelTrainer)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>Trainable.</B><B><A HREF="../../../../org/biojava/bio/dp/Trainable.html#registerWithTrainer(org.biojava.bio.dp.ModelTrainer)">registerWithTrainer</A></B>(<A HREF="../../../../org/biojava/bio/dp/ModelTrainer.html" title="interface in org.biojava.bio.dp">ModelTrainer</A>&nbsp;mt)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Perform any registration that is necessary with mt.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/dp/State.html" title="interface in org.biojava.bio.dp">State</A>[]</CODE></FONT></TD>
<TD><CODE><B>DP.</B><B><A HREF="../../../../org/biojava/bio/dp/DP.html#stateList(org.biojava.bio.dp.MarkovModel)">stateList</A></B>(<A HREF="../../../../org/biojava/bio/dp/MarkovModel.html" title="interface in org.biojava.bio.dp">MarkovModel</A>&nbsp;mm)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TrainingAlgorithm.</B><B><A HREF="../../../../org/biojava/bio/dp/TrainingAlgorithm.html#train(org.biojava.bio.seq.db.SequenceDB, double, org.biojava.bio.dp.StoppingCriteria)">train</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db,
      double&nbsp;nullWeight,
      <A HREF="../../../../org/biojava/bio/dp/StoppingCriteria.html" title="interface in org.biojava.bio.dp">StoppingCriteria</A>&nbsp;stopper)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trains the sequences in db untill stopper says to finnish.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractTrainer.</B><B><A HREF="../../../../org/biojava/bio/dp/AbstractTrainer.html#train(org.biojava.bio.seq.db.SequenceDB, double, org.biojava.bio.dp.StoppingCriteria)">train</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db,
      double&nbsp;nullModelWeight,
      <A HREF="../../../../org/biojava/bio/dp/StoppingCriteria.html" title="interface in org.biojava.bio.dp">StoppingCriteria</A>&nbsp;stopper)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trains the sequences in db until stopper says to finnish.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp.onehead"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/dp/onehead/package-summary.html">org.biojava.bio.dp.onehead</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/dp/onehead/package-summary.html">org.biojava.bio.dp.onehead</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/dp/onehead/SingleDP.html#SingleDP(org.biojava.bio.dp.MarkovModel)">SingleDP</A></B>(<A HREF="../../../../org/biojava/bio/dp/MarkovModel.html" title="interface in org.biojava.bio.dp">MarkovModel</A>&nbsp;model)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp.twohead"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/dp/twohead/package-summary.html">org.biojava.bio.dp.twohead</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/dp/twohead/package-summary.html">org.biojava.bio.dp.twohead</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/dp/twohead/CellCalculatorFactory.html" title="interface in org.biojava.bio.dp.twohead">CellCalculatorFactory</A></CODE></FONT></TD>
<TD><CODE><B>CellCalculatorFactoryMaker.</B><B><A HREF="../../../../org/biojava/bio/dp/twohead/CellCalculatorFactoryMaker.html#make(org.biojava.bio.dp.DP)">make</A></B>(<A HREF="../../../../org/biojava/bio/dp/DP.html" title="class in org.biojava.bio.dp">DP</A>&nbsp;dp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/dp/twohead/package-summary.html">org.biojava.bio.dp.twohead</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/dp/twohead/PairwiseDP.html#PairwiseDP(org.biojava.bio.dp.MarkovModel, org.biojava.bio.dp.twohead.CellCalculatorFactoryMaker)">PairwiseDP</A></B>(<A HREF="../../../../org/biojava/bio/dp/MarkovModel.html" title="interface in org.biojava.bio.dp">MarkovModel</A>&nbsp;mm,
           <A HREF="../../../../org/biojava/bio/dp/twohead/CellCalculatorFactoryMaker.html" title="interface in org.biojava.bio.dp.twohead">CellCalculatorFactoryMaker</A>&nbsp;ccfm)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.gui"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/gui/package-summary.html">org.biojava.bio.gui</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/gui/package-summary.html">org.biojava.bio.gui</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureTree.</B><B><A HREF="../../../../org/biojava/bio/gui/FeatureTree.html#setSequenceDB(org.biojava.bio.seq.db.SequenceDB)">setSequenceDB</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Use this method to provide the sequences for the tree to work with.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.molbio"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/molbio/package-summary.html">org.biojava.bio.molbio</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/molbio/package-summary.html">org.biojava.bio.molbio</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[]</CODE></FONT></TD>
<TD><CODE><B>RestrictionEnzyme.</B><B><A HREF="../../../../org/biojava/bio/molbio/RestrictionEnzyme.html#getUpstreamCut()">getUpstreamCut</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getUpstreamCut</code> returns the cut site upstream of
 the recognition site.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>RestrictionEnzyme.</B><B><A HREF="../../../../org/biojava/bio/molbio/RestrictionEnzyme.html#getUpstreamEndType()">getUpstreamEndType</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getUpstreamEndType</code> returns the double-stranded end
 type produced by the secondary (upstream) cut.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.das"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DASSequence.</B><B><A HREF="../../../../org/biojava/bio/program/das/DASSequence.html#addAnnotationSource(java.net.URL)">addAnnotationSource</A></B>(java.net.URL&nbsp;dataSourceURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DAS.</B><B><A HREF="../../../../org/biojava/bio/program/das/DAS.html#addDasURL(java.net.URL)">addDasURL</A></B>(java.net.URL&nbsp;dasURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>ReferenceServer.</B><B><A HREF="../../../../org/biojava/bio/program/das/ReferenceServer.html#getDB()">getDB</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html" title="interface in org.biojava.bio.seq.distributed">DistDataSource</A></CODE></FONT></TD>
<TD><CODE><B>DataSource.</B><B><A HREF="../../../../org/biojava/bio/program/das/DataSource.html#getDistDataSource()">getDistDataSource</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html" title="interface in org.biojava.bio.seq.distributed">DistDataSource</A></CODE></FONT></TD>
<TD><CODE><B>ReferenceServer.</B><B><A HREF="../../../../org/biojava/bio/program/das/ReferenceServer.html#getDistDataSource()">getDistDataSource</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>DASSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/program/das/DASSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html" title="interface in org.biojava.bio.seq.db">SequenceDBLite</A></CODE></FONT></TD>
<TD><CODE><B>DASSequenceDBProvider.</B><B><A HREF="../../../../org/biojava/bio/program/das/DASSequenceDBProvider.html#getSequenceDB(java.util.Map)">getSequenceDB</A></B>(java.util.Map&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A></CODE></FONT></TD>
<TD><CODE><B>DASSequence.</B><B><A HREF="../../../../org/biojava/bio/program/das/DASSequence.html#getSymbols()">getSymbols</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.util.Set</CODE></FONT></TD>
<TD><CODE><B>DAS.</B><B><A HREF="../../../../org/biojava/bio/program/das/DAS.html#getTypes(java.net.URL)">getTypes</A></B>(java.net.URL&nbsp;dasURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>DASSequence.</B><B><A HREF="../../../../org/biojava/bio/program/das/DASSequence.html#realizeFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;dest,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DASSequence.</B><B><A HREF="../../../../org/biojava/bio/program/das/DASSequence.html#removeAnnotationSource(java.net.URL)">removeAnnotationSource</A></B>(java.net.URL&nbsp;dataSourceURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/program/das/DASSequenceDB.html#DASSequenceDB(java.net.URL)">DASSequenceDB</A></B>(java.net.URL&nbsp;dataSourceURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.das.dasalignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/das/dasalignment/package-summary.html">org.biojava.bio.program.das.dasalignment</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/das/dasalignment/package-summary.html">org.biojava.bio.program.das.dasalignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/program/das/dasalignment/DASException.html" title="class in org.biojava.bio.program.das.dasalignment">DASException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An exception of one of the DAS classes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.formats"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/formats/package-summary.html">org.biojava.bio.program.formats</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/formats/package-summary.html">org.biojava.bio.program.formats</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/formats/Format.html" title="interface in org.biojava.bio.program.formats">Format</A></CODE></FONT></TD>
<TD><CODE><B>FormatTools.</B><B><A HREF="../../../../org/biojava/bio/program/formats/FormatTools.html#getFormat(java.lang.String)">getFormat</A></B>(java.lang.String&nbsp;formatName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Attempt to find aformat for a format identifer string.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#annotateSequences(org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.program.gff.GFFEntrySet)">annotateSequences</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqs,
                  <A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A>&nbsp;ents)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Annotates all sequences in a sequence DB with features from a GFF entry set.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/program/gff/SimpleGFFRecord.html" title="class in org.biojava.bio.program.gff">SimpleGFFRecord</A></CODE></FONT></TD>
<TD><CODE><B>SequencesAsGFF.</B><B><A HREF="../../../../org/biojava/bio/program/gff/SequencesAsGFF.html#createGFFRecord(org.biojava.bio.seq.Feature, java.lang.String)">createGFFRecord</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;feature,
                java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Internal method to create a <span class="type">GFFRecord</span>
 from an individual <span class="type">Feature</span>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFRecord.html" title="interface in org.biojava.bio.program.gff">GFFRecord</A></CODE></FONT></TD>
<TD><CODE><B>GFFParser.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFParser.html#createRecord(org.biojava.bio.program.gff.GFFDocumentHandler, java.util.List, java.lang.String, java.lang.String)">createRecord</A></B>(<A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler,
             java.util.List&nbsp;aList,
             java.lang.String&nbsp;rest,
             java.lang.String&nbsp;comment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Actually turns a list of tokens, some value string and a comment into a
 <span class="type">GFFRecord</span> and informs
 <span class="arg">handler</span>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequencesAsGFF.</B><B><A HREF="../../../../org/biojava/bio/program/gff/SequencesAsGFF.html#doPreProcessSequence(org.biojava.bio.seq.Sequence, org.biojava.bio.program.gff.GFFDocumentHandler, java.lang.String)">doPreProcessSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                     <A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler,
                     java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Emit any per-sequence header information.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequencesAsGFF.</B><B><A HREF="../../../../org/biojava/bio/program/gff/SequencesAsGFF.html#doProcessFeature(org.biojava.bio.seq.Feature, org.biojava.bio.program.gff.GFFDocumentHandler, java.lang.String)">doProcessFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;feature,
                 <A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler,
                 java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Internal method to process an individual <span class="type">Feature</span>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequencesAsGFF.</B><B><A HREF="../../../../org/biojava/bio/program/gff/SequencesAsGFF.html#doProcessSequence(org.biojava.bio.seq.Sequence, org.biojava.bio.program.gff.GFFDocumentHandler, java.lang.String)">doProcessSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                  <A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler,
                  java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Internal method to process an individual <span class="type">Sequence</span>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#gffFromSeqDB(org.biojava.bio.seq.db.SequenceDB)">gffFromSeqDB</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a GFFEntrySet containing one entry for each feature on each
 sequence of a SequenceDB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#gffFromSequence(org.biojava.bio.seq.Sequence)">gffFromSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a GFFEntrySet containing one entry for each feature on a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFFWriter.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFWriter.html#invalidEnd(java.lang.String, java.lang.NumberFormatException)">invalidEnd</A></B>(java.lang.String&nbsp;token,
           java.lang.NumberFormatException&nbsp;nfe)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFFWriter.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFWriter.html#invalidFrame(java.lang.String, java.lang.NumberFormatException)">invalidFrame</A></B>(java.lang.String&nbsp;token,
             java.lang.NumberFormatException&nbsp;nfe)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFFWriter.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFWriter.html#invalidScore(java.lang.String, java.lang.NumberFormatException)">invalidScore</A></B>(java.lang.String&nbsp;token,
             java.lang.NumberFormatException&nbsp;nfe)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFFWriter.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFWriter.html#invalidStart(java.lang.String, java.lang.NumberFormatException)">invalidStart</A></B>(java.lang.String&nbsp;token,
             java.lang.NumberFormatException&nbsp;nfe)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFFWriter.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFWriter.html#invalidStrand(java.lang.String)">invalidStrand</A></B>(java.lang.String&nbsp;token)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFFParser.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFParser.html#parse(java.io.BufferedReader, org.biojava.bio.program.gff.GFFDocumentHandler)">parse</A></B>(java.io.BufferedReader&nbsp;bReader,
      <A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Informs <span class="arg">handler</span> of each line of
 gff read from <span class="arg">bReader</span>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFFParser.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFParser.html#parse(java.io.BufferedReader, org.biojava.bio.program.gff.GFFDocumentHandler, java.lang.String)">parse</A></B>(java.io.BufferedReader&nbsp;bReader,
      <A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler,
      java.lang.String&nbsp;locator)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Informs <span class="arg">handler</span> of each line of
 GFF read from <span class="arg">bReader</span></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequencesAsGFF.</B><B><A HREF="../../../../org/biojava/bio/program/gff/SequencesAsGFF.html#processDB(org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.program.gff.GFFDocumentHandler)">processDB</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
          <A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Process all <span class="type">Sequence</span>s within a
 <span class="type">SequenceDB</span>, informing
 <span class="arg">handler</span> of any suitable features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequencesAsGFF.</B><B><A HREF="../../../../org/biojava/bio/program/gff/SequencesAsGFF.html#processSequence(org.biojava.bio.seq.Sequence, org.biojava.bio.program.gff.GFFDocumentHandler)">processSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
                <A HREF="../../../../org/biojava/bio/program/gff/GFFDocumentHandler.html" title="interface in org.biojava.bio.program.gff">GFFDocumentHandler</A>&nbsp;handler)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Process an individual <span class="type">Sequence</span>, informing
 <span class="arg">handler</span> of any suitable features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#readGFF(java.io.BufferedReader)">readGFF</A></B>(java.io.BufferedReader&nbsp;gffIn)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read all GFF entries from a buffered reader.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#readGFF(java.io.BufferedReader, org.biojava.bio.program.gff.GFFRecordFilter)">readGFF</A></B>(java.io.BufferedReader&nbsp;gffIn,
        <A HREF="../../../../org/biojava/bio/program/gff/GFFRecordFilter.html" title="interface in org.biojava.bio.program.gff">GFFRecordFilter</A>&nbsp;recFilt)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read all GFF entries matching a filter from a buffered reader.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#readGFF(java.io.File)">readGFF</A></B>(java.io.File&nbsp;inFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a <code>GFFEntrySet</code> from a file with no filtering.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#readGFF(java.io.File, org.biojava.bio.program.gff.GFFRecordFilter)">readGFF</A></B>(java.io.File&nbsp;inFile,
        <A HREF="../../../../org/biojava/bio/program/gff/GFFRecordFilter.html" title="interface in org.biojava.bio.program.gff">GFFRecordFilter</A>&nbsp;recFilt)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a GFFEntrySet from a file with the specified filter.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#readGFF(java.lang.String)">readGFF</A></B>(java.lang.String&nbsp;fileName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use: readGff(File)</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/gff/GFFEntrySet.html" title="class in org.biojava.bio.program.gff">GFFEntrySet</A></CODE></FONT></TD>
<TD><CODE><B>GFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/gff/GFFTools.html#readGFF(java.lang.String, org.biojava.bio.program.gff.GFFRecordFilter)">readGFF</A></B>(java.lang.String&nbsp;fileName,
        <A HREF="../../../../org/biojava/bio/program/gff/GFFRecordFilter.html" title="interface in org.biojava.bio.program.gff">GFFRecordFilter</A>&nbsp;recFilt)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use: readGff(File,GFFRecordFilter)</I></TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff3"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/gff3/package-summary.html">org.biojava.bio.program.gff3</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/gff3/package-summary.html">org.biojava.bio.program.gff3</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/program/gff3/GFF3Record.html" title="interface in org.biojava.bio.program.gff3">GFF3Record</A></CODE></FONT></TD>
<TD><CODE><B>GFF3Parser.</B><B><A HREF="../../../../org/biojava/bio/program/gff3/GFF3Parser.html#createRecord(org.biojava.bio.program.gff3.GFF3DocumentHandler, java.util.List, java.lang.String, java.lang.String, org.biojava.ontology.Ontology, org.biojava.ontology.Ontology)">createRecord</A></B>(<A HREF="../../../../org/biojava/bio/program/gff3/GFF3DocumentHandler.html" title="interface in org.biojava.bio.program.gff3">GFF3DocumentHandler</A>&nbsp;handler,
             java.util.List&nbsp;aList,
             java.lang.String&nbsp;rest,
             java.lang.String&nbsp;comment,
             <A HREF="../../../../org/biojava/ontology/Ontology.html" title="interface in org.biojava.ontology">Ontology</A>&nbsp;ontology,
             <A HREF="../../../../org/biojava/ontology/Ontology.html" title="interface in org.biojava.ontology">Ontology</A>&nbsp;fallBack)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Actually turns a list of tokens, some value string and a comment into a
 <span class="type">GFF3Record</span> and informs
 <span class="arg">handler</span>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFF3Parser.</B><B><A HREF="../../../../org/biojava/bio/program/gff3/GFF3Parser.html#parse(java.io.BufferedReader, org.biojava.bio.program.gff3.GFF3DocumentHandler, org.biojava.ontology.Ontology)">parse</A></B>(java.io.BufferedReader&nbsp;bReader,
      <A HREF="../../../../org/biojava/bio/program/gff3/GFF3DocumentHandler.html" title="interface in org.biojava.bio.program.gff3">GFF3DocumentHandler</A>&nbsp;handler,
      <A HREF="../../../../org/biojava/ontology/Ontology.html" title="interface in org.biojava.ontology">Ontology</A>&nbsp;ontology)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Informs <span class="arg">handler</span> of each line of
 gff read from <span class="arg">bReader</span>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>GFF3Parser.</B><B><A HREF="../../../../org/biojava/bio/program/gff3/GFF3Parser.html#parse(java.io.BufferedReader, org.biojava.bio.program.gff3.GFF3DocumentHandler, org.biojava.ontology.Ontology, java.lang.String)">parse</A></B>(java.io.BufferedReader&nbsp;bReader,
      <A HREF="../../../../org/biojava/bio/program/gff3/GFF3DocumentHandler.html" title="interface in org.biojava.bio.program.gff3">GFF3DocumentHandler</A>&nbsp;handler,
      <A HREF="../../../../org/biojava/ontology/Ontology.html" title="interface in org.biojava.ontology">Ontology</A>&nbsp;ontology,
      java.lang.String&nbsp;locator)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Informs <span class="arg">handler</span> of each line of
 GFF read from <span class="arg">bReader</span></TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.homologene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/homologene/package-summary.html">org.biojava.bio.program.homologene</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/homologene/package-summary.html">org.biojava.bio.program.homologene</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/program/homologene/DuplicateTaxonException.html" title="class in org.biojava.bio.program.homologene">DuplicateTaxonException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.indexdb"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/indexdb/package-summary.html">org.biojava.bio.program.indexdb</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/indexdb/package-summary.html">org.biojava.bio.program.indexdb</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/indexdb/BioStore.html" title="class in org.biojava.bio.program.indexdb">BioStore</A></CODE></FONT></TD>
<TD><CODE><B>BioStoreFactory.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/BioStoreFactory.html#createBioStore()">createBioStore</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>createBioStore</code> creates a <code>BioStore</code>
 reflecting the current state of the factory and returns a
 reference to it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/indexdb/Record.html" title="interface in org.biojava.bio.program.indexdb">Record</A></CODE></FONT></TD>
<TD><CODE><B>IndexStore.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/IndexStore.html#get(java.lang.String)">get</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>get</code> returns a record specified by a primary
 identifier.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B>BioStore.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/BioStore.html#get(java.lang.String, java.lang.String)">get</A></B>(java.lang.String&nbsp;id,
    java.lang.String&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B>IndexStore.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/IndexStore.html#get(java.lang.String, java.lang.String)">get</A></B>(java.lang.String&nbsp;id,
    java.lang.String&nbsp;namespace)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>get</code> returns a list of <code>Record</code>s by
 searching against the primary identifiers if the namespace
 argument is equal to the primary namespace or otherwise by
 searching the secondary namespaces.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>IndexTools.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/IndexTools.html#indexEmbl(java.lang.String, java.io.File, java.io.File[], int)">indexEmbl</A></B>(java.lang.String&nbsp;name,
          java.io.File&nbsp;location,
          java.io.File[]&nbsp;seqFiles,
          int&nbsp;alphabetIdentifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>indexEmbl</code> indexes DNA, RNA or protein EMBL format
 sequence files on ID as primary identifier and AC as secondary.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>IndexTools.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/IndexTools.html#indexFasta(java.lang.String, java.io.File, java.io.File[], int)">indexFasta</A></B>(java.lang.String&nbsp;name,
           java.io.File&nbsp;location,
           java.io.File[]&nbsp;seqFiles,
           int&nbsp;alphabetIdentifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>indexFasta</code> indexes DNA, RNA or protein Fasta
 format sequence files on primary identifier.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>IndexTools.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/IndexTools.html#indexGenbank(java.lang.String, java.io.File, java.io.File[], int)">indexGenbank</A></B>(java.lang.String&nbsp;name,
             java.io.File&nbsp;location,
             java.io.File[]&nbsp;seqFiles,
             int&nbsp;alphabetIdentifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>indexGenbank</code> indexes DNA, RNA or protein Genbank
 format sequence files on LOCUS as primary identifier and
 ACCESSION as secondary.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>IndexTools.</B><B><A HREF="../../../../org/biojava/bio/program/indexdb/IndexTools.html#indexSwissprot(java.lang.String, java.io.File, java.io.File[])">indexSwissprot</A></B>(java.lang.String&nbsp;name,
               java.io.File&nbsp;location,
               java.io.File[]&nbsp;seqFiles)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>indexSwissprot</code> indexes Swissprot format protein
 sequence files on ID as primary identifier.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/program/indexdb/package-summary.html">org.biojava.bio.program.indexdb</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/program/indexdb/BioStore.html#BioStore(java.io.File, boolean)">BioStore</A></B>(java.io.File&nbsp;location,
         boolean&nbsp;cache)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new <code>BioStore</code> flatfile index at the
 specified location with the specified caching behaviour.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PhredTools.</B><B><A HREF="../../../../org/biojava/bio/program/phred/PhredTools.html#writePhredQuality(java.io.OutputStream, java.io.OutputStream, org.biojava.bio.seq.db.SequenceDB)">writePhredQuality</A></B>(java.io.OutputStream&nbsp;qual,
                  java.io.OutputStream&nbsp;seq,
                  <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes Phred quality data in a Fasta type format.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.sax.blastxml"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/sax/blastxml/package-summary.html">org.biojava.bio.program.sax.blastxml</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/program/sax/blastxml/package-summary.html">org.biojava.bio.program.sax.blastxml</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/program/sax/blastxml/BlastXMLParserFacade.html#BlastXMLParserFacade()">BlastXMLParserFacade</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssaha"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>DataStoreFactory.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/DataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Build a new DataStore.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>CompactedDataStoreFactory.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/CompactedDataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>MappedDataStoreFactory.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/MappedDataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>NIODataStoreFactory.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/NIODataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.seq.db.SequenceDB, org.biojava.bio.symbol.Packing, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;seqDB,
               <A HREF="../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/ssaha/DataStore.html" title="interface in org.biojava.bio.program.ssaha">DataStore</A></CODE></FONT></TD>
<TD><CODE><B>CompactedDataStoreFactory.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/CompactedDataStoreFactory.html#buildDataStore(java.io.File, org.biojava.bio.program.ssaha.SequenceStreamer, org.biojava.bio.symbol.Packing, int, int, int)">buildDataStore</A></B>(java.io.File&nbsp;storeFile,
               <A HREF="../../../../org/biojava/bio/program/ssaha/SequenceStreamer.html" title="interface in org.biojava.bio.program.ssaha">SequenceStreamer</A>&nbsp;streamer,
               <A HREF="../../../../org/biojava/bio/symbol/Packing.html" title="interface in org.biojava.bio.symbol">Packing</A>&nbsp;packing,
               int&nbsp;wordLength,
               int&nbsp;stepSize,
               int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceStreamer.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/SequenceStreamer.html#reset()">reset</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceStreamer.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/SequenceStreamer.html#streamNext(org.biojava.bio.seq.io.SeqIOListener)">streamNext</A></B>(<A HREF="../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceStreamer.SequenceDBStreamer.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/SequenceStreamer.SequenceDBStreamer.html#streamNext(org.biojava.bio.seq.io.SeqIOListener)">streamNext</A></B>(<A HREF="../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceStreamer.FileStreamer.</B><B><A HREF="../../../../org/biojava/bio/program/ssaha/SequenceStreamer.FileStreamer.html#streamNext(org.biojava.bio.seq.io.SeqIOListener)">streamNext</A></B>(<A HREF="../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssbind"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>ViewSequenceFactory.</B><B><A HREF="../../../../org/biojava/bio/program/ssbind/ViewSequenceFactory.html#makeQueryViewSequence(java.lang.String)">makeQueryViewSequence</A></B>(java.lang.String&nbsp;queryID)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/search/SeqSimilaritySearchResult.html" title="interface in org.biojava.bio.search">SeqSimilaritySearchResult</A></CODE></FONT></TD>
<TD><CODE><B>BlastLikeSearchBuilder.</B><B><A HREF="../../../../org/biojava/bio/program/ssbind/BlastLikeSearchBuilder.html#makeSearchResult()">makeSearchResult</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>ViewSequenceFactory.</B><B><A HREF="../../../../org/biojava/bio/program/ssbind/ViewSequenceFactory.html#makeSubjectViewSequence(java.lang.String)">makeSubjectViewSequence</A></B>(java.lang.String&nbsp;subjectID)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A></CODE></FONT></TD>
<TD><CODE><B>AlphabetResolver.</B><B><A HREF="../../../../org/biojava/bio/program/ssbind/AlphabetResolver.html#resolveAlphabet(java.lang.String)">resolveAlphabet</A></B>(java.lang.String&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>resolveAlphabet</code> returns an appropriate
 <code>Alphabet</code> for an arbitrary identifier.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.unigene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneCluster.html" title="interface in org.biojava.bio.program.unigene">UnigeneCluster</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneDB.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html#addCluster(org.biojava.bio.program.unigene.UnigeneCluster)">addCluster</A></B>(<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneCluster.html" title="interface in org.biojava.bio.program.unigene">UnigeneCluster</A>&nbsp;cluster)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a cluster to a database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneTools.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/UnigeneTools.html#createUnigene(java.net.URL)">createUnigene</A></B>(java.net.URL&nbsp;dbURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new UnigeneDB instance referred to by a URL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>FlatFileUnigeneFactory.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/FlatFileUnigeneFactory.html#createUnigene(java.net.URL)">createUnigene</A></B>(java.net.URL&nbsp;unigeneLoc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneFactory.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/UnigeneFactory.html#createUnigene(java.net.URL)">createUnigene</A></B>(java.net.URL&nbsp;unigeneURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>SQLUnigeneFactory.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/SQLUnigeneFactory.html#createUnigene(java.net.URL)">createUnigene</A></B>(java.net.URL&nbsp;dbURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneFactory.html" title="interface in org.biojava.bio.program.unigene">UnigeneFactory</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneTools.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/UnigeneTools.html#findFactory(java.net.URL)">findFactory</A></B>(java.net.URL&nbsp;dbURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Find the UnigeneFactory that can accept a URL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneCluster.html" title="interface in org.biojava.bio.program.unigene">UnigeneCluster</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneDB.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html#getCluster(java.lang.String)">getCluster</A></B>(java.lang.String&nbsp;clusterID)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetch a cluster by its cluster id.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneTools.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/UnigeneTools.html#loadUnigene(java.net.URL)">loadUnigene</A></B>(java.net.URL&nbsp;dbURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Load a UnigeneDB instance referred to by a URL.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>FlatFileUnigeneFactory.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/FlatFileUnigeneFactory.html#loadUnigene(java.net.URL)">loadUnigene</A></B>(java.net.URL&nbsp;unigeneLoc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>UnigeneFactory.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/UnigeneFactory.html#loadUnigene(java.net.URL)">loadUnigene</A></B>(java.net.URL&nbsp;unigeneURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/program/unigene/UnigeneDB.html" title="interface in org.biojava.bio.program.unigene">UnigeneDB</A></CODE></FONT></TD>
<TD><CODE><B>SQLUnigeneFactory.</B><B><A HREF="../../../../org/biojava/bio/program/unigene/SQLUnigeneFactory.html#loadUnigene(java.net.URL)">loadUnigene</A></B>(java.net.URL&nbsp;dbURL)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.xff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/program/xff/package-summary.html">org.biojava.bio.program.xff</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/program/xff/package-summary.html">org.biojava.bio.program.xff</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>XFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/xff/XFFTools.html#annotateXFF(java.io.File, org.biojava.bio.seq.Sequence)">annotateXFF</A></B>(java.io.File&nbsp;xffFile,
            <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;sequence)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>XFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/xff/XFFTools.html#annotateXFF(java.io.File, org.biojava.bio.seq.Sequence, org.biojava.bio.Annotation)">annotateXFF</A></B>(java.io.File&nbsp;xffFile,
            <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;sequence,
            <A HREF="../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A>&nbsp;ann)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>XFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/xff/XFFTools.html#readXFF(java.io.File, java.lang.String)">readXFF</A></B>(java.io.File&nbsp;xffFile,
        java.lang.String&nbsp;seqID)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>XFFTools.</B><B><A HREF="../../../../org/biojava/bio/program/xff/XFFTools.html#readXFF(java.io.File, java.lang.String, org.biojava.bio.symbol.FiniteAlphabet)">readXFF</A></B>(java.io.File&nbsp;xffFile,
        java.lang.String&nbsp;seqID,
        <A HREF="../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.proteomics"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/proteomics/package-summary.html">org.biojava.bio.proteomics</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/proteomics/package-summary.html">org.biojava.bio.proteomics</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>Digest.</B><B><A HREF="../../../../org/biojava/bio/proteomics/Digest.html#addDigestFeatures()">addDigestFeatures</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds peptides as features to the Sequence in this class.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/proteomics/Protease.html" title="class in org.biojava.bio.proteomics">Protease</A></CODE></FONT></TD>
<TD><CODE><B>ProteaseManager.</B><B><A HREF="../../../../org/biojava/bio/proteomics/ProteaseManager.html#createProtease(java.lang.String, boolean, java.lang.String)">createProtease</A></B>(java.lang.String&nbsp;cleaveRes,
               boolean&nbsp;endoProtease,
               java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/proteomics/Protease.html" title="class in org.biojava.bio.proteomics">Protease</A></CODE></FONT></TD>
<TD><CODE><B>ProteaseManager.</B><B><A HREF="../../../../org/biojava/bio/proteomics/ProteaseManager.html#createProtease(java.lang.String, boolean, java.lang.String, java.lang.String)">createProtease</A></B>(java.lang.String&nbsp;cleaveRes,
               boolean&nbsp;endoProtease,
               java.lang.String&nbsp;notCleaveRes,
               java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/proteomics/Protease.html" title="class in org.biojava.bio.proteomics">Protease</A></CODE></FONT></TD>
<TD><CODE><B>ProteaseManager.</B><B><A HREF="../../../../org/biojava/bio/proteomics/ProteaseManager.html#createProtease(org.biojava.bio.symbol.SymbolList, boolean, java.lang.String)">createProtease</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;cleaveRes,
               boolean&nbsp;endoProtease,
               java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/proteomics/Protease.html" title="class in org.biojava.bio.proteomics">Protease</A></CODE></FONT></TD>
<TD><CODE><B>ProteaseManager.</B><B><A HREF="../../../../org/biojava/bio/proteomics/ProteaseManager.html#createProtease(org.biojava.bio.symbol.SymbolList, boolean, org.biojava.bio.symbol.SymbolList, java.lang.String)">createProtease</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;cleaveRes,
               boolean&nbsp;endoProtease,
               <A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;notCleaveRes,
               java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates and registers a new Protease.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B>IsoelectricPointCalc.</B><B><A HREF="../../../../org/biojava/bio/proteomics/IsoelectricPointCalc.html#getIsoelectricPoint(org.biojava.bio.symbol.SymbolList)">getIsoelectricPoint</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;peptide)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Static public method to compute the pI for a polypeptide in
 denaturating and reduced conditions with both free ends.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B>IsoelectricPointCalc.</B><B><A HREF="../../../../org/biojava/bio/proteomics/IsoelectricPointCalc.html#getPI(org.biojava.bio.symbol.SymbolList, boolean, boolean)">getPI</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;peptide,
      boolean&nbsp;hasFreeNTerm,
      boolean&nbsp;hasFreeCTerm)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Computes isoelectric point of specified peptide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/proteomics/Protease.html" title="class in org.biojava.bio.proteomics">Protease</A></CODE></FONT></TD>
<TD><CODE><B>ProteaseManager.</B><B><A HREF="../../../../org/biojava/bio/proteomics/ProteaseManager.html#getProteaseByName(java.lang.String)">getProteaseByName</A></B>(java.lang.String&nbsp;proteaseName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets a Protease instance by name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/proteomics/Protease.html" title="class in org.biojava.bio.proteomics">Protease</A></CODE></FONT></TD>
<TD><CODE><B>Protease.</B><B><A HREF="../../../../org/biojava/bio/proteomics/Protease.html#getProteaseByName(java.lang.String)">getProteaseByName</A></B>(java.lang.String&nbsp;proteaseName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a reference to the named Protease.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProteaseManager.</B><B><A HREF="../../../../org/biojava/bio/proteomics/ProteaseManager.html#registerProtease(org.biojava.bio.proteomics.Protease)">registerProtease</A></B>(<A HREF="../../../../org/biojava/bio/proteomics/Protease.html" title="class in org.biojava.bio.proteomics">Protease</A>&nbsp;prot)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Registers a protease and ensures its flyweight status</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/proteomics/package-summary.html">org.biojava.bio.proteomics</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/proteomics/Protease.html#Protease(java.lang.String, boolean)">Protease</A></B>(java.lang.String&nbsp;cleavageRes,
         boolean&nbsp;endoProtease)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>Creating a Protease with this constructor will not register it
 with the ProteaseManager (use ProteaseManager.createProtease())</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/proteomics/Protease.html#Protease(java.lang.String, boolean, java.lang.String)">Protease</A></B>(java.lang.String&nbsp;cleaveRes,
         boolean&nbsp;endoProtease,
         java.lang.String&nbsp;notCleaveRes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>Creating a Protease with this constructor will not register it
 with the ProteaseManager (use ProteaseManager.createProtease())</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/proteomics/Protease.html#Protease(org.biojava.bio.symbol.SymbolList, boolean, org.biojava.bio.symbol.SymbolList)">Protease</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;cleaveRes,
         boolean&nbsp;endoProtease,
         <A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;notCleaveRes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>Creating a Protease with this constructor will not register it
 with the ProteaseManager (use ProteaseManager.createProtease())</I></TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.proteomics.aaindex"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/proteomics/aaindex/package-summary.html">org.biojava.bio.proteomics.aaindex</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/proteomics/aaindex/package-summary.html">org.biojava.bio.proteomics.aaindex</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/symbol/SymbolPropertyTable.html" title="interface in org.biojava.bio.symbol">SymbolPropertyTable</A></CODE></FONT></TD>
<TD><CODE><B>SymbolPropertyTableIterator.</B><B><A HREF="../../../../org/biojava/bio/proteomics/aaindex/SymbolPropertyTableIterator.html#nextTable()">nextTable</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the next <A HREF="../../../../org/biojava/bio/symbol/SymbolPropertyTable.html" title="interface in org.biojava.bio.symbol"><CODE>SymbolPropertyTable</CODE></A> 
 object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/symbol/SymbolPropertyTable.html" title="interface in org.biojava.bio.symbol">SymbolPropertyTable</A></CODE></FONT></TD>
<TD><CODE><B>AAindexStreamReader.</B><B><A HREF="../../../../org/biojava/bio/proteomics/aaindex/AAindexStreamReader.html#nextTable()">nextTable</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the next <A HREF="../../../../org/biojava/bio/symbol/SymbolPropertyTable.html" title="interface in org.biojava.bio.symbol"><CODE>SymbolPropertyTable</CODE></A> 
 object.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/proteomics/aaindex/package-summary.html">org.biojava.bio.proteomics.aaindex</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/proteomics/aaindex/SimpleSymbolPropertyTableDB.html#SimpleSymbolPropertyTableDB(org.biojava.bio.proteomics.aaindex.SymbolPropertyTableIterator)">SimpleSymbolPropertyTableDB</A></B>(<A HREF="../../../../org/biojava/bio/proteomics/aaindex/SymbolPropertyTableIterator.html" title="interface in org.biojava.bio.proteomics.aaindex">SymbolPropertyTableIterator</A>&nbsp;tableIterator)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Initializes the database by copying all symbol property tables from 
 a given iterator into the database.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.search"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/search/SeqSimilaritySearchResult.html" title="interface in org.biojava.bio.search">SeqSimilaritySearchResult</A></CODE></FONT></TD>
<TD><CODE><B>SearchBuilder.</B><B><A HREF="../../../../org/biojava/bio/search/SearchBuilder.html#makeSearchResult()">makeSearchResult</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The <code>makeSearchResult</code> method returns a
 <code>SeqSimilaritySearchResult</code> instance created from
 accumulated data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/search/SeqSimilaritySearchResult.html" title="interface in org.biojava.bio.search">SeqSimilaritySearchResult</A></CODE></FONT></TD>
<TD><CODE><B>SeqSimilaritySearcher.</B><B><A HREF="../../../../org/biojava/bio/search/SeqSimilaritySearcher.html#search(org.biojava.bio.symbol.SymbolList, org.biojava.bio.seq.db.SequenceDB, java.util.Map)">search</A></B>(<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;querySeq,
       <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;db,
       java.util.Map&nbsp;searchParameters)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Using this sequence similarity searcher, search with the given
 sequence against the given sequence database.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/search/FilteringContentHandler.html#FilteringContentHandler(org.biojava.bio.search.BlastLikeSearchFilter)">FilteringContentHandler</A></B>(<A HREF="../../../../org/biojava/bio/search/BlastLikeSearchFilter.html" title="interface in org.biojava.bio.search">BlastLikeSearchFilter</A>&nbsp;filter)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/search/FilteringContentHandler.html#FilteringContentHandler(org.biojava.bio.search.BlastLikeSearchFilter, org.biojava.bio.search.SearchContentHandler)">FilteringContentHandler</A></B>(<A HREF="../../../../org/biojava/bio/search/BlastLikeSearchFilter.html" title="interface in org.biojava.bio.search">BlastLikeSearchFilter</A>&nbsp;filter,
                        <A HREF="../../../../org/biojava/bio/search/SearchContentHandler.html" title="interface in org.biojava.bio.search">SearchContentHandler</A>&nbsp;delegate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceTools.</B><B><A HREF="../../../../org/biojava/bio/seq/SequenceTools.html#addAllFeatures(org.biojava.bio.seq.Sequence, org.biojava.bio.seq.FeatureHolder)">addAllFeatures</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
               <A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;fh)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add features to a sequence that contain the same information as all
 those in a feature holder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleFeatureRealizer.</B><B><A HREF="../../../../org/biojava/bio/seq/SimpleFeatureRealizer.html#addImplementation(java.lang.Class, java.lang.Class)">addImplementation</A></B>(java.lang.Class&nbsp;template,
                  java.lang.Class&nbsp;impl)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Install a new mapping from a class of Feature.Template to
 a class of Feature implementations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceAnnotator.</B><B><A HREF="../../../../org/biojava/bio/seq/SequenceAnnotator.html#annotate(org.biojava.bio.seq.Sequence)">annotate</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return an annotated version of a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>FeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;ft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new Feature, and add it to this FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>NewSimpleAssembly.</B><B><A HREF="../../../../org/biojava/bio/seq/NewSimpleAssembly.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>CircularView.</B><B><A HREF="../../../../org/biojava/bio/seq/CircularView.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Over rides ViewSequence to allow the use of locations that have
 coordinates outside of the sequence length (which are needed to
 describe locations that overlap the origin of a circular sequence).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleAssembly.</B><B><A HREF="../../../../org/biojava/bio/seq/SimpleAssembly.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>LazyFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/LazyFeatureHolder.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>AbstractFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/AbstractFeatureHolder.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>RemoteFeature.</B><B><A HREF="../../../../org/biojava/bio/seq/RemoteFeature.html#getRemoteFeature()">getRemoteFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve the Feature on some assembly Sequence that can represent this
 RemoteFeature properly.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceIterator.</B><B><A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html#nextSequence()">nextSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the next sequence in the iterator.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>RealizingFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/RealizingFeatureHolder.html#realizeFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;parent,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Realize a feature template.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>NewSimpleAssembly.</B><B><A HREF="../../../../org/biojava/bio/seq/NewSimpleAssembly.html#realizeFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;fh,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleAssembly.</B><B><A HREF="../../../../org/biojava/bio/seq/SimpleAssembly.html#realizeFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;fh,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>FeatureRealizer.</B><B><A HREF="../../../../org/biojava/bio/seq/FeatureRealizer.html#realizeFeature(org.biojava.bio.seq.Sequence, org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
               <A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;parent,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Install a feature on the specified sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleFeatureRealizer.</B><B><A HREF="../../../../org/biojava/bio/seq/SimpleFeatureRealizer.html#realizeFeature(org.biojava.bio.seq.Sequence, org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq,
               <A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;parent,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove a feature from this FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>LazyFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/LazyFeatureHolder.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/AbstractFeatureHolder.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>RemoteFeature.Resolver.</B><B><A HREF="../../../../org/biojava/bio/seq/RemoteFeature.Resolver.html#resolve(org.biojava.bio.seq.RemoteFeature)">resolve</A></B>(<A HREF="../../../../org/biojava/bio/seq/RemoteFeature.html" title="interface in org.biojava.bio.seq">RemoteFeature</A>&nbsp;rFeat)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Resolve rFeat.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>FilterUtils.</B><B><A HREF="../../../../org/biojava/bio/seq/FilterUtils.html#visitFilter(org.biojava.bio.seq.FeatureFilter, org.biojava.utils.walker.Visitor)">visitFilter</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A>&nbsp;filter,
            <A HREF="../../../../org/biojava/utils/walker/Visitor.html" title="interface in org.biojava.utils.walker">Visitor</A>&nbsp;visitor)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Applies a visitor to a filter, and returns the visitor's result or null.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/IllegalIDException.html" title="class in org.biojava.bio.seq.db">IllegalIDException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>IndexedSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/IndexedSequenceDB.html#addFile(java.io.File)">addFile</A></B>(java.io.File&nbsp;seqFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add sequences from a file to the sequence database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceDBLite.</B><B><A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html#addSequence(org.biojava.bio.seq.Sequence)">addSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds a sequence to the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DummySequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/DummySequenceDB.html#addSequence(org.biojava.bio.seq.Sequence)">addSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/AbstractSequenceDB.html#addSequence(org.biojava.bio.seq.Sequence)">addSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TabIndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/TabIndexStore.html#commit()">commit</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblCDROMIndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/EmblCDROMIndexStore.html#commit()">commit</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>commit</code> commits changes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioIndex.</B><B><A HREF="../../../../org/biojava/bio/seq/db/BioIndex.html#commit()">commit</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>IndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/IndexStore.html#commit()">commit</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Commit the stored indices to permanent storage.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>AnnotatedSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/AnnotatedSequenceDB.html#doAnnotation(org.biojava.bio.seq.Sequence)">doAnnotation</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Apply the annotation to a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/Index.html" title="interface in org.biojava.bio.seq.db">Index</A></CODE></FONT></TD>
<TD><CODE><B>TabIndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/TabIndexStore.html#fetch(java.lang.String)">fetch</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/Index.html" title="interface in org.biojava.bio.seq.db">Index</A></CODE></FONT></TD>
<TD><CODE><B>EmblCDROMIndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/EmblCDROMIndexStore.html#fetch(java.lang.String)">fetch</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/Index.html" title="interface in org.biojava.bio.seq.db">Index</A></CODE></FONT></TD>
<TD><CODE><B>BioIndex.</B><B><A HREF="../../../../org/biojava/bio/seq/db/BioIndex.html#fetch(java.lang.String)">fetch</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/Index.html" title="interface in org.biojava.bio.seq.db">Index</A></CODE></FONT></TD>
<TD><CODE><B>IndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/IndexStore.html#fetch(java.lang.String)">fetch</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetch an Index based upon an ID.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBLite.</B><B><A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve a single sequence by its id.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>DummySequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/DummySequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SubSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/SubSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>AnnotatedSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/AnnotatedSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SwissprotSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/SwissprotSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>WebSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/WebSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets a sequence using its unique ID (eg for GenBank this would be the GI number)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>ViewingSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/ViewingSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>CachingSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/CachingSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>IndexedSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/IndexedSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>GenpeptSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/GenpeptSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenbankSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/GenbankSequenceDB.html#getSequences(java.util.Set)">getSequences</A></B>(java.util.Set&nbsp;list)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve sequences from a Genbank</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenbankSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/GenbankSequenceDB.html#getSequences(java.util.Set, org.biojava.bio.seq.db.SequenceDB)">getSequences</A></B>(java.util.Set&nbsp;list,
             <A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;database)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve sequences from a Genbank</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SequenceDBLite.</B><B><A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html#removeSequence(java.lang.String)">removeSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove the sequence associated with an ID from the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DummySequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/DummySequenceDB.html#removeSequence(java.lang.String)">removeSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HashSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/HashSequenceDB.html#removeSequence(java.lang.String)">removeSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/AbstractSequenceDB.html#removeSequence(java.lang.String)">removeSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NCBISequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/NCBISequenceDB.html#setDatabase(java.lang.String)">setDatabase</A></B>(java.lang.String&nbsp;dataBase)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NCBISequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/NCBISequenceDB.html#setSequenceFormat(org.biojava.bio.seq.io.SequenceFormat)">setSequenceFormat</A></B>(<A HREF="../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TabIndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/TabIndexStore.html#store(org.biojava.bio.seq.db.Index)">store</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/Index.html" title="interface in org.biojava.bio.seq.db">Index</A>&nbsp;indx)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>EmblCDROMIndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/EmblCDROMIndexStore.html#store(org.biojava.bio.seq.db.Index)">store</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/Index.html" title="interface in org.biojava.bio.seq.db">Index</A>&nbsp;index)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>store</code> adds an <code>Index</code> to the store.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>IndexStore.</B><B><A HREF="../../../../org/biojava/bio/seq/db/IndexStore.html#store(org.biojava.bio.seq.db.Index)">store</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/Index.html" title="interface in org.biojava.bio.seq.db">Index</A>&nbsp;indx)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add the Index to the store.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/BioIndex.html#BioIndex(java.io.File)">BioIndex</A></B>(java.io.File&nbsp;indexDirectory)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Load an existing index file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/BioIndex.html#BioIndex(java.io.File, java.lang.String, int)">BioIndex</A></B>(java.io.File&nbsp;indexDirectory,
         java.lang.String&nbsp;namespace,
         int&nbsp;idLength)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/SubSequenceDB.html#SubSequenceDB(org.biojava.bio.seq.db.SequenceDB, java.util.Set)">SubSequenceDB</A></B>(<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A>&nbsp;parent,
              java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/TabIndexStore.html#TabIndexStore(java.io.File, java.io.File, java.lang.String, org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.seq.io.SequenceBuilderFactory, org.biojava.bio.seq.io.SymbolTokenization)">TabIndexStore</A></B>(java.io.File&nbsp;storeFile,
              java.io.File&nbsp;indexFile,
              java.lang.String&nbsp;name,
              <A HREF="../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
              <A HREF="../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A>&nbsp;sbFactory,
              <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symbolParser)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new TabIndexStore.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biofetch"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/db/biofetch/package-summary.html">org.biojava.bio.seq.db.biofetch</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/db/biofetch/package-summary.html">org.biojava.bio.seq.db.biofetch</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>BioFetchSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biofetch/BioFetchSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html" title="interface in org.biojava.bio.seq.db">SequenceDBLite</A></CODE></FONT></TD>
<TD><CODE><B>BioFetchSequenceDBProvider.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biofetch/BioFetchSequenceDBProvider.html#getSequenceDB(java.util.Map)">getSequenceDB</A></B>(java.util.Map&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A></CODE></FONT></TD>
<TD><CODE><B>TaxonSQL.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/TaxonSQL.html#addName(java.sql.Connection, org.biojava.bio.taxa.Taxon, java.lang.String, java.lang.String)">addName</A></B>(java.sql.Connection&nbsp;conn,
        <A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A>&nbsp;taxon,
        java.lang.String&nbsp;nameClass,
        java.lang.String&nbsp;newName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Adds a new name of the given <code>nameClass</code> to the taxon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#addSequence(org.biojava.bio.seq.Sequence)">addSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#createDummySequence(java.lang.String, org.biojava.bio.symbol.Alphabet, int)">createDummySequence</A></B>(java.lang.String&nbsp;id,
                    <A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alphabet,
                    int&nbsp;length)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.util.Stack</CODE></FONT></TD>
<TD><CODE><B>TaxonSQL.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/TaxonSQL.html#getChildrenOf(java.sql.Connection, java.lang.String)">getChildrenOf</A></B>(java.sql.Connection&nbsp;conn,
              java.lang.String&nbsp;scientificName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Returns all children of the specified taxon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.util.Stack</CODE></FONT></TD>
<TD><CODE><B>TaxonSQL.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/TaxonSQL.html#getChildrenOf(java.sql.Connection, org.biojava.bio.taxa.Taxon)">getChildrenOf</A></B>(java.sql.Connection&nbsp;conn,
              <A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A>&nbsp;t)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Returns the children as a <code>Stack</code> of this given taxon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#getSequence(int)">getSequence</A></B>(int&nbsp;bioentry_id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html" title="interface in org.biojava.bio.seq.db">SequenceDBLite</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLSequenceDBProvider.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDBProvider.html#getSequenceDB(java.util.Map)">getSequenceDB</A></B>(java.util.Map&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A></CODE></FONT></TD>
<TD><CODE><B>TaxonSQL.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/TaxonSQL.html#removeName(java.sql.Connection, org.biojava.bio.seq.db.biosql.DBHelper, org.biojava.bio.taxa.Taxon, java.lang.String, java.lang.String)">removeName</A></B>(java.sql.Connection&nbsp;conn,
           <A HREF="../../../../org/biojava/bio/seq/db/biosql/DBHelper.html" title="class in org.biojava.bio.seq.db.biosql">DBHelper</A>&nbsp;helper,
           <A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A>&nbsp;taxon,
           java.lang.String&nbsp;nameClass,
           java.lang.String&nbsp;oldName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Deletes the specified name from of the taxon from the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#removeSequence(java.lang.String)">removeSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A></CODE></FONT></TD>
<TD><CODE><B>TaxonSQL.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/TaxonSQL.html#removeTaxon(java.sql.Connection, org.biojava.bio.seq.db.biosql.DBHelper, java.lang.String)">removeTaxon</A></B>(java.sql.Connection&nbsp;conn,
            <A HREF="../../../../org/biojava/bio/seq/db/biosql/DBHelper.html" title="class in org.biojava.bio.seq.db.biosql">DBHelper</A>&nbsp;helper,
            java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Deletes a taxon specified by one of it's names with all it's different
 names, annotations and sequences from the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A></CODE></FONT></TD>
<TD><CODE><B>TaxonSQL.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/TaxonSQL.html#removeTaxon(java.sql.Connection, int, org.biojava.bio.seq.db.biosql.DBHelper)">removeTaxon</A></B>(java.sql.Connection&nbsp;conn,
            int&nbsp;ncbi_id,
            <A HREF="../../../../org/biojava/bio/seq/db/biosql/DBHelper.html" title="class in org.biojava.bio.seq.db.biosql">DBHelper</A>&nbsp;helper)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Deletes the taxon given by it's NCBI-Taxon-ID from the database and returns the
 removed taxon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A></CODE></FONT></TD>
<TD><CODE><B>TaxonSQL.</B><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/TaxonSQL.html#setCommonName(java.sql.Connection, org.biojava.bio.taxa.Taxon, java.lang.String)">setCommonName</A></B>(java.sql.Connection&nbsp;conn,
              <A HREF="../../../../org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">Taxon</A>&nbsp;taxon,
              java.lang.String&nbsp;newName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;With this method the common name of the given taxon can be changed or created,
 if there was none before.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#BioSQLSequenceDB(javax.sql.DataSource, java.lang.String, boolean)">BioSQLSequenceDB</A></B>(javax.sql.DataSource&nbsp;ds,
                 java.lang.String&nbsp;biodatabase,
                 boolean&nbsp;create)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#BioSQLSequenceDB(java.lang.String, java.lang.String, java.lang.String, java.lang.String, boolean)">BioSQLSequenceDB</A></B>(java.lang.String&nbsp;dbURL,
                 java.lang.String&nbsp;dbUser,
                 java.lang.String&nbsp;dbPass,
                 java.lang.String&nbsp;biodatabase,
                 boolean&nbsp;create)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Connect to a BioSQL database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html#BioSQLSequenceDB(java.lang.String, java.lang.String, java.lang.String, java.lang.String, java.lang.String, boolean)">BioSQLSequenceDB</A></B>(java.lang.String&nbsp;dbDriver,
                 java.lang.String&nbsp;dbURL,
                 java.lang.String&nbsp;dbUser,
                 java.lang.String&nbsp;dbPass,
                 java.lang.String&nbsp;biodatabase,
                 boolean&nbsp;create)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Connect to a BioSQL database.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.flat"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/db/flat/package-summary.html">org.biojava.bio.seq.db.flat</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/db/flat/package-summary.html">org.biojava.bio.seq.db.flat</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>FlatSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/db/flat/FlatSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html" title="interface in org.biojava.bio.seq.db">SequenceDBLite</A></CODE></FONT></TD>
<TD><CODE><B>FlatSequenceDBProvider.</B><B><A HREF="../../../../org/biojava/bio/seq/db/flat/FlatSequenceDBProvider.html#getSequenceDB(java.util.Map)">getSequenceDB</A></B>(java.util.Map&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/seq/db/flat/package-summary.html">org.biojava.bio.seq.db.flat</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/db/flat/FlatSequenceDB.html#FlatSequenceDB(java.lang.String, java.lang.String)">FlatSequenceDB</A></B>(java.lang.String&nbsp;location,
               java.lang.String&nbsp;dbName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.distributed"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></CODE></FONT></TD>
<TD><CODE><B>DistDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html#getFeatures(org.biojava.bio.seq.FeatureFilter)">getFeatures</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A>&nbsp;ff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get all features matching a FeatureFilter provided by this DistDataSource.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></CODE></FONT></TD>
<TD><CODE><B>GFFDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/GFFDataSource.html#getFeatures(org.biojava.bio.seq.FeatureFilter)">getFeatures</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A>&nbsp;ff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/SequenceDBDataSource.html#getFeatures(org.biojava.bio.seq.FeatureFilter)">getFeatures</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A>&nbsp;ff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></CODE></FONT></TD>
<TD><CODE><B>DistDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html#getFeatures(java.lang.String, org.biojava.bio.seq.FeatureFilter, boolean)">getFeatures</A></B>(java.lang.String&nbsp;id,
            <A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A>&nbsp;ff,
            boolean&nbsp;recurse)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get all features matching a FeatureFilter on a Sequence with an ID and recurse flats.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></CODE></FONT></TD>
<TD><CODE><B>GFFDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/GFFDataSource.html#getFeatures(java.lang.String, org.biojava.bio.seq.FeatureFilter, boolean)">getFeatures</A></B>(java.lang.String&nbsp;id,
            <A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A>&nbsp;ff,
            boolean&nbsp;recurse)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/SequenceDBDataSource.html#getFeatures(java.lang.String, org.biojava.bio.seq.FeatureFilter, boolean)">getFeatures</A></B>(java.lang.String&nbsp;id,
            <A HREF="../../../../org/biojava/bio/seq/FeatureFilter.html" title="interface in org.biojava.bio.seq">FeatureFilter</A>&nbsp;ff,
            boolean&nbsp;recurse)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>DistDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get a Sequence object for an ID.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>GFFDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/GFFDataSource.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/SequenceDBDataSource.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>DistributedSequenceDB.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/DistributedSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>DistDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html#hasFeatures(java.lang.String)">hasFeatures</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Find out if this DistDataSource can provide features on a sequence with a particular ID.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GFFDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/GFFDataSource.html#hasFeatures(java.lang.String)">hasFeatures</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>SequenceDBDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/SequenceDBDataSource.html#hasFeatures(java.lang.String)">hasFeatures</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>DistDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html#hasSequence(java.lang.String)">hasSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Find out if this DistDataSource provides the sequence information for a sequence ID.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>GFFDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/GFFDataSource.html#hasSequence(java.lang.String)">hasSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>SequenceDBDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/SequenceDBDataSource.html#hasSequence(java.lang.String)">hasSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set</CODE></FONT></TD>
<TD><CODE><B>DistDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/DistDataSource.html#ids(boolean)">ids</A></B>(boolean&nbsp;topLevel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the complete set of sequence IDs provided by this DistDataSource.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set</CODE></FONT></TD>
<TD><CODE><B>GFFDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/GFFDataSource.html#ids(boolean)">ids</A></B>(boolean&nbsp;topLevel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Set</CODE></FONT></TD>
<TD><CODE><B>SequenceDBDataSource.</B><B><A HREF="../../../../org/biojava/bio/seq/distributed/SequenceDBDataSource.html#ids(boolean)">ids</A></B>(boolean&nbsp;topLevel)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.homol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleHomology.</B><B><A HREF="../../../../org/biojava/bio/seq/homol/SimpleHomology.html#setAlignment(org.biojava.bio.symbol.Alignment)">setAlignment</A></B>(<A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;alignment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>setAlignment</code> sets the alignment which describes
 the homology.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>DummySequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/DummySequence.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SubSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SubSequence.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleSequence.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleFeature.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleFeature.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>RevCompSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/RevCompSequence.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;ft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;createFeature() will call createFeature() on the underlying Sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ViewSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/ViewSequence.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleGappedSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleGappedSequence.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>LazyFilterFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/LazyFilterFeatureHolder.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleSequence.html#createFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;fh,
              <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>Please use new 1-arg createFeature instead.</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleRemoteFeature.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleRemoteFeature.html#getRemoteFeature()">getRemoteFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A></CODE></FONT></TD>
<TD><CODE><B>TemplateUtils.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/TemplateUtils.html#instantiateTemplate(org.biojava.bio.seq.Feature)">instantiateTemplate</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;feat)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This attempts to divine the 'best' template class for a feature and return
 a new instance readly for pupulating.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A></CODE></FONT></TD>
<TD><CODE><B>TemplateUtils.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/TemplateUtils.html#makeTemplate(org.biojava.bio.seq.Feature)">makeTemplate</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;feat)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>TemplateUtils.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/TemplateUtils.html#populate(org.biojava.bio.seq.Feature.Template, org.biojava.bio.seq.Feature)">populate</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ,
         <A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;feat)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This attempts to populate the fields of a template using
 the publically accessible information in a feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleSequence.html#realizeFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;parent,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleFeature.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleFeature.html#realizeFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;fh,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ViewSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/ViewSequence.html#realizeFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">realizeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/FeatureHolder.html" title="interface in org.biojava.bio.seq">FeatureHolder</A>&nbsp;parent,
               <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;template)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DummySequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/DummySequence.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;feature)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SubSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SubSequence.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleSequence.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove a feature attached to this sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RevCompSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/RevCompSequence.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleGappedSequence.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleGappedSequence.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>LazyFilterFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/LazyFilterFeatureHolder.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleRemoteFeature.DBResolver.</B><B><A HREF="../../../../org/biojava/bio/seq/impl/SimpleRemoteFeature.DBResolver.html#resolve(org.biojava.bio.seq.RemoteFeature)">resolve</A></B>(<A HREF="../../../../org/biojava/bio/seq/RemoteFeature.html" title="interface in org.biojava.bio.seq">RemoteFeature</A>&nbsp;rFeat)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/io/ParseException.html" title="class in org.biojava.bio.seq.io">ParseException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ParseException should be thrown to indicate that there was a problem with
 parsing sequence information.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/ComponentFeature.html" title="interface in org.biojava.bio.seq">ComponentFeature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleAssemblyBuilder.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SimpleAssemblyBuilder.html#addComponentSequence(org.biojava.bio.seq.ComponentFeature.Template)">addComponentSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/ComponentFeature.Template.html" title="class in org.biojava.bio.seq">ComponentFeature.Template</A>&nbsp;cft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#biojavaToFile(int, java.io.OutputStream, java.lang.Object)">biojavaToFile</A></B>(int&nbsp;fileType,
              java.io.OutputStream&nbsp;os,
              java.lang.Object&nbsp;biojava)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Converts a Biojava object to the given filetype.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#biojavaToFile(java.lang.String, java.lang.String, java.io.OutputStream, java.lang.Object)">biojavaToFile</A></B>(java.lang.String&nbsp;formatName,
              java.lang.String&nbsp;alphabetName,
              java.io.OutputStream&nbsp;os,
              java.lang.Object&nbsp;biojava)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a Biojava <code>SequenceIterator</code>,
 <code>SequenceDB</code>, <code>Sequence</code> or <code>Aligment</code>
 to an <code>OutputStream</code></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureTableParser.</B><B><A HREF="../../../../org/biojava/bio/seq/io/FeatureTableParser.html#endFeature()">endFeature</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureTableParser.</B><B><A HREF="../../../../org/biojava/bio/seq/io/FeatureTableParser.html#featureData(java.lang.String)">featureData</A></B>(java.lang.String&nbsp;line)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#fileToBiojava(int, java.io.BufferedReader)">fileToBiojava</A></B>(int&nbsp;fileType,
              java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads a file and returns the corresponding Biojava object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#fileToBiojava(java.lang.String, java.lang.String, java.io.BufferedReader)">fileToBiojava</A></B>(java.lang.String&nbsp;formatName,
              java.lang.String&nbsp;alphabetName,
              java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Reads a file with the specified format and alphabet</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#formatToFactory(org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.symbol.Alphabet)">formatToFactory</A></B>(<A HREF="../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A>&nbsp;format,
                <A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>as this essentially duplicates the operation
 available in the method <code>identifyBuilderFactory</code>.</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#getAlphabet(int)">getAlphabet</A></B>(int&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>getAlphabet</code> accepts a value which represents a
 sequence format and returns the relevant
 <code>FiniteAlphabet</code> object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SequenceBuilderFactory.html" title="interface in org.biojava.bio.seq.io">SequenceBuilderFactory</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#getBuilderFactory(int)">getBuilderFactory</A></B>(int&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>getBuilderFactory</code> accepts a value which represents
 a sequence format and returns the relevant
 <code>SequenceBuilderFactory</code> object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SequenceFormat.html" title="interface in org.biojava.bio.seq.io">SequenceFormat</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#getSequenceFormat(int)">getSequenceFormat</A></B>(int&nbsp;identifier)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>getSequenceFormat</code> accepts a value which represents
 a sequence format and returns the relevant
 <code>SequenceFormat</code> object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A></CODE></FONT></TD>
<TD><CODE><B>ChunkedSymbolListFactory.</B><B><A HREF="../../../../org/biojava/bio/seq/io/ChunkedSymbolListFactory.html#make(org.biojava.bio.seq.io.SymbolReader)">make</A></B>(<A HREF="../../../../org/biojava/bio/seq/io/SymbolReader.html" title="interface in org.biojava.bio.seq.io">SymbolReader</A>&nbsp;sr)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Method to create a Sequence with a SymbolReader.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderFilter.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SequenceBuilderFilter.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceBuilder.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SequenceBuilder.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the Sequence object which has been constructed
 by this builder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SmartSequenceBuilder.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SmartSequenceBuilder.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SimpleAssemblyBuilder.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SimpleAssemblyBuilder.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBSequenceBuilder.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SequenceDBSequenceBuilder.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;create the sequence</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SequenceBuilderBase.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SequenceBuilderBase.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SimpleSequenceBuilder.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SimpleSequenceBuilder.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>StreamReader.</B><B><A HREF="../../../../org/biojava/bio/seq/io/StreamReader.html#nextSequence()">nextSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Pull the next sequence out of the stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A></CODE></FONT></TD>
<TD><CODE><B>EmblLikeLocationParser.</B><B><A HREF="../../../../org/biojava/bio/seq/io/EmblLikeLocationParser.html#parseLocation(java.lang.String, org.biojava.bio.seq.Feature.Template)">parseLocation</A></B>(java.lang.String&nbsp;location,
              <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;theTemplate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<code>parseLocation</code> creates a <code>Location</code> from
 the String and returns a stranded location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDB.html" title="interface in org.biojava.bio.seq.db">SequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#readFasta(java.io.InputStream, org.biojava.bio.symbol.Alphabet)">readFasta</A></B>(java.io.InputStream&nbsp;seqFile,
          <A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Create a sequence database from a fasta file provided as an
 input stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>SequenceFormat.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SequenceFormat.html#readSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojava.bio.seq.io.SeqIOListener)">readSequence</A></B>(java.io.BufferedReader&nbsp;reader,
             <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
             <A HREF="../../../../org/biojava/bio/seq/io/SeqIOListener.html" title="interface in org.biojava.bio.seq.io">SeqIOListener</A>&nbsp;listener)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read a sequence and pass data on to a SeqIOListener.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FeatureTableParser.</B><B><A HREF="../../../../org/biojava/bio/seq/io/FeatureTableParser.html#startFeature(java.lang.String)">startFeature</A></B>(java.lang.String&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaAlignmentFormat.</B><B><A HREF="../../../../org/biojava/bio/seq/io/FastaAlignmentFormat.html#write(java.io.OutputStream, org.biojava.bio.symbol.Alignment, int)">write</A></B>(java.io.OutputStream&nbsp;os,
      <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align,
      int&nbsp;fileType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes out the alignment to an FASTA file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MSFAlignmentFormat.</B><B><A HREF="../../../../org/biojava/bio/seq/io/MSFAlignmentFormat.html#write(java.io.OutputStream, org.biojava.bio.symbol.Alignment, int)">write</A></B>(java.io.OutputStream&nbsp;os,
      <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align,
      int&nbsp;fileType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaAlignmentFormat.</B><B><A HREF="../../../../org/biojava/bio/seq/io/FastaAlignmentFormat.html#writeDna(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeDna</A></B>(java.io.OutputStream&nbsp;os,
         <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MSFAlignmentFormat.</B><B><A HREF="../../../../org/biojava/bio/seq/io/MSFAlignmentFormat.html#writeDna(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeDna</A></B>(java.io.OutputStream&nbsp;os,
         <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeGenpept(java.io.OutputStream, org.biojava.bio.seq.Sequence)">writeGenpept</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a single Sequence to an OutputStream in Genpept format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeGenpept(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)">writeGenpept</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a stream of Sequences to an OutputStream in Genpept
 format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaAlignmentFormat.</B><B><A HREF="../../../../org/biojava/bio/seq/io/FastaAlignmentFormat.html#writeProtein(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeProtein</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MSFAlignmentFormat.</B><B><A HREF="../../../../org/biojava/bio/seq/io/MSFAlignmentFormat.html#writeProtein(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeProtein</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeSwissprot(java.io.OutputStream, org.biojava.bio.seq.Sequence)">writeSwissprot</A></B>(java.io.OutputStream&nbsp;os,
               <A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a single Sequence to an OutputStream in SwissProt format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SeqIOTools.</B><B><A HREF="../../../../org/biojava/bio/seq/io/SeqIOTools.html#writeSwissprot(java.io.OutputStream, org.biojava.bio.seq.SequenceIterator)">writeSwissprot</A></B>(java.io.OutputStream&nbsp;os,
               <A HREF="../../../../org/biojava/bio/seq/SequenceIterator.html" title="interface in org.biojava.bio.seq">SequenceIterator</A>&nbsp;in)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Writes a stream of Sequences to an OutputStream in SwissProt
 format.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/biojava/bio/seq/io/CrossProductTokenization.html#CrossProductTokenization(org.biojava.bio.symbol.Alphabet)">CrossProductTokenization</A></B>(<A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.projection"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/seq/projection/package-summary.html">org.biojava.bio.seq.projection</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/seq/projection/package-summary.html">org.biojava.bio.seq.projection</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ProjectedFeature.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectedFeature.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;temp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ProjectedFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectedFeatureHolder.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;templ)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ReparentContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ReparentContext.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;projTempl)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ProjectionContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectionContext.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;projTempl)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a projected feature with properties matching the template.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ReparentContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ReparentContext.html#createFeature(org.biojava.bio.seq.Feature, org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f,
              <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;projTempl)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ProjectionContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectionContext.html#createFeature(org.biojava.bio.seq.Feature, org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;projParent,
              <A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;projTempl)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new projected feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProjectedFeature.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectedFeature.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProjectedFeatureHolder.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectedFeatureHolder.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;dyingChild)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ReparentContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ReparentContext.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;dyingChild)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProjectionContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectionContext.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;dyingChild)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove the dying child.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ReparentContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ReparentContext.html#removeFeature(org.biojava.bio.seq.Feature, org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f,
              <A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f2)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ProjectionContext.</B><B><A HREF="../../../../org/biojava/bio/seq/projection/ProjectionContext.html#removeFeature(org.biojava.bio.seq.Feature, org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;projParent,
              <A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;dyingChild)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove the dying child.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.structure"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/structure/package-summary.html">org.biojava.bio.structure</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/structure/package-summary.html">org.biojava.bio.structure</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/StructureException.html" title="class in org.biojava.bio.structure">StructureException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An exception during the parsing of a PDB file.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.structure.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/structure/io/package-summary.html">org.biojava.bio.structure.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/structure/io/package-summary.html">org.biojava.bio.structure.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/structure/io/PDBParseException.html" title="class in org.biojava.bio.structure.io">PDBParseException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An exception during the parsing of a PDB file.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/structure/io/package-summary.html">org.biojava.bio.structure.io</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/alignment/SubstitutionMatrix.html" title="class in org.biojava.bio.alignment">SubstitutionMatrix</A></CODE></FONT></TD>
<TD><CODE><B>SeqRes2AtomAligner.</B><B><A HREF="../../../../org/biojava/bio/structure/io/SeqRes2AtomAligner.html#getSubstitutionMatrix(org.biojava.bio.symbol.FiniteAlphabet)">getSubstitutionMatrix</A></B>(<A HREF="../../../../org/biojava/bio/symbol/FiniteAlphabet.html" title="interface in org.biojava.bio.symbol">FiniteAlphabet</A>&nbsp;alphabet)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.symbol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/symbol/IllegalAlphabetException.html" title="class in org.biojava.bio.symbol">IllegalAlphabetException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 The exception to indicate that an invalid alphabet has been used.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/symbol/IllegalSymbolException.html" title="class in org.biojava.bio.symbol">IllegalSymbolException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The exception to indicate that a symbol is not valid within a context.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/symbol/AlphabetIndex.html" title="interface in org.biojava.bio.symbol">AlphabetIndex</A></CODE></FONT></TD>
<TD><CODE><B>AlphabetManager.</B><B><A HREF="../../../../org/biojava/bio/symbol/AlphabetManager.html#getAlphabetIndex(org.biojava.bio.symbol.Symbol[])">getAlphabetIndex</A></B>(<A HREF="../../../../org/biojava/bio/symbol/Symbol.html" title="interface in org.biojava.bio.symbol">Symbol</A>[]&nbsp;syms)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get an indexer for an array of symbols.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>DoubleAlphabet.SubDoubleAlphabet.</B><B><A HREF="../../../../org/biojava/bio/symbol/DoubleAlphabet.SubDoubleAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>Alphabet.</B><B><A HREF="../../../../org/biojava/bio/symbol/Alphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Get a SymbolTokenization by name.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>AbstractAlphabet.</B><B><A HREF="../../../../org/biojava/bio/symbol/AbstractAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A></CODE></FONT></TD>
<TD><CODE><B>SoftMaskedAlphabet.</B><B><A HREF="../../../../org/biojava/bio/symbol/SoftMaskedAlphabet.html#getTokenization(java.lang.String)">getTokenization</A></B>(java.lang.String&nbsp;type)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AlphabetManager.</B><B><A HREF="../../../../org/biojava/bio/symbol/AlphabetManager.html#loadAlphabets(org.xml.sax.InputSource)">loadAlphabets</A></B>(org.xml.sax.InputSource&nbsp;is)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Load additional Alphabets, defined in XML format, into the AlphabetManager's registry.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/symbol/MergeLocation.html" title="class in org.biojava.bio.symbol">MergeLocation</A></CODE></FONT></TD>
<TD><CODE><B>MergeLocation.</B><B><A HREF="../../../../org/biojava/bio/symbol/MergeLocation.html#mergeLocations(java.util.List)">mergeLocations</A></B>(java.util.List&nbsp;componentLocations)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Static Factory method for getting an instance of a <code>MergeLocation</code></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/symbol/MergeLocation.html" title="class in org.biojava.bio.symbol">MergeLocation</A></CODE></FONT></TD>
<TD><CODE><B>MergeLocation.</B><B><A HREF="../../../../org/biojava/bio/symbol/MergeLocation.html#mergeLocations(org.biojava.bio.symbol.Location, org.biojava.bio.symbol.Location)">mergeLocations</A></B>(<A HREF="../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;locA,
               <A HREF="../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;locB)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CodonPrefFilter.</B><B><A HREF="../../../../org/biojava/bio/symbol/CodonPrefFilter.html#put(org.biojava.bio.symbol.CodonPref)">put</A></B>(<A HREF="../../../../org/biojava/bio/symbol/CodonPref.html" title="interface in org.biojava.bio.symbol">CodonPref</A>&nbsp;codonPref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;handles storage of a CodonPref object</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CodonPrefFilter.ByName.</B><B><A HREF="../../../../org/biojava/bio/symbol/CodonPrefFilter.ByName.html#put(org.biojava.bio.symbol.CodonPref)">put</A></B>(<A HREF="../../../../org/biojava/bio/symbol/CodonPref.html" title="interface in org.biojava.bio.symbol">CodonPref</A>&nbsp;codonPref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CodonPrefFilter.EverythingToXML.</B><B><A HREF="../../../../org/biojava/bio/symbol/CodonPrefFilter.EverythingToXML.html#put(org.biojava.bio.symbol.CodonPref)">put</A></B>(<A HREF="../../../../org/biojava/bio/symbol/CodonPref.html" title="interface in org.biojava.bio.symbol">CodonPref</A>&nbsp;codonPref)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/symbol/CodonPref.html" title="interface in org.biojava.bio.symbol">CodonPref</A>[]</CODE></FONT></TD>
<TD><CODE><B>CodonPrefTools.</B><B><A HREF="../../../../org/biojava/bio/symbol/CodonPrefTools.html#readFromXML(java.io.InputStream)">readFromXML</A></B>(java.io.InputStream&nbsp;prefStream)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CodonPrefTools.</B><B><A HREF="../../../../org/biojava/bio/symbol/CodonPrefTools.html#readFromXML(java.io.InputStream, org.biojava.bio.symbol.CodonPrefFilter)">readFromXML</A></B>(java.io.InputStream&nbsp;prefStream,
            <A HREF="../../../../org/biojava/bio/symbol/CodonPrefFilter.html" title="interface in org.biojava.bio.symbol">CodonPrefFilter</A>&nbsp;filter)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;read an CodonPref XML stream and handle it with a CodonPrefFilter object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojava/bio/symbol/CodonPref.html" title="interface in org.biojava.bio.symbol">CodonPref</A></CODE></FONT></TD>
<TD><CODE><B>CodonPrefTools.</B><B><A HREF="../../../../org/biojava/bio/symbol/CodonPrefTools.html#readFromXML(java.io.InputStream, java.lang.String)">readFromXML</A></B>(java.io.InputStream&nbsp;prefStream,
            java.lang.String&nbsp;name)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reads a specified CodonPref from an file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>CodonPrefTools.</B><B><A HREF="../../../../org/biojava/bio/symbol/CodonPrefTools.html#writeToXML(org.biojava.bio.symbol.CodonPref, java.io.PrintWriter)">writeToXML</A></B>(<A HREF="../../../../org/biojava/bio/symbol/CodonPref.html" title="interface in org.biojava.bio.symbol">CodonPref</A>&nbsp;codonPref,
           java.io.PrintWriter&nbsp;writer)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;write out a specified CodonPref object in XML format.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.taxa"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/taxa/package-summary.html">org.biojava.bio.taxa</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/bio/taxa/package-summary.html">org.biojava.bio.taxa</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/bio/taxa/CircularReferenceException.html" title="class in org.biojava.bio.taxa">CircularReferenceException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>replaced by classes in <A HREF="../../../../org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I></TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.directory"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/directory/package-summary.html">org.biojava.directory</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/directory/package-summary.html">org.biojava.directory</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html" title="interface in org.biojava.bio.seq.db">SequenceDBLite</A></CODE></FONT></TD>
<TD><CODE><B>Registry.</B><B><A HREF="../../../../org/biojava/directory/Registry.html#getDatabase(java.lang.String)">getDatabase</A></B>(java.lang.String&nbsp;dbName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getDatabase</code> retrieves a database instance known by
 a name <code>String</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/db/SequenceDBLite.html" title="interface in org.biojava.bio.seq.db">SequenceDBLite</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBProvider.</B><B><A HREF="../../../../org/biojava/directory/SequenceDBProvider.html#getSequenceDB(java.util.Map)">getSequenceDB</A></B>(java.util.Map&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get a sequence database.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.automata"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/utils/automata/package-summary.html">org.biojava.utils.automata</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/utils/automata/package-summary.html">org.biojava.utils.automata</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/utils/automata/AutomatonException.html" title="class in org.biojava.utils.automata">AutomatonException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An exception thrown by classes of this package.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.math"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/utils/math/package-summary.html">org.biojava.utils.math</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/utils/math/package-summary.html">org.biojava.utils.math</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;double</CODE></FONT></TD>
<TD><CODE><B>BinarySearch.</B><B><A HREF="../../../../org/biojava/utils/math/BinarySearch.html#solve(double, double, double, org.biojava.utils.math.ComputeObject)">solve</A></B>(double&nbsp;min,
      double&nbsp;max,
      double&nbsp;tolerance,
      <A HREF="../../../../org/biojava/utils/math/ComputeObject.html" title="interface in org.biojava.utils.math">ComputeObject</A>&nbsp;obj)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;method that will attempt solving the equation.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.regex"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/utils/regex/package-summary.html">org.biojava.utils.regex</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/utils/regex/package-summary.html">org.biojava.utils.regex</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/biojava/utils/regex/RegexException.html" title="class in org.biojava.utils.regex">RegexException</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An exception thrown by classes of this package.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.walker"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojava/utils/walker/package-summary.html">org.biojava.utils.walker</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojava/utils/walker/package-summary.html">org.biojava.utils.walker</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/utils/walker/Walker.html" title="interface in org.biojava.utils.walker">Walker</A></CODE></FONT></TD>
<TD><CODE><B>WalkerFactory.</B><B><A HREF="../../../../org/biojava/utils/walker/WalkerFactory.html#getWalker(org.biojava.utils.walker.Visitor)">getWalker</A></B>(<A HREF="../../../../org/biojava/utils/walker/Visitor.html" title="interface in org.biojava.utils.walker">Visitor</A>&nbsp;visitor)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get a Walker that is customosed to a particular visitor.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojavax/bio/package-summary.html">org.biojavax.bio</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojavax/bio/package-summary.html">org.biojavax.bio</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A></CODE></FONT></TD>
<TD><CODE><B>BioEntryIterator.</B><B><A HREF="../../../../org/biojavax/bio/BioEntryIterator.html#nextBioEntry()">nextBioEntry</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#addBioEntry(org.biojavax.bio.BioEntry)">addBioEntry</A></B>(<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HashBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashBioEntryDB.html#addBioEntry(org.biojavax.bio.BioEntry)">addBioEntry</A></B>(<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a BioEntry, the name of the BioEntry will be used as the ID</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractBioEntryDB.html#addBioEntry(org.biojavax.bio.BioEntry)">addBioEntry</A></B>(<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioEntryDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/BioEntryDBLite.html#addBioEntry(org.biojavax.bio.BioEntry)">addBioEntry</A></B>(<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds a sequence to the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HashBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashBioEntryDB.html#addBioEntry(java.lang.String, org.biojavax.bio.BioEntry)">addBioEntry</A></B>(java.lang.String&nbsp;id,
            <A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HashRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashRichSequenceDB.html#addRichSequence(org.biojavax.bio.seq.RichSequence)">addRichSequence</A></B>(<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#addRichSequence(org.biojavax.bio.seq.RichSequence)">addRichSequence</A></B>(<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequenceDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/RichSequenceDBLite.html#addRichSequence(org.biojavax.bio.seq.RichSequence)">addRichSequence</A></B>(<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds a sequence to the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HashRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashRichSequenceDB.html#addRichSequence(java.lang.String, org.biojavax.bio.seq.RichSequence)">addRichSequence</A></B>(java.lang.String&nbsp;id,
                <A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#addSequence(org.biojava.bio.seq.Sequence)">addSequence</A></B>(<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A></CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#getBioEntry(java.lang.String)">getBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A></CODE></FONT></TD>
<TD><CODE><B>HashBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashBioEntryDB.html#getBioEntry(java.lang.String)">getBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A></CODE></FONT></TD>
<TD><CODE><B>BioEntryDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/BioEntryDBLite.html#getBioEntry(java.lang.String)">getBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve a single BioEntry by its id.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#getBioEntrys(java.util.Set)">getBioEntrys</A></B>(java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>HashBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashBioEntryDB.html#getBioEntrys(java.util.Set)">getBioEntrys</A></B>(java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>BioEntryDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/BioEntryDBLite.html#getBioEntrys(java.util.Set)">getBioEntrys</A></B>(java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve multiple BioEntry by their ids.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#getBioEntrys(java.util.Set, org.biojavax.bio.db.BioEntryDB)">getBioEntrys</A></B>(java.util.Set&nbsp;ids,
             <A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>HashBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashBioEntryDB.html#getBioEntrys(java.util.Set, org.biojavax.bio.db.BioEntryDB)">getBioEntrys</A></B>(java.util.Set&nbsp;ids,
             <A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>BioEntryDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/BioEntryDBLite.html#getBioEntrys(java.util.Set, org.biojavax.bio.db.BioEntryDB)">getBioEntrys</A></B>(java.util.Set&nbsp;ids,
             <A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve multiple BioEntry into a specific sequence database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>HashRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashRichSequenceDB.html#getRichSequence(java.lang.String)">getRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/RichSequenceDBLite.html#getRichSequence(java.lang.String)">getRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve a single RichSequence by its id.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>HashRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashRichSequenceDB.html#getRichSequences(java.util.Set)">getRichSequences</A></B>(java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/RichSequenceDBLite.html#getRichSequences(java.util.Set)">getRichSequences</A></B>(java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve multiple RichSequence by its id.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>HashRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashRichSequenceDB.html#getRichSequences(java.util.Set, org.biojavax.bio.db.RichSequenceDB)">getRichSequences</A></B>(java.util.Set&nbsp;ids,
                 <A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/RichSequenceDBLite.html#getRichSequences(java.util.Set, org.biojavax.bio.db.RichSequenceDB)">getRichSequences</A></B>(java.util.Set&nbsp;ids,
                 <A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve multiple RichSequence into a specific sequence database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#getSequence(java.lang.String)">getSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#removeBioEntry(java.lang.String)">removeBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HashBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashBioEntryDB.html#removeBioEntry(java.lang.String)">removeBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractBioEntryDB.html#removeBioEntry(java.lang.String)">removeBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioEntryDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/BioEntryDBLite.html#removeBioEntry(java.lang.String)">removeBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove the BioEntry associated with an ID from the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#removeRichSequence(java.lang.String)">removeRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RichSequenceDBLite.</B><B><A HREF="../../../../org/biojavax/bio/db/RichSequenceDBLite.html#removeRichSequence(java.lang.String)">removeRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove the RichSequence associated with an ID from the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HashRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/HashRichSequenceDB.html#removeSequence(java.lang.String)">removeSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/AbstractRichSequenceDB.html#removeSequence(java.lang.String)">removeSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLBioEntryDB.html#_addBioEntry(org.biojavax.bio.BioEntry)">_addBioEntry</A></B>(<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLBioEntryDB.html#addBioEntry(org.biojavax.bio.BioEntry)">addBioEntry</A></B>(<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLRichSequenceDB.html#addRichSequence(org.biojavax.bio.seq.RichSequence)">addRichSequence</A></B>(<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A>&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLRichSequenceDB.html#fullyLoadRichSequence(org.biojavax.bio.seq.RichSequence)">fullyLoadRichSequence</A></B>(<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A>&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLBioEntryDB.html#getBioEntry(java.lang.String)">getBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLBioEntryDB.html#getBioEntrys(java.util.Set)">getBioEntrys</A></B>(java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLBioEntryDB.html#getBioEntrys(java.util.Set, org.biojavax.bio.db.BioEntryDB)">getBioEntrys</A></B>(java.util.Set&nbsp;ids,
             <A HREF="../../../../org/biojavax/bio/db/BioEntryDB.html" title="interface in org.biojavax.bio.db">BioEntryDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLRichSequenceDB.html#getRichSequence(java.lang.String)">getRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLRichSequenceDB.html#getRichSequences(java.util.Set)">getRichSequences</A></B>(java.util.Set&nbsp;ids)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>BioSQLRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLRichSequenceDB.html#getRichSequences(java.util.Set, org.biojavax.bio.db.RichSequenceDB)">getRichSequences</A></B>(java.util.Set&nbsp;ids,
                 <A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A>&nbsp;db)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLBioEntryDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLBioEntryDB.html#removeBioEntry(java.lang.String)">removeBioEntry</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>BioSQLRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/biosql/BioSQLRichSequenceDB.html#removeRichSequence(java.lang.String)">removeRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.ncbi"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>GenbankRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html#getRichSequence(java.lang.String)">getRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given the appropriate Genbank ID, return the matching RichSequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>GenpeptRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html#getRichSequence(java.lang.String)">getRichSequence</A></B>(java.lang.String&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given the appropriate Genbank ID, return the matching RichSequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>GenbankRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html#getRichSequence(java.lang.String, org.biojavax.Namespace)">getRichSequence</A></B>(java.lang.String&nbsp;id,
                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;nsp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given the appropriate Genbank ID, return the matching RichSequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>GenpeptRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html#getRichSequence(java.lang.String, org.biojavax.Namespace)">getRichSequence</A></B>(java.lang.String&nbsp;id,
                <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;nsp)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given the appropriate Genbank ID, return the matching RichSequence object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenbankRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html#getRichSequences(java.util.Set)">getRichSequences</A></B>(java.util.Set&nbsp;list)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve rich sequences from a Genbank</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenpeptRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html#getRichSequences(java.util.Set)">getRichSequences</A></B>(java.util.Set&nbsp;list)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve rich sequences from a Genbank</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenbankRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenbankRichSequenceDB.html#getRichSequences(java.util.Set, org.biojavax.bio.db.RichSequenceDB)">getRichSequences</A></B>(java.util.Set&nbsp;list,
                 <A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A>&nbsp;database)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve rich sequences from a Genbank</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A></CODE></FONT></TD>
<TD><CODE><B>GenpeptRichSequenceDB.</B><B><A HREF="../../../../org/biojavax/bio/db/ncbi/GenpeptRichSequenceDB.html#getRichSequences(java.util.Set, org.biojavax.bio.db.RichSequenceDB)">getRichSequences</A></B>(java.util.Set&nbsp;list,
                 <A HREF="../../../../org/biojavax/bio/db/RichSequenceDB.html" title="interface in org.biojavax.bio.db">RichSequenceDB</A>&nbsp;database)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve rich sequences from a Genbank</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>SimpleRichFeature.</B><B><A HREF="../../../../org/biojavax/bio/seq/SimpleRichFeature.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;ft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new Feature, and add it to this FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A></CODE></FONT></TD>
<TD><CODE><B>ThinRichSequence.</B><B><A HREF="../../../../org/biojavax/bio/seq/ThinRichSequence.html#createFeature(org.biojava.bio.seq.Feature.Template)">createFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.Template.html" title="class in org.biojava.bio.seq">Feature.Template</A>&nbsp;ft)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new Feature, and add it to this FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.Tools.</B><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.Tools.html#createRichSequence(org.biojavax.Namespace, java.lang.String, java.lang.String, org.biojava.bio.symbol.Alphabet)">createRichSequence</A></B>(<A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns,
                   java.lang.String&nbsp;name,
                   java.lang.String&nbsp;seqString,
                   <A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new RichSequence in the specified namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.Tools.</B><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.Tools.html#createRichSequence(java.lang.String, java.lang.String, org.biojava.bio.symbol.Alphabet)">createRichSequence</A></B>(java.lang.String&nbsp;name,
                   java.lang.String&nbsp;seqString,
                   <A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new RichSequence in the default namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequence.Tools.</B><B><A HREF="../../../../org/biojavax/bio/seq/RichSequence.Tools.html#createRichSequence(java.lang.String, java.lang.String, java.lang.String, org.biojava.bio.symbol.Alphabet)">createRichSequence</A></B>(java.lang.String&nbsp;namespace,
                   java.lang.String&nbsp;name,
                   java.lang.String&nbsp;seqString,
                   <A HREF="../../../../org/biojava/bio/symbol/Alphabet.html" title="interface in org.biojava.bio.symbol">Alphabet</A>&nbsp;alpha)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create a new RichSequence in the specified namespace.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceIterator.</B><B><A HREF="../../../../org/biojavax/bio/seq/RichSequenceIterator.html#nextRichSequence()">nextRichSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichFeature.</B><B><A HREF="../../../../org/biojavax/bio/seq/SimpleRichFeature.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove a feature from this FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ThinRichSequence.</B><B><A HREF="../../../../org/biojavax/bio/seq/ThinRichSequence.html#removeFeature(org.biojava.bio.seq.Feature)">removeFeature</A></B>(<A HREF="../../../../org/biojava/bio/seq/Feature.html" title="interface in org.biojava.bio.seq">Feature</A>&nbsp;f)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Remove a feature from this FeatureHolder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequence.</B><B><A HREF="../../../../org/biojavax/bio/seq/SimpleRichSequence.html#setAlphabetName(java.lang.String)">setAlphabetName</A></B>(java.lang.String&nbsp;alphaname)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ThinRichSequence.</B><B><A HREF="../../../../org/biojavax/bio/seq/ThinRichSequence.html#setAlphabetName(java.lang.String)">setAlphabetName</A></B>(java.lang.String&nbsp;alphaname)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequence.</B><B><A HREF="../../../../org/biojavax/bio/seq/SimpleRichSequence.html#setStringSequence(java.lang.String)">setStringSequence</A></B>(java.lang.String&nbsp;seq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A> in <A HREF="../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A> that throw <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio">BioException</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceBuilder.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilder.html#makeRichSequence()">makeRichSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Build a RichSequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#makeRichSequence()">makeRichSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Build a RichSequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>RichSequenceBuilder.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceBuilder.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the Sequence object which has been constructed
 by this builder.
 Implementations of this for a RichSequenceBuilder should
 delegate to makeRichSequence() and return only RichSequence objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>SimpleRichSequenceBuilder.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/SimpleRichSequenceBuilder.html#makeSequence()">makeSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the Sequence object which has been constructed
 by this builder.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/BioEntry.html" title="interface in org.biojavax.bio">BioEntry</A></CODE></FONT></TD>
<TD><CODE><B>RichStreamReader.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/RichStreamReader.html#nextBioEntry()">nextBioEntry</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojavax/bio/seq/RichSequence.html" title="interface in org.biojavax.bio.seq">RichSequence</A></CODE></FONT></TD>
<TD><CODE><B>RichStreamReader.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/RichStreamReader.html#nextRichSequence()">nextRichSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/biojava/bio/seq/Sequence.html" title="interface in org.biojava.bio.seq">Sequence</A></CODE></FONT></TD>
<TD><CODE><B>RichStreamReader.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/RichStreamReader.html#nextSequence()">nextSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the next sequence in the iterator.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>RichSequenceFormat.</B><B><A HREF="../../../../org/biojavax/bio/seq/io/RichSequenceFormat.html#readRichSequence(java.io.BufferedReader, org.biojava.bio.seq.io.SymbolTokenization, org.biojavax.bio.seq.io.RichSeqIOListener, org.biojavax.Namespace)">readRichSequence</A></B>(java.io.BufferedReader&nbsp;reader,
                 <A HREF="../../../../org/biojava/bio/seq/io/SymbolTokenization.html" title="interface in org.biojava.bio.seq.io">SymbolTokenization</A>&nbsp;symParser,
                 <A HREF="../../../../org/biojavax/bio/seq/io/RichSeqIOListener.html" title="interface in org.biojavax.bio.seq.io">RichSeqIOListener</A>&nbsp;listener,
                 <A HREF="../../../../org/biojavax/Namespace.html" title="interface in org.biojavax">Namespace</A>&nbsp;ns)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads a sequence from the given buffered reader using the given tokenizer to parse
 sequence symbols.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../org/biojava/bio/BioException.html" title="class in org.biojava.bio"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?org/biojava/bio//class-useBioException.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="BioException.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
