<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:51:07 BST 2008 -->
<TITLE>
Uses of Package org.biojava.bio.seq.db (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Package org.biojava.bio.seq.db (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/seq/db/package-use.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Package<br>org.biojava.bio.seq.db</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.alignment"><B>org.biojava.bio.alignment</B></A></TD>
<TD>
Classes to generate and describe sequence alignments.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp"><B>org.biojava.bio.dp</B></A></TD>
<TD>HMM and Dynamic Programming Algorithms.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.gui"><B>org.biojava.bio.gui</B></A></TD>
<TD>Graphical interfaces for biojava objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.das"><B>org.biojava.bio.program.das</B></A></TD>
<TD>Development client for the Distributed Annotation System.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff"><B>org.biojava.bio.program.gff</B></A></TD>
<TD>GFF manipulation.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssaha"><B>org.biojava.bio.program.ssaha</B></A></TD>
<TD>SSAHA sequence searching API.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssbind"><B>org.biojava.bio.program.ssbind</B></A></TD>
<TD>
Creation of objects from SAX events using the BioJava BlastLikeDataSetCollection DTD.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.unigene"><B>org.biojava.bio.program.unigene</B></A></TD>
<TD>Objects for representing Unigene clusters.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.proteomics.aaindex"><B>org.biojava.bio.proteomics.aaindex</B></A></TD>
<TD>Classes and interfaces to load Amino Acid Index database files.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.search"><B>org.biojava.bio.search</B></A></TD>
<TD>
Interfaces and classes for representing sequence similarity search results.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq"><B>org.biojava.bio.seq</B></A></TD>
<TD>Classes and interfaces for defining biological sequences and informatics
objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db"><B>org.biojava.bio.seq.db</B></A></TD>
<TD>
Collections of biological sequence data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biofetch"><B>org.biojava.bio.seq.db.biofetch</B></A></TD>
<TD>Client for the OBDA BioFetch protocol.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biosql"><B>org.biojava.bio.seq.db.biosql</B></A></TD>
<TD>General purpose Sequence storage in a relational database.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.flat"><B>org.biojava.bio.seq.db.flat</B></A></TD>
<TD>
Support for OBDA flatfile databases.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.distributed"><B>org.biojava.bio.seq.distributed</B></A></TD>
<TD>Sequences and SequenceDBs which are composed from data taken
from a number of data sources.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.impl"><B>org.biojava.bio.seq.impl</B></A></TD>
<TD>Standard in-memory implementations of <code>Sequence</code> and
<code>Feature</code>.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.directory"><B>org.biojava.directory</B></A></TD>
<TD>Open Bio Sequence Database Access (OBDA) registry support.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db"><B>org.biojavax.bio.db</B></A></TD>
<TD>Interactions between biojavax objects and a DB.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.biosql"><B>org.biojavax.bio.db.biosql</B></A></TD>
<TD>Interface between biojava and biosql databases&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.ncbi"><B>org.biojavax.bio.db.ncbi</B></A></TD>
<TD>Interfaces to NCBI data.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.alignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.alignment"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.dp"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.gui"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/gui/package-summary.html">org.biojava.bio.gui</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.gui"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.das"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.program.das"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.program.das"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojava.bio.program.das"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.program.gff"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.program.gff"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.program.phred"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssaha"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.program.ssaha"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssbind"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.program.ssbind"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.program.ssbind"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBInstallation.html#org.biojava.bio.program.ssbind"><B>SequenceDBInstallation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A SequenceDBInstallation has the functionality of a factory for
 SequenceDB objects and additionally manages the SequenceDB objects
 created by itself such that the minimum number of SequenceDB
 objects is created by a particular SequenceDBInstallation
 object.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.unigene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.program.unigene"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.proteomics.aaindex"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/proteomics/aaindex/package-summary.html">org.biojava.bio.proteomics.aaindex</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.proteomics.aaindex"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.search"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.search"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.seq"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/AbstractSequenceDB.html#org.biojava.bio.seq.db"><B>AbstractSequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of SequenceDB that provides the sequenceIterator
 method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IDMaker.html#org.biojava.bio.seq.db"><B>IDMaker</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interface for objects that define how to make an ID for a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.seq.db"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/Index.html#org.biojava.bio.seq.db"><B>Index</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This defines an index entry for an individual sequence within a set of
 indexed files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IndexStore.html#org.biojava.bio.seq.db"><B>IndexStore</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This defines the objects that IndexedSequenceDB uses to store all of the
 database state, such as name, format, sequence builder and the actual file
 offsets.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.seq.db"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBInstallation.html#org.biojava.bio.seq.db"><B>SequenceDBInstallation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 A SequenceDBInstallation has the functionality of a factory for
 SequenceDB objects and additionally manages the SequenceDB objects
 created by itself such that the minimum number of SequenceDB
 objects is created by a particular SequenceDBInstallation
 object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojava.bio.seq.db"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBWrapper.html#org.biojava.bio.seq.db"><B>SequenceDBWrapper</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of SequenceDB that wraps up another database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/TabIndexStore.html#org.biojava.bio.seq.db"><B>TabIndexStore</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Implements IndexStore as a serialized file for the java data and a
 tab-delimited file of offsets.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/WebSequenceDB.html#org.biojava.bio.seq.db"><B>WebSequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Functions for access to a web based database that returns sequences
 in a variety of formats.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biofetch"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/db/biofetch/package-summary.html">org.biojava.bio.seq.db.biofetch</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.seq.db.biofetch"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojava.bio.seq.db.biofetch"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.seq.db.biosql"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojava.bio.seq.db.biosql"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.flat"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/db/flat/package-summary.html">org.biojava.bio.seq.db.flat</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.seq.db.flat"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojava.bio.seq.db.flat"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.distributed"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/AbstractSequenceDB.html#org.biojava.bio.seq.distributed"><B>AbstractSequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract implementation of SequenceDB that provides the sequenceIterator
 method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojava.bio.seq.distributed"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.seq.distributed"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojava.bio.seq.distributed"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.seq.impl"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojava.bio.seq.io"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.directory"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojava/directory/package-summary.html">org.biojava.directory</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojava.directory"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IDMaker.html#org.biojavax.bio.db"><B>IDMaker</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Interface for objects that define how to make an ID for a sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojavax.bio.db"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojavax.bio.db"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojavax.bio.db"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojavax.bio.db.biosql"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojavax.bio.db.biosql"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojavax.bio.db.biosql"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.ncbi"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A> used by <A HREF="../../../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/IllegalIDException.html#org.biojavax.bio.db.ncbi"><B>IllegalIDException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDB.html#org.biojavax.bio.db.ncbi"><B>SequenceDB</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences with accessible keys and iterators over all
 sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../../../org/biojava/bio/seq/db/class-use/SequenceDBLite.html#org.biojavax.bio.db.ncbi"><B>SequenceDBLite</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A database of sequences.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/seq/db/package-use.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
