<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:50:59 BST 2008 -->
<TITLE>
Uses of Interface org.biojava.bio.symbol.Alignment (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Interface org.biojava.bio.symbol.Alignment (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/symbol//class-useAlignment.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Alignment.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Interface<br>org.biojava.bio.symbol.Alignment</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.alignment"><B>org.biojava.bio.alignment</B></A></TD>
<TD>
Classes to generate and describe sequence alignments.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.chromatogram"><B>org.biojava.bio.chromatogram</B></A></TD>
<TD>
Interfaces and classes for chromatogram data, as produced by DNA sequencing equipment.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dist"><B>org.biojava.bio.dist</B></A></TD>
<TD>
Probability distributions over Alphabets.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp"><B>org.biojava.bio.dp</B></A></TD>
<TD>HMM and Dynamic Programming Algorithms.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.abi"><B>org.biojava.bio.program.abi</B></A></TD>
<TD>ABI Trace Handling.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.search"><B>org.biojava.bio.search</B></A></TD>
<TD>
Interfaces and classes for representing sequence similarity search results.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.homol"><B>org.biojava.bio.seq.homol</B></A></TD>
<TD>The classes and interfaces for defining sequence similarity and
honology.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.impl"><B>org.biojava.bio.seq.impl</B></A></TD>
<TD>Standard in-memory implementations of <code>Sequence</code> and
<code>Feature</code>.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.symbol"><B>org.biojava.bio.symbol</B></A></TD>
<TD>Representation of the Symbols that make up a sequence, and locations within
them.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.phylo.io.phylip"><B>org.biojavax.bio.phylo.io.phylip</B></A></TD>
<TD>Classes to support the reading and writing of PHYLIP format.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.alignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subinterfaces of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;interface</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/alignment/ARAlignment.html" title="interface in org.biojava.bio.alignment">ARAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ARAlignment is an interface that defines methods for adding and
 removing seqeunces from an Alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;interface</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/alignment/UnequalLengthAlignment.html" title="interface in org.biojava.bio.alignment">UnequalLengthAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;UnequalLengthAlignment has the following behavior.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A> that implement <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/alignment/AbstractULAlignment.html" title="class in org.biojava.bio.alignment">AbstractULAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/alignment/AbstractULAlignment.SubULAlignment.html" title="class in org.biojava.bio.alignment">AbstractULAlignment.SubULAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/alignment/FlexibleAlignment.html" title="class in org.biojava.bio.alignment">FlexibleAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;FlexibleAlignment is a class which implements
 UnequalLengthAlignment, ARAlignment and EditableAlignment <b>It
 places no restriction on where any sequence can be in the alignment
 so there could be gaps in the alignment.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Fields in <A HREF="../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A> declared as <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>NeedlemanWunsch.</B><B><A HREF="../../../../../org/biojava/bio/alignment/NeedlemanWunsch.html#pairalign">pairalign</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The result of a successfull alignment</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>abstract &nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SequenceAlignment.</B><B><A HREF="../../../../../org/biojava/bio/alignment/SequenceAlignment.html#getAlignment(org.biojava.bio.symbol.SymbolList, org.biojava.bio.symbol.SymbolList)">getAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;query,
             <A HREF="../../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;subject)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This method also performs a sequence alignment of the two given sequences
 but it returns an Alignment object instead of the score.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>NeedlemanWunsch.</B><B><A HREF="../../../../../org/biojava/bio/alignment/NeedlemanWunsch.html#getAlignment(org.biojava.bio.symbol.SymbolList, org.biojava.bio.symbol.SymbolList)">getAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;query,
             <A HREF="../../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;target)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This method is good if one wants to reuse the alignment calculated by this
 class in another BioJava class.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>AbstractULAlignment.</B><B><A HREF="../../../../../org/biojava/bio/alignment/AbstractULAlignment.html#subAlignment(java.util.Set, int, int)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             int&nbsp;min,
             int&nbsp;max)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retreives a subAlignment</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>AbstractULAlignment.</B><B><A HREF="../../../../../org/biojava/bio/alignment/AbstractULAlignment.html#subAlignment(java.util.Set, org.biojava.bio.symbol.Location)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             <A HREF="../../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a subalignment specified by the location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>AbstractULAlignment.SubULAlignment.</B><B><A HREF="../../../../../org/biojava/bio/alignment/AbstractULAlignment.SubULAlignment.html#subAlignment(java.util.Set, org.biojava.bio.symbol.Location)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             <A HREF="../../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.chromatogram"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/chromatogram/package-summary.html">org.biojava.bio.chromatogram</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/chromatogram/package-summary.html">org.biojava.bio.chromatogram</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>AbstractChromatogram.</B><B><A HREF="../../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#createImmutableAlignment(java.util.Map)">createImmutableAlignment</A></B>(java.util.Map&nbsp;labelsToSymLists)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A factory method for creating new immutable alignments, particularly
 for use as base call alignments.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>AbstractChromatogram.</B><B><A HREF="../../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#getBaseCalls()">getBaseCalls</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return the total number of base calls.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>Chromatogram.</B><B><A HREF="../../../../../org/biojava/bio/chromatogram/Chromatogram.html#getBaseCalls()">getBaseCalls</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns an alignment that describes the base calls for this chromatogram.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>AbstractChromatogram.</B><B><A HREF="../../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#reverseComplementBaseCalls()">reverseComplementBaseCalls</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a new base call alignment that is the reverse complement of
 one in this chromatogram.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/chromatogram/package-summary.html">org.biojava.bio.chromatogram</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AbstractChromatogram.</B><B><A HREF="../../../../../org/biojava/bio/chromatogram/AbstractChromatogram.html#setBaseCallAlignment(org.biojava.bio.symbol.Alignment)">setBaseCallAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Provides the list of base calls.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dist"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/dist/package-summary.html">org.biojava.bio.dist</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/dist/package-summary.html">org.biojava.bio.dist</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>[]</CODE></FONT></TD>
<TD><CODE><B>DistributionTools.</B><B><A HREF="../../../../../org/biojava/bio/dist/DistributionTools.html#distOverAlignment(org.biojava.bio.symbol.Alignment)">distOverAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;a)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Equivalent to distOverAlignment(a, false, 0.0).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>[]</CODE></FONT></TD>
<TD><CODE><B>DistributionTools.</B><B><A HREF="../../../../../org/biojava/bio/dist/DistributionTools.html#distOverAlignment(org.biojava.bio.symbol.Alignment, boolean)">distOverAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;a,
                  boolean&nbsp;countGaps)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates an array of distributions, one for each column of the alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>[]</CODE></FONT></TD>
<TD><CODE><B>DistributionTools.</B><B><A HREF="../../../../../org/biojava/bio/dist/DistributionTools.html#distOverAlignment(org.biojava.bio.symbol.Alignment, boolean, double)">distOverAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;a,
                  boolean&nbsp;countGaps,
                  double&nbsp;nullWeight)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates an array of distributions, one for each column of the alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A></CODE></FONT></TD>
<TD><CODE><B>DistributionTools.</B><B><A HREF="../../../../../org/biojava/bio/dist/DistributionTools.html#jointDistOverAlignment(org.biojava.bio.symbol.Alignment, boolean, double, int[])">jointDistOverAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;a,
                       boolean&nbsp;countGaps,
                       double&nbsp;nullWeight,
                       int[]&nbsp;cols)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a joint distribution.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subinterfaces of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;interface</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/dp/StatePath.html" title="interface in org.biojava.bio.dp">StatePath</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Extends the Alignment interface so that it is explicitly used to represent
 a state path through an HMM, and the associated emitted sequence and
 likelihoods.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A> that implement <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/dp/SimpleStatePath.html" title="class in org.biojava.bio.dp">SimpleStatePath</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A no-frills implementation of StatePath.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimpleStatePath.</B><B><A HREF="../../../../../org/biojava/bio/dp/SimpleStatePath.html#subAlignment(java.util.Set, org.biojava.bio.symbol.Location)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             <A HREF="../../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.abi"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/program/abi/package-summary.html">org.biojava.bio.program.abi</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/program/abi/package-summary.html">org.biojava.bio.program.abi</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>ABITools.</B><B><A HREF="../../../../../org/biojava/bio/program/abi/ABITools.html#getAlignment(org.biojava.bio.symbol.SymbolList)">getAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;abiSeq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 View a symbol list over the QUALITY alphabet as an alignment.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/dist/Distribution.html" title="interface in org.biojava.bio.dist">Distribution</A>[]</CODE></FONT></TD>
<TD><CODE><B>PhredTools.</B><B><A HREF="../../../../../org/biojava/bio/program/phred/PhredTools.html#phredAlignmentToDistArray(org.biojava.bio.symbol.Alignment)">phredAlignmentToDistArray</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;a)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;converts an Alignment of PhredSequences to a Distribution[] where each position is the average
 distribution of the underlying column of the alignment.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.search"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SequenceDBSearchSubHit.</B><B><A HREF="../../../../../org/biojava/bio/search/SequenceDBSearchSubHit.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimpleSeqSimilaritySearchSubHit.</B><B><A HREF="../../../../../org/biojava/bio/search/SimpleSeqSimilaritySearchSubHit.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SeqSimilaritySearchSubHit.</B><B><A HREF="../../../../../org/biojava/bio/search/SeqSimilaritySearchSubHit.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return an alignment of (possibly part of) the query sequence
 against (possibly part of) this hit sequence.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/search/SequenceDBSearchSubHit.html#SequenceDBSearchSubHit(double, double, double, int, int, org.biojava.bio.seq.StrandedFeature.Strand, int, int, org.biojava.bio.seq.StrandedFeature.Strand, org.biojava.bio.symbol.Alignment, org.biojava.bio.Annotation)">SequenceDBSearchSubHit</A></B>(double&nbsp;score,
                       double&nbsp;eValue,
                       double&nbsp;pValue,
                       int&nbsp;queryStart,
                       int&nbsp;queryEnd,
                       <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.Strand.html" title="class in org.biojava.bio.seq">StrandedFeature.Strand</A>&nbsp;queryStrand,
                       int&nbsp;subjectStart,
                       int&nbsp;subjectEnd,
                       <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.Strand.html" title="class in org.biojava.bio.seq">StrandedFeature.Strand</A>&nbsp;subjectStrand,
                       <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;alignment,
                       <A HREF="../../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A>&nbsp;annotation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Creates a new <code>SequenceDBSearchSubHit</code> object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/search/SimpleSeqSimilaritySearchSubHit.html#SimpleSeqSimilaritySearchSubHit(double, double, double, int, int, org.biojava.bio.seq.StrandedFeature.Strand, int, int, org.biojava.bio.seq.StrandedFeature.Strand, org.biojava.bio.symbol.Alignment, org.biojava.bio.Annotation)">SimpleSeqSimilaritySearchSubHit</A></B>(double&nbsp;score,
                                double&nbsp;eValue,
                                double&nbsp;pValue,
                                int&nbsp;queryStart,
                                int&nbsp;queryEnd,
                                <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.Strand.html" title="class in org.biojava.bio.seq">StrandedFeature.Strand</A>&nbsp;queryStrand,
                                int&nbsp;subjectStart,
                                int&nbsp;subjectEnd,
                                <A HREF="../../../../../org/biojava/bio/seq/StrandedFeature.Strand.html" title="class in org.biojava.bio.seq">StrandedFeature.Strand</A>&nbsp;subjectStrand,
                                <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;alignment,
                                <A HREF="../../../../../org/biojava/bio/Annotation.html" title="interface in org.biojava.bio">Annotation</A>&nbsp;annotation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new <code>SimpleSeqSimilaritySearchSubHit</code>
 object.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.homol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A> that implement <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.EmptyPairwiseAlignment.html" title="class in org.biojava.bio.seq.homol">SimilarityPairFeature.EmptyPairwiseAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>EmptyPairwiseAlignment</code> empty pairwise alignment
 which has labels to empty symbol lists.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Fields in <A HREF="../../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A> declared as <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimilarityPairFeature.Template.</B><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.Template.html#alignment">alignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>alignment</code> <code>Alignment</code> field.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimilarityPairFeature.</B><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#EMPTY_PAIRWISE">EMPTY_PAIRWISE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constant <code>EMPTY_PAIRWISE</code> is an empty alignment for
 situations where there is no available alignment data or the
 implementation does not want to create one.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimilarityPairFeature.</B><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getAlignment</code> returns the <code>Alignment</code> of
 two similar features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>Homology.</B><B><A HREF="../../../../../org/biojava/bio/seq/homol/Homology.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve the Alignment that specifies how the homologous regions are
 aligned.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimpleHomology.</B><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimpleHomology.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getAlignment</code> returns the alignment, which uses the
 <code>HomologyFeature</code>s as keys.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimilarityPairFeature.EmptyPairwiseAlignment.</B><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimilarityPairFeature.EmptyPairwiseAlignment.html#subAlignment(java.util.Set, org.biojava.bio.symbol.Location)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             <A HREF="../../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleHomology.</B><B><A HREF="../../../../../org/biojava/bio/seq/homol/SimpleHomology.html#setAlignment(org.biojava.bio.symbol.Alignment)">setAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;alignment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>setAlignment</code> sets the alignment which describes
 the homology.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimpleSimilarityPairFeature.</B><B><A HREF="../../../../../org/biojava/bio/seq/impl/SimpleSimilarityPairFeature.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>getAlignment</code> returns the alignment between the two
 features.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>FastaAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/FastaAlignmentFormat.html#read(java.io.BufferedReader)">read</A></B>(java.io.BufferedReader&nbsp;br)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads an alignment in FASTA format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>MSFAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/MSFAlignmentFormat.html#read(java.io.BufferedReader)">read</A></B>(java.io.BufferedReader&nbsp;reader)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads an MSF Alignment File</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>AlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/AlignmentFormat.html#read(java.io.BufferedReader)">read</A></B>(java.io.BufferedReader&nbsp;reader)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Read in an alignment from a buffered reader object</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/FastaAlignmentFormat.html#write(java.io.OutputStream, org.biojava.bio.symbol.Alignment, int)">write</A></B>(java.io.OutputStream&nbsp;os,
      <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align,
      int&nbsp;fileType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes out the alignment to an FASTA file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MSFAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/MSFAlignmentFormat.html#write(java.io.OutputStream, org.biojava.bio.symbol.Alignment, int)">write</A></B>(java.io.OutputStream&nbsp;os,
      <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align,
      int&nbsp;fileType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/FastaAlignmentFormat.html#writeDna(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeDna</A></B>(java.io.OutputStream&nbsp;os,
         <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MSFAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/MSFAlignmentFormat.html#writeDna(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeDna</A></B>(java.io.OutputStream&nbsp;os,
         <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FastaAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/FastaAlignmentFormat.html#writeProtein(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeProtein</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MSFAlignmentFormat.</B><B><A HREF="../../../../../org/biojava/bio/seq/io/MSFAlignmentFormat.html#writeProtein(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeProtein</A></B>(java.io.OutputStream&nbsp;os,
             <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.symbol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Classes in <A HREF="../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> that implement <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/symbol/RelabeledAlignment.html" title="class in org.biojava.bio.symbol">RelabeledAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An alignment that relabels another alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/symbol/SimpleAlignment.html" title="class in org.biojava.bio.symbol">SimpleAlignment</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A simple implementation of an Alignment.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SymbolListViews.</B><B><A HREF="../../../../../org/biojava/bio/symbol/SymbolListViews.html#alignment(java.util.List, org.biojava.bio.symbol.SymbolList)">alignment</A></B>(java.util.List&nbsp;labels,
          <A HREF="../../../../../org/biojava/bio/symbol/SymbolList.html" title="interface in org.biojava.bio.symbol">SymbolList</A>&nbsp;symList)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;View a SymbolList over a cross-product Alphabet as an Alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SymbolListViews.</B><B><A HREF="../../../../../org/biojava/bio/symbol/SymbolListViews.html#alignment(java.util.Map)">alignment</A></B>(java.util.Map&nbsp;labelToSymList)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Construct an alignment of the SymbolLists contained in the values collection
 of <code>labelToSymList</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>SimpleAlignment.</B><B><A HREF="../../../../../org/biojava/bio/symbol/SimpleAlignment.html#subAlignment(java.util.Set, org.biojava.bio.symbol.Location)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             <A HREF="../../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>Alignment.</B><B><A HREF="../../../../../org/biojava/bio/symbol/Alignment.html#subAlignment(java.util.Set, org.biojava.bio.symbol.Location)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             <A HREF="../../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Make a view onto this alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>RelabeledAlignment.</B><B><A HREF="../../../../../org/biojava/bio/symbol/RelabeledAlignment.html#subAlignment(java.util.Set, org.biojava.bio.symbol.Location)">subAlignment</A></B>(java.util.Set&nbsp;labels,
             <A HREF="../../../../../org/biojava/bio/symbol/Location.html" title="interface in org.biojava.bio.symbol">Location</A>&nbsp;loc)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Constructors in <A HREF="../../../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/symbol/Alignment.SymbolListIterator.html#Alignment.SymbolListIterator(org.biojava.bio.symbol.Alignment)">Alignment.SymbolListIterator</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;ali)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../org/biojava/bio/symbol/RelabeledAlignment.html#RelabeledAlignment(org.biojava.bio.symbol.Alignment)">RelabeledAlignment</A></B>(<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;delegate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.phylo.io.phylip"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A> in <A HREF="../../../../../org/biojavax/bio/phylo/io/phylip/package-summary.html">org.biojavax.bio.phylo.io.phylip</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojavax/bio/phylo/io/phylip/package-summary.html">org.biojavax.bio.phylo.io.phylip</A> that return <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileBuilder.</B><B><A HREF="../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileBuilder.html#getAlignment()">getAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../org/biojavax/bio/phylo/io/phylip/package-summary.html">org.biojavax.bio.phylo.io.phylip</A> with parameters of type <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileFormat.</B><B><A HREF="../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileFormat.html#writeFile(java.io.File, org.biojava.bio.symbol.Alignment)">writeFile</A></B>(java.io.File&nbsp;file,
          <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;alignment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the given Alignment in PHYLIP format to a file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileFormat.</B><B><A HREF="../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileFormat.html#writeStream(java.io.OutputStream, org.biojava.bio.symbol.Alignment)">writeStream</A></B>(java.io.OutputStream&nbsp;os,
            <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;alignment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the given Alignment in PHYLIP format to a stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PHYLIPFileFormat.</B><B><A HREF="../../../../../org/biojavax/bio/phylo/io/phylip/PHYLIPFileFormat.html#writeWriter(java.io.Writer, org.biojava.bio.symbol.Alignment)">writeWriter</A></B>(java.io.Writer&nbsp;writer,
            <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol">Alignment</A>&nbsp;alignment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Writes the given Alignment in PHYLIP format to a writer.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../org/biojava/bio/symbol/Alignment.html" title="interface in org.biojava.bio.symbol"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../index.html?org/biojava/bio/symbol//class-useAlignment.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Alignment.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
