<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:51:06 BST 2008 -->
<TITLE>
Uses of Package org.biojava.bio (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Package org.biojava.bio (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/biojava/bio/package-use.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Package<br>org.biojava.bio</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio"><B>org.biojava.bio</B></A></TD>
<TD>The core classes that will be used throughout the bio packages.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.alignment"><B>org.biojava.bio.alignment</B></A></TD>
<TD>
Classes to generate and describe sequence alignments.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.annodb"><B>org.biojava.bio.annodb</B></A></TD>
<TD>
Databases of generic structured data (annotation) objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.chromatogram"><B>org.biojava.bio.chromatogram</B></A></TD>
<TD>
Interfaces and classes for chromatogram data, as produced by DNA sequencing equipment.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dist"><B>org.biojava.bio.dist</B></A></TD>
<TD>
Probability distributions over Alphabets.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp"><B>org.biojava.bio.dp</B></A></TD>
<TD>HMM and Dynamic Programming Algorithms.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp.onehead"><B>org.biojava.bio.dp.onehead</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.dp.twohead"><B>org.biojava.bio.dp.twohead</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.gui"><B>org.biojava.bio.gui</B></A></TD>
<TD>Graphical interfaces for biojava objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.molbio"><B>org.biojava.bio.molbio</B></A></TD>
<TD>The classes and interfaces in this package cover common molecular
biological techniques such as restriction digests and PCR.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.abi"><B>org.biojava.bio.program.abi</B></A></TD>
<TD>ABI Trace Handling.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.das"><B>org.biojava.bio.program.das</B></A></TD>
<TD>Development client for the Distributed Annotation System.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.das.dasalignment"><B>org.biojava.bio.program.das.dasalignment</B></A></TD>
<TD>
request alignments via DAS&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.formats"><B>org.biojava.bio.program.formats</B></A></TD>
<TD>
Experimental parsers using the tagvalue framework.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff"><B>org.biojava.bio.program.gff</B></A></TD>
<TD>GFF manipulation.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.gff3"><B>org.biojava.bio.program.gff3</B></A></TD>
<TD>
Support for reading and writing GFF3.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.hmmer"><B>org.biojava.bio.program.hmmer</B></A></TD>
<TD>
Tools for working with profile Hidden Markov Models from the HMMer package.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.homologene"><B>org.biojava.bio.program.homologene</B></A></TD>
<TD>
Support classes for Homologene data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.indexdb"><B>org.biojava.bio.program.indexdb</B></A></TD>
<TD>A flat-file ascii index of ascii flat files as per the OBDA specification.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.phred"><B>org.biojava.bio.program.phred</B></A></TD>
<TD>Parser for Phred output&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.sax.blastxml"><B>org.biojava.bio.program.sax.blastxml</B></A></TD>
<TD>
Parsers for the XML output from NCBI blast.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssaha"><B>org.biojava.bio.program.ssaha</B></A></TD>
<TD>SSAHA sequence searching API.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.ssbind"><B>org.biojava.bio.program.ssbind</B></A></TD>
<TD>
Creation of objects from SAX events using the BioJava BlastLikeDataSetCollection DTD.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.tagvalue"><B>org.biojava.bio.program.tagvalue</B></A></TD>
<TD> Process files as streams of records, each with tags with values.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.unigene"><B>org.biojava.bio.program.unigene</B></A></TD>
<TD>Objects for representing Unigene clusters.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.program.xff"><B>org.biojava.bio.program.xff</B></A></TD>
<TD>Event-driven parsing system for the Extensible Feature Format (XFF).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.proteomics"><B>org.biojava.bio.proteomics</B></A></TD>
<TD>Utilities to aid in performing various physical analysis of proteins.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.proteomics.aaindex"><B>org.biojava.bio.proteomics.aaindex</B></A></TD>
<TD>Classes and interfaces to load Amino Acid Index database files.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.search"><B>org.biojava.bio.search</B></A></TD>
<TD>
Interfaces and classes for representing sequence similarity search results.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq"><B>org.biojava.bio.seq</B></A></TD>
<TD>Classes and interfaces for defining biological sequences and informatics
objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db"><B>org.biojava.bio.seq.db</B></A></TD>
<TD>
Collections of biological sequence data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biofetch"><B>org.biojava.bio.seq.db.biofetch</B></A></TD>
<TD>Client for the OBDA BioFetch protocol.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.biosql"><B>org.biojava.bio.seq.db.biosql</B></A></TD>
<TD>General purpose Sequence storage in a relational database.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.db.flat"><B>org.biojava.bio.seq.db.flat</B></A></TD>
<TD>
Support for OBDA flatfile databases.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.distributed"><B>org.biojava.bio.seq.distributed</B></A></TD>
<TD>Sequences and SequenceDBs which are composed from data taken
from a number of data sources.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.homol"><B>org.biojava.bio.seq.homol</B></A></TD>
<TD>The classes and interfaces for defining sequence similarity and
honology.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.impl"><B>org.biojava.bio.seq.impl</B></A></TD>
<TD>Standard in-memory implementations of <code>Sequence</code> and
<code>Feature</code>.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io"><B>org.biojava.bio.seq.io</B></A></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io.agave"><B>org.biojava.bio.seq.io.agave</B></A></TD>
<TD>Classes for converting between AGAVE XML and BioJava objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.io.filterxml"><B>org.biojava.bio.seq.io.filterxml</B></A></TD>
<TD>Tools for reading and writing an XML representation of BioJava's FeatureFilter language.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.seq.projection"><B>org.biojava.bio.seq.projection</B></A></TD>
<TD>Code for projecting Feature objects and systematically altering their
properties.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.structure"><B>org.biojava.bio.structure</B></A></TD>
<TD>
Interfaces and classes for protein structure (PDB).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.structure.io"><B>org.biojava.bio.structure.io</B></A></TD>
<TD>Input and Output of Structures&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.symbol"><B>org.biojava.bio.symbol</B></A></TD>
<TD>Representation of the Symbols that make up a sequence, and locations within
them.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.bio.taxa"><B>org.biojava.bio.taxa</B></A></TD>
<TD>Taxonomy object for representing species information.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.directory"><B>org.biojava.directory</B></A></TD>
<TD>Open Bio Sequence Database Access (OBDA) registry support.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.ontology"><B>org.biojava.ontology</B></A></TD>
<TD>
A general-purpose API for ontologies.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.automata"><B>org.biojava.utils.automata</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.math"><B>org.biojava.utils.math</B></A></TD>
<TD>
Mathematical utility classes.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.regex"><B>org.biojava.utils.regex</B></A></TD>
<TD>This package is used to perform regular expression searches of
SymbolLists defined in arbitrary Alphabets.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojava.utils.walker"><B>org.biojava.utils.walker</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax"><B>org.biojavax</B></A></TD>
<TD>The Biojava extensions packages, classes that extend the core biojava 
     functionality&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio"><B>org.biojavax.bio</B></A></TD>
<TD>Classes to represent biological entities and their relationships.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db"><B>org.biojavax.bio.db</B></A></TD>
<TD>Interactions between biojavax objects and a DB.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.biosql"><B>org.biojavax.bio.db.biosql</B></A></TD>
<TD>Interface between biojava and biosql databases&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.db.ncbi"><B>org.biojavax.bio.db.ncbi</B></A></TD>
<TD>Interfaces to NCBI data.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq"><B>org.biojavax.bio.seq</B></A></TD>
<TD>Rich implementations of Sequences, Locations and Features.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.bio.seq.io"><B>org.biojavax.bio.seq.io</B></A></TD>
<TD>Classes to support the I/O of RichSequence and 
     Bioentry objects.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.biojavax.ontology"><B>org.biojavax.ontology</B></A></TD>
<TD>Extensions to the biojava ontology model that represent BioSQL ontology.&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AbstractAnnotation.html#org.biojava.bio"><B>AbstractAnnotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A utility class to ease the problem of implementing an Annotation to that of
 providing an apropreate implementation of Map.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.Abstract.html#org.biojava.bio"><B>AnnotationType.Abstract</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An abstract base class useful for implementing AnnotationType
 instances.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/CollectionConstraint.html#org.biojava.bio"><B>CollectionConstraint</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Used by <code>AnnotationType</code> to represent the constraint on
 the collection of values in a property-slot.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/EcNumber.html#org.biojava.bio"><B>EcNumber</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;An ec (enzyme classification) number.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/EcNumber.Impl.html#org.biojava.bio"><B>EcNumber.Impl</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A simple implementation of EcNumber.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/PropertyConstraint.html#org.biojava.bio"><B>PropertyConstraint</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>PropertyConstraint</code>s describes a constraint applied
 to the members of an annotation bundle.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.alignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioError.html#org.biojava.bio.alignment"><B>BioError</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological error.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.alignment"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioRuntimeException.html#org.biojava.bio.alignment"><B>BioRuntimeException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.annodb"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/annodb/package-summary.html">org.biojava.bio.annodb</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.annodb"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.annodb"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.chromatogram"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/chromatogram/package-summary.html">org.biojava.bio.chromatogram</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.chromatogram"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dist"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/dist/package-summary.html">org.biojava.bio.dist</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.dist"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.dp"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.dp"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.dp"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp.onehead"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/dp/onehead/package-summary.html">org.biojava.bio.dp.onehead</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.dp.onehead"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.dp.twohead"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/dp/twohead/package-summary.html">org.biojava.bio.dp.twohead</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.dp.twohead"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.gui"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/gui/package-summary.html">org.biojava.bio.gui</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.gui"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.molbio"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/molbio/package-summary.html">org.biojava.bio.molbio</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.molbio"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.molbio"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.molbio"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.abi"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/abi/package-summary.html">org.biojava.bio.program.abi</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioError.html#org.biojava.bio.program.abi"><B>BioError</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological error.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.das"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.das"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.das"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.das.dasalignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/das/dasalignment/package-summary.html">org.biojava.bio.program.das.dasalignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.das.dasalignment"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.program.das.dasalignment"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.das.dasalignment"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.formats"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/formats/package-summary.html">org.biojava.bio.program.formats</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.program.formats"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.formats"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.gff"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.gff3"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/gff3/package-summary.html">org.biojava.bio.program.gff3</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.program.gff3"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.gff3"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.gff3"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.hmmer"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/hmmer/package-summary.html">org.biojava.bio.program.hmmer</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.program.hmmer"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.hmmer"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.homologene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/homologene/package-summary.html">org.biojava.bio.program.homologene</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.homologene"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.indexdb"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/indexdb/package-summary.html">org.biojava.bio.program.indexdb</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.indexdb"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.program.indexdb"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.indexdb"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.phred"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.program.phred"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.phred"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.phred"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.sax.blastxml"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/sax/blastxml/package-summary.html">org.biojava.bio.program.sax.blastxml</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.sax.blastxml"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssaha"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.ssaha"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.ssbind"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.ssbind"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.ssbind"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.tagvalue"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/tagvalue/package-summary.html">org.biojava.bio.program.tagvalue</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.tagvalue"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.program.tagvalue"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.unigene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.program.unigene"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.program.unigene"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.unigene"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.program.xff"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/program/xff/package-summary.html">org.biojava.bio.program.xff</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.program.xff"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.program.xff"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.proteomics"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/proteomics/package-summary.html">org.biojava.bio.proteomics</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.proteomics"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.proteomics.aaindex"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/proteomics/aaindex/package-summary.html">org.biojava.bio.proteomics.aaindex</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.proteomics.aaindex"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.search"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.search"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.search"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.search"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.seq"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.seq"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.seq"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.db"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioRuntimeException.html#org.biojava.bio.seq.db"><B>BioRuntimeException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biofetch"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/db/biofetch/package-summary.html">org.biojava.bio.seq.db.biofetch</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.db.biofetch"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.db.biosql"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioRuntimeException.html#org.biojava.bio.seq.db.biosql"><B>BioRuntimeException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.db.flat"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/db/flat/package-summary.html">org.biojava.bio.seq.db.flat</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.db.flat"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.distributed"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.distributed"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.homol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.seq.homol"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.homol"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.seq.impl"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.seq.impl"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.impl"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AbstractAnnotation.html#org.biojava.bio.seq.io"><B>AbstractAnnotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A utility class to ease the problem of implementing an Annotation to that of
 providing an apropreate implementation of Map.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.seq.io"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.seq.io"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.io"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io.agave"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/io/agave/package-summary.html">org.biojava.bio.seq.io.agave</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.seq.io.agave"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.seq.io.agave"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/SimpleAnnotation.html#org.biojava.bio.seq.io.agave"><B>SimpleAnnotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A no-frills implementation of Annotation that is just a wrapper
 around a <code>LinkedHashMap</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.io.filterxml"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/io/filterxml/package-summary.html">org.biojava.bio.seq.io.filterxml</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/AnnotationType.html#org.biojava.bio.seq.io.filterxml"><B>AnnotationType</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A set of constraints on the data contained in an <code>Annotation</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/CollectionConstraint.html#org.biojava.bio.seq.io.filterxml"><B>CollectionConstraint</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Used by <code>AnnotationType</code> to represent the constraint on
 the collection of values in a property-slot.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/PropertyConstraint.html#org.biojava.bio.seq.io.filterxml"><B>PropertyConstraint</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>PropertyConstraint</code>s describes a constraint applied
 to the members of an annotation bundle.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.seq.projection"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/seq/projection/package-summary.html">org.biojava.bio.seq.projection</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.seq.projection"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.seq.projection"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.structure"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/structure/package-summary.html">org.biojava.bio.structure</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.structure"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.structure"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.structure.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/structure/io/package-summary.html">org.biojava.bio.structure.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.structure.io"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.symbol"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.symbol"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.symbol"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.symbol"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.bio.taxa"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/bio/taxa/package-summary.html">org.biojava.bio.taxa</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.bio.taxa"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.bio.taxa"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.bio.taxa"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.directory"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/directory/package-summary.html">org.biojava.directory</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.directory"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.ontology"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/ontology/package-summary.html">org.biojava.ontology</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojava.ontology"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojava.ontology"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.automata"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/utils/automata/package-summary.html">org.biojava.utils.automata</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.utils.automata"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.math"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/utils/math/package-summary.html">org.biojava.utils.math</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.utils.math"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.regex"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/utils/regex/package-summary.html">org.biojava.utils.regex</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.utils.regex"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojava.utils.walker"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojava/utils/walker/package-summary.html">org.biojava.utils.walker</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojava.utils.walker"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/package-summary.html">org.biojavax</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojavax"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojavax"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/bio/package-summary.html">org.biojavax.bio</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojavax.bio"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojavax.bio"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojavax.bio"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojavax.bio.db"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.biosql"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojavax.bio.db.biosql"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.db.ncbi"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojavax.bio.db.ncbi"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojavax.bio.seq"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojavax.bio.seq"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojavax.bio.seq"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.bio.seq.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/BioException.html#org.biojavax.bio.seq.io"><B>BioException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A nestable biological exception.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.biojavax.ontology"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Classes in <A HREF="../../../org/biojava/bio/package-summary.html">org.biojava.bio</A> used by <A HREF="../../../org/biojavax/ontology/package-summary.html">org.biojavax.ontology</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotatable.html#org.biojavax.ontology"><B>Annotatable</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Indicates that an object has an associated annotation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../../org/biojava/bio/class-use/Annotation.html#org.biojavax.ontology"><B>Annotation</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Arbitrary annotation associated with one or more objects.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/biojava/bio/package-use.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
