<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:51:09 BST 2008 -->
<TITLE>
Deprecated List (BioJava-live API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Deprecated List (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Use</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Deprecated</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="index.html?deprecated-list.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="deprecated-list.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Deprecated API</B></H2>
</CENTER>
<HR SIZE="4" NOSHADE>
<B>Contents</B><UL>
<LI><A HREF="#interface">Deprecated Interfaces</A>
<LI><A HREF="#class">Deprecated Classes</A>
<LI><A HREF="#exception">Deprecated Exceptions</A>
<LI><A HREF="#field">Deprecated Fields</A>
<LI><A HREF="#method">Deprecated Methods</A>
<LI><A HREF="#constructor">Deprecated Constructors</A>
</UL>

<A NAME="interface"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Deprecated Interfaces</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SeqFileFormer.html" title="interface in org.biojava.bio.seq.io">org.biojava.bio.seq.io.SeqFileFormer</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/SequenceFactory.html" title="interface in org.biojava.bio.seq">org.biojava.bio.seq.SequenceFactory</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use org.biojavax.bio.seq.io.RichSequenceBuilder or 
 use org.biojavax.bio.seq.io.SequenceBuilder</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/Taxon.html" title="interface in org.biojava.bio.taxa">org.biojava.bio.taxa.Taxon</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/TaxonFactory.html" title="interface in org.biojava.bio.taxa">org.biojava.bio.taxa.TaxonFactory</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/TaxonParser.html" title="interface in org.biojava.bio.taxa">org.biojava.bio.taxa.TaxonParser</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="class"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Deprecated Classes</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/AbstractTaxon.html" title="class in org.biojava.bio.taxa">org.biojava.bio.taxa.AbstractTaxon</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/Annotatable.AnnotationForwarder.html" title="class in org.biojava.bio">org.biojava.bio.Annotatable.AnnotationForwarder</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use
   <code>new ChangeForwarder.Retyper(source, cs, Annotation.PROPERTY)</code>
   instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/BioSQLSequenceDB.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.BioSQLSequenceDB</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/BioSQLSequenceDBProvider.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.BioSQLSequenceDBProvider</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/DBHelper.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.DBHelper</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/dist/Distribution.NullModelForwarder.html" title="class in org.biojava.bio.dist">org.biojava.bio.dist.Distribution.NullModelForwarder</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use
    <code>new ChangeForwarder.Retyper(this, cs, Annotation.PROPERTY)</code>
    instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/EbiFormat.html" title="class in org.biojava.bio.taxa">org.biojava.bio.taxa.EbiFormat</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/EmblFileFormer.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.EmblFileFormer</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/EmblLikeFormat.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.EmblLikeFormat</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io.EMBLFormat instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/EmblLikeLocationParser.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.EmblLikeLocationParser</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/EmblProcessor.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.EmblProcessor</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/FastaDescriptionLineParser.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.FastaDescriptionLineParser</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io.FastaFormat</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/FastaFormat.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.FastaFormat</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io.FastaFormat</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/FeatureTableParser.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.FeatureTableParser</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenbankFileFormer.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.GenbankFileFormer</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenbankFormat.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.GenbankFormat</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io.GenbankFormat</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenbankProcessor.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.GenbankProcessor</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenbankXmlFormat.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.GenbankXmlFormat</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io.INSDseqFormat</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenEmblFeatureComparator.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.GenEmblFeatureComparator</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenEmblPropertyComparator.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.GenEmblPropertyComparator</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenpeptFormat.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.GenpeptFormat</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/HypersonicDBHelper.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.HypersonicDBHelper</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/stats/svm/LinearKernel.html" title="class in org.biojava.stats.svm">org.biojava.stats.svm.LinearKernel</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Just use SparseVector.kernel instead...</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/MySQLDBHelper.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.MySQLDBHelper</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/OracleDBHelper.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.OracleDBHelper</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/OrganismParser.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.OrganismParser</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.taxa framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/utils/ProcessTools.html" title="class in org.biojava.utils">org.biojava.utils.ProcessTools</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>preferable to use org.biojava.utils.ExecRunner 
 or the org.biojava.utils.process package.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/ProteinRefSeqFileFormer.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.ProteinRefSeqFileFormer</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/ProteinRefSeqProcessor.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.ProteinRefSeqProcessor</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/ReferenceAnnotation.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.ReferenceAnnotation</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SeqIOEventEmitter.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.SeqIOEventEmitter</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SeqIOTools.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.SeqIOTools</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use org.biojavax.bio.seq.RichSequence.IOTools</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/search/SequenceDBSearchHit.html" title="class in org.biojava.bio.search">org.biojava.bio.search.SequenceDBSearchHit</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>SimpleSeqSimilaritySearchHit has been made Annotatable
 and is now functionally identical.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/search/SequenceDBSearchResult.html" title="class in org.biojava.bio.search">org.biojava.bio.search.SequenceDBSearchResult</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>SimpleSeqSimilaritySearchResult has been made
 Annotatable and is now functionally identical.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/search/SequenceDBSearchSubHit.html" title="class in org.biojava.bio.search">org.biojava.bio.search.SequenceDBSearchSubHit</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>SimpleSeqSimilaritySearchSubHit has been made
 Annotatable and is now functionally identical.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/gui/sequence/SequencePoster.html" title="class in org.biojava.bio.gui.sequence">org.biojava.bio.gui.sequence.SequencePoster</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>This doesn't handle loads of stuff. Use SequencePoster.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/dist/SimpleDistributionTrainer.html" title="class in org.biojava.bio.dist">org.biojava.bio.dist.SimpleDistributionTrainer</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Distribution impls should be providing custom trainers.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/SimpleTaxon.html" title="class in org.biojava.bio.taxa">org.biojava.bio.taxa.SimpleTaxon</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/SimpleTaxonFactory.html" title="class in org.biojava.bio.taxa">org.biojava.bio.taxa.SimpleTaxonFactory</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SwissprotFileFormer.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.SwissprotFileFormer</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SwissprotProcessor.html" title="class in org.biojava.bio.seq.io">org.biojava.bio.seq.io.SwissprotProcessor</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use org.biojavax.bio.seq.io framework instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/TaxonSQL.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.TaxonSQL</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/db/biosql/UnknownDBHelper.html" title="class in org.biojava.bio.seq.db.biosql">org.biojava.bio.seq.db.biosql.UnknownDBHelper</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use hibernate and org.biojavax.bio.db.*</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/WeakTaxon.html" title="class in org.biojava.bio.taxa">org.biojava.bio.taxa.WeakTaxon</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/WeakTaxonFactory.html" title="class in org.biojava.bio.taxa">org.biojava.bio.taxa.WeakTaxonFactory</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="exception"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Deprecated Exceptions</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/taxa/CircularReferenceException.html" title="class in org.biojava.bio.taxa">org.biojava.bio.taxa.CircularReferenceException</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>replaced by classes in <A HREF="org/biojavax/bio/taxa/package-summary.html"><CODE>org.biojavax.bio.taxa</CODE></A></I>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="field"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Deprecated Fields</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/gff/GFFRecord.html#NO_FRAME">org.biojava.bio.program.gff.GFFRecord.NO_FRAME</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use GFFTools.NO_FRAME instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/gff/GFFRecord.html#NO_SCORE">org.biojava.bio.program.gff.GFFRecord.NO_SCORE</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use GFFTools.NO_SCORE instead</I>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="method"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Deprecated Methods</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/utils/Changeable.html#addChangeListener(org.biojava.utils.ChangeListener)">org.biojava.utils.Changeable.addChangeListener(ChangeListener)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use addChangeListener(cl, ChangeType.UNKNOWN)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/impl/SimpleSequence.html#createFeature(org.biojava.bio.seq.FeatureHolder, org.biojava.bio.seq.Feature.Template)">org.biojava.bio.seq.impl.SimpleSequence.createFeature(FeatureHolder, Feature.Template)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Please use new 1-arg createFeature instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/symbol/AlphabetManager.html#createSymbol(char, org.biojava.bio.Annotation, java.util.List, org.biojava.bio.symbol.Alphabet)">org.biojava.bio.symbol.AlphabetManager.createSymbol(char, Annotation, List, Alphabet)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use the new version, without the token argument</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/symbol/AlphabetManager.html#createSymbol(char, org.biojava.bio.Annotation, java.util.Set, org.biojava.bio.symbol.Alphabet)">org.biojava.bio.symbol.AlphabetManager.createSymbol(char, Annotation, Set, Alphabet)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use the three-arg version of this method instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/symbol/AlphabetManager.html#createSymbol(char, java.lang.String, org.biojava.bio.Annotation)">org.biojava.bio.symbol.AlphabetManager.createSymbol(char, String, Annotation)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use the two-arg version of this method instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SeqIOTools.html#formatToFactory(org.biojava.bio.seq.io.SequenceFormat, org.biojava.bio.symbol.Alphabet)">org.biojava.bio.seq.io.SeqIOTools.formatToFactory(SequenceFormat, Alphabet)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>as this essentially duplicates the operation
 available in the method <code>identifyBuilderFactory</code>.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/dist/DistributionTools.html#generateOrderNSequence(java.lang.String, org.biojava.bio.dist.OrderNDistribution, int)">org.biojava.bio.dist.DistributionTools.generateOrderNSequence(String, OrderNDistribution, int)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use generateSequence() or generateSymbolList() instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/phred/PhredFormat.html#getDefaultFormat()">org.biojava.bio.program.phred.PhredFormat.getDefaultFormat()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I></I>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/EmblLikeFormat.html#getDefaultFormat()">org.biojava.bio.seq.io.EmblLikeFormat.getDefaultFormat()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I></I>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GAMEFormat.html#getDefaultFormat()">org.biojava.bio.seq.io.GAMEFormat.getDefaultFormat()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I></I>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SequenceFormat.html#getDefaultFormat()">org.biojava.bio.seq.io.SequenceFormat.getDefaultFormat()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>new implementations should only write a single
 format.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/FastaFormat.html#getDefaultFormat()">org.biojava.bio.seq.io.FastaFormat.getDefaultFormat()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I></I>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenbankFormat.html#getDefaultFormat()">org.biojava.bio.seq.io.GenbankFormat.getDefaultFormat()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I></I>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/Chain.html#getGroup(int)">org.biojava.bio.structure.Chain.getGroup(int)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomGroup or getSeqResGroup instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/ChainImpl.html#getGroup(int)">org.biojava.bio.structure.ChainImpl.getGroup(int)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomGroup or getSeqResGroup instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/Chain.html#getGroups()">org.biojava.bio.structure.Chain.getGroups()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomGroups or getSeqResGroups instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/ChainImpl.html#getGroups()">org.biojava.bio.structure.ChainImpl.getGroups()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomGroups instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/Chain.html#getGroups(java.lang.String)">org.biojava.bio.structure.Chain.getGroups(String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomGroups or getSeqResGroups instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/ChainImpl.html#getGroups(java.lang.String)">org.biojava.bio.structure.ChainImpl.getGroups(String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomGroups instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/Structure.html#getHeader()">org.biojava.bio.structure.Structure.getHeader()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getPDBHeader instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/Chain.html#getLength()">org.biojava.bio.structure.Chain.getLength()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>please use getAtomLength or getLengthSeqRes instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/ChainImpl.html#getLength()">org.biojava.bio.structure.ChainImpl.getLength()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomLength instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/Chain.html#getLengthAminos()">org.biojava.bio.structure.Chain.getLengthAminos()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomGroups("amino").size() instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/Chain.html#getSequence()">org.biojava.bio.structure.Chain.getSequence()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomSequence instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/ChainImpl.html#getSequence()">org.biojava.bio.structure.ChainImpl.getSequence()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use getAtomSequence instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SeqIOTools.html#guessFileType(java.io.File)">org.biojava.bio.seq.io.SeqIOTools.guessFileType(File)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>because there is no standard file naming convention
 and guessing by file name is inherantly error prone and bad.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/utils/AbstractChangeable.html#hasListeners()">org.biojava.utils.AbstractChangeable.hasListeners()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use hasListeners(ChangeType) if at all possible</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/dp/MarkovModel.html#heads()">org.biojava.bio.dp.MarkovModel.heads()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use <code>advance().length</code></I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/symbol/AlphabetManager.html#instance()">org.biojava.bio.symbol.AlphabetManager.instance()</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>all AlphabetManager methods have become static</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/gff/GFFTools.html#readGFF(java.lang.String)">org.biojava.bio.program.gff.GFFTools.readGFF(String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use: readGff(File)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/gff/GFFTools.html#readGFF(java.lang.String, org.biojava.bio.program.gff.GFFRecordFilter)">org.biojava.bio.program.gff.GFFTools.readGFF(String, GFFRecordFilter)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use: readGff(File,GFFRecordFilter)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/utils/Changeable.html#removeChangeListener(org.biojava.utils.ChangeListener)">org.biojava.utils.Changeable.removeChangeListener(ChangeListener)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use removeChangeListener(cl, ChangeType.UNKNOWN)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/sax/FastaSearchSAXParser.html#setQuerySeq(java.lang.String)">org.biojava.bio.program.sax.FastaSearchSAXParser.setQuerySeq(String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use <code>setQueryID</code> instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/sax/FastaSearchSAXParser.html#setSubjectDB(java.lang.String)">org.biojava.bio.program.sax.FastaSearchSAXParser.setSubjectDB(String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use <code>setDatabaseID</code> instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojavax/bio/seq/io/RichSeqIOListener.html#setURI(java.lang.String)">org.biojavax.bio.seq.io.RichSeqIOListener.setURI(String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>There is no clear mapping between URI and BioSQL. This
  method is no longer used or supported in biojavax. Don't use it. Calling
  it may result in exceptions. Use instead setName(String name),
  setAccession(String accession), setVersion(int version) etc as
  appropriate.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/symbol/AlphabetManager.html#symbolForName(java.lang.String)">org.biojava.bio.symbol.AlphabetManager.symbolForName(String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use symbolForLifeScienceID() instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/RNATools.html#transcribe(org.biojava.bio.symbol.SymbolList)">org.biojava.bio.seq.RNATools.transcribe(SymbolList)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>The naming of this method is confusing and inconsistent use either DNATools.toRNA(SymbolList list) or
 DNATools.transcribeToRNA(SymbolList list) depending on the desired behaivour.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/program/phred/PhredFormat.html#writeSequence(org.biojava.bio.seq.Sequence, java.lang.String, java.io.PrintStream)">org.biojava.bio.program.phred.PhredFormat.writeSequence(Sequence, String, PrintStream)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use writeSequence(Sequence seq, PrintStream os)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/EmblLikeFormat.html#writeSequence(org.biojava.bio.seq.Sequence, java.lang.String, java.io.PrintStream)">org.biojava.bio.seq.io.EmblLikeFormat.writeSequence(Sequence, String, PrintStream)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use writeSequence(Sequence seq, PrintStream os)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/SequenceFormat.html#writeSequence(org.biojava.bio.seq.Sequence, java.lang.String, java.io.PrintStream)">org.biojava.bio.seq.io.SequenceFormat.writeSequence(Sequence, String, PrintStream)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use writeSequence(Sequence seq, PrintStream os)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/FastaFormat.html#writeSequence(org.biojava.bio.seq.Sequence, java.lang.String, java.io.PrintStream)">org.biojava.bio.seq.io.FastaFormat.writeSequence(Sequence, String, PrintStream)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use writeSequence(Sequence seq, PrintStream os)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/seq/io/GenbankFormat.html#writeSequence(org.biojava.bio.seq.Sequence, java.lang.String, java.io.PrintStream)">org.biojava.bio.seq.io.GenbankFormat.writeSequence(Sequence, String, PrintStream)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use writeSequence(Sequence seq, PrintStream os)</I>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="constructor"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Deprecated Constructors</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/BioError.html#BioError(java.lang.Throwable, java.lang.String)">org.biojava.bio.BioError(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Use BioError(message, ex) instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/BioException.html#BioException(java.lang.Throwable, java.lang.String)">org.biojava.bio.BioException(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use new BioException(message, ex) instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/BioRuntimeException.html#BioRuntimeException(java.lang.Throwable, java.lang.String)">org.biojava.bio.BioRuntimeException(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use new BioRuntimeException(message, ex) instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/utils/ChangeVetoException.html#ChangeVetoException(java.lang.Throwable, java.lang.String)">org.biojava.utils.ChangeVetoException(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use new ChangeVetoException(reason, ex);</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/symbol/PackedSymbolListFactory.html#PackedSymbolListFactory(boolean)">org.biojava.bio.symbol.PackedSymbolListFactory(boolean)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>the argumentless constructor creates a SymbolListFactory
   that will autoselect the packing appropriately.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/utils/ParserException.html#ParserException(java.lang.Throwable, java.lang.String)">org.biojava.utils.ParserException(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use new ParserException(detail, t)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/io/PDBParseException.html#PDBParseException(java.lang.Throwable, java.lang.String)">org.biojava.bio.structure.io.PDBParseException(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use new BioException(message, ex) instead</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/dp/ProfileHMM.html#ProfileHMM(org.biojava.bio.symbol.Alphabet, int, org.biojava.bio.dist.DistributionFactory, org.biojava.bio.dist.DistributionFactory)">org.biojava.bio.dp.ProfileHMM(Alphabet, int, DistributionFactory, DistributionFactory)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I></I>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/proteomics/Protease.html#Protease(java.lang.String, boolean)">org.biojava.bio.proteomics.Protease(String, boolean)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Creating a Protease with this constructor will not register it
 with the ProteaseManager (use ProteaseManager.createProtease())</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/proteomics/Protease.html#Protease(java.lang.String, boolean, java.lang.String)">org.biojava.bio.proteomics.Protease(String, boolean, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Creating a Protease with this constructor will not register it
 with the ProteaseManager (use ProteaseManager.createProtease())</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/proteomics/Protease.html#Protease(org.biojava.bio.symbol.SymbolList, boolean, org.biojava.bio.symbol.SymbolList)">org.biojava.bio.proteomics.Protease(SymbolList, boolean, SymbolList)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>Creating a Protease with this constructor will not register it
 with the ProteaseManager (use ProteaseManager.createProtease())</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/directory/RegistryException.html#RegistryException(java.lang.Throwable, java.lang.String)">org.biojava.directory.RegistryException(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use new RegistryException(message, cause)</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/dp/SimpleDotState.html#SimpleDotState(char, java.lang.String, org.biojava.bio.Annotation)">org.biojava.bio.dp.SimpleDotState(char, String, Annotation)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>token is ignored since 1.2.  Use the 2-arg constructor instead.</I>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/dp/SimpleMarkovModel.html#SimpleMarkovModel(int, org.biojava.bio.symbol.Alphabet)">org.biojava.bio.dp.SimpleMarkovModel(int, Alphabet)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I></I>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="org/biojava/bio/structure/StructureException.html#StructureException(java.lang.Throwable, java.lang.String)">org.biojava.bio.structure.StructureException(Throwable, String)</A>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<I>use new StructureException(message, ex) instead</I>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Use</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Deprecated</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="index.html?deprecated-list.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="deprecated-list.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
