<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_06) on Thu May 22 13:51:09 BST 2008 -->
<TITLE>
Overview (BioJava-live API)
</TITLE>

<META NAME="keywords" CONTENT="Overview, BioJava-live">

<LINK REL ="stylesheet" TYPE="text/css" HREF="stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Overview (BioJava-live API)";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Overview</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Use</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="index.html?overview-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="overview-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H1>
BioJava-live
</H1>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Core biological packages</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bibliography/package-summary.html">org.biojava.bibliography</A></B></TD>
<TD>The data structures and interfaces included in this package define an
interoperable framework for bibliographic searches.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/package-summary.html">org.biojava.bio</A></B></TD>
<TD>The core classes that will be used throughout the bio packages.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/alignment/package-summary.html">org.biojava.bio.alignment</A></B></TD>
<TD>
Classes to generate and describe sequence alignments.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/dist/package-summary.html">org.biojava.bio.dist</A></B></TD>
<TD>
Probability distributions over Alphabets.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/search/package-summary.html">org.biojava.bio.search</A></B></TD>
<TD>
Interfaces and classes for representing sequence similarity search results.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/package-summary.html">org.biojava.bio.seq</A></B></TD>
<TD>Classes and interfaces for defining biological sequences and informatics
objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/db/package-summary.html">org.biojava.bio.seq.db</A></B></TD>
<TD>
Collections of biological sequence data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/filter/package-summary.html">org.biojava.bio.seq.filter</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/io/package-summary.html">org.biojava.bio.seq.io</A></B></TD>
<TD>Classes and interfaces for processing and producing flat-file representations
of sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/symbol/package-summary.html">org.biojava.bio.symbol</A></B></TD>
<TD>Representation of the Symbols that make up a sequence, and locations within
them.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/taxa/package-summary.html">org.biojava.bio.taxa</A></B></TD>
<TD>Taxonomy object for representing species information.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/directory/package-summary.html">org.biojava.directory</A></B></TD>
<TD>Open Bio Sequence Database Access (OBDA) registry support.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>User interface components</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/gui/package-summary.html">org.biojava.bio.gui</A></B></TD>
<TD>Graphical interfaces for biojava objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/gui/glyph/package-summary.html">org.biojava.bio.gui.glyph</A></B></TD>
<TD>Glyphs implementations that draw shapes into SequenceRenderers</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/gui/sequence/package-summary.html">org.biojava.bio.gui.sequence</A></B></TD>
<TD>
Graphical displays of biological sequences and associated annotations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/gui/sequence/tracklayout/package-summary.html">org.biojava.bio.gui.sequence.tracklayout</A></B></TD>
<TD>Classes for the handling of the layout of a WrappedSequencePanel.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Sequence databases and formats</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/das/package-summary.html">org.biojava.bio.program.das</A></B></TD>
<TD>Development client for the Distributed Annotation System.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/das/dasalignment/package-summary.html">org.biojava.bio.program.das.dasalignment</A></B></TD>
<TD>
request alignments via DAS</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/homologene/package-summary.html">org.biojava.bio.program.homologene</A></B></TD>
<TD>
Support classes for Homologene data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/indexdb/package-summary.html">org.biojava.bio.program.indexdb</A></B></TD>
<TD>A flat-file ascii index of ascii flat files as per the OBDA specification.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/xff/package-summary.html">org.biojava.bio.program.xff</A></B></TD>
<TD>Event-driven parsing system for the Extensible Feature Format (XFF).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/db/biofetch/package-summary.html">org.biojava.bio.seq.db.biofetch</A></B></TD>
<TD>Client for the OBDA BioFetch protocol.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/db/biosql/package-summary.html">org.biojava.bio.seq.db.biosql</A></B></TD>
<TD>General purpose Sequence storage in a relational database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/db/emblcd/package-summary.html">org.biojava.bio.seq.db.emblcd</A></B></TD>
<TD>
Readers for the EMBL CD-ROM format binary index files used by EMBOSS and Staden packages.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/db/flat/package-summary.html">org.biojava.bio.seq.db.flat</A></B></TD>
<TD>
Support for OBDA flatfile databases.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/io/agave/package-summary.html">org.biojava.bio.seq.io.agave</A></B></TD>
<TD>Classes for converting between AGAVE XML and BioJava objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/io/game/package-summary.html">org.biojava.bio.seq.io.game</A></B></TD>
<TD>Event-driven parsing system for the Gene Annotation Markup Elements (GAME).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/io/game12/package-summary.html">org.biojava.bio.seq.io.game12</A></B></TD>
<TD>
Event-driven parsing system for the Gene Annotation Markup Elements (GAME).</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Handling output from external tools</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/package-summary.html">org.biojava.bio.program</A></B></TD>
<TD>
Java wrappers for interacting with external bioinformatics tools.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/blast2html/package-summary.html">org.biojava.bio.program.blast2html</A></B></TD>
<TD>Code for generating HTML reports from blast output</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/gff/package-summary.html">org.biojava.bio.program.gff</A></B></TD>
<TD>GFF manipulation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/gff3/package-summary.html">org.biojava.bio.program.gff3</A></B></TD>
<TD>
Support for reading and writing GFF3.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/hmmer/package-summary.html">org.biojava.bio.program.hmmer</A></B></TD>
<TD>
Tools for working with profile Hidden Markov Models from the HMMer package.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/phred/package-summary.html">org.biojava.bio.program.phred</A></B></TD>
<TD>Parser for Phred output</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/sax/package-summary.html">org.biojava.bio.program.sax</A></B></TD>
<TD>Parsers which offer XML representations of the output from common
bioinformatics tools.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/sax/blastxml/package-summary.html">org.biojava.bio.program.sax.blastxml</A></B></TD>
<TD>
Parsers for the XML output from NCBI blast.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/ssbind/package-summary.html">org.biojava.bio.program.ssbind</A></B></TD>
<TD>
Creation of objects from SAX events using the BioJava BlastLikeDataSetCollection DTD.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/xml/package-summary.html">org.biojava.bio.program.xml</A></B></TD>
<TD>Utility classes for the <code>org.biojava.bio.program.sax</code>
package.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Dynamic programming packages</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/dp/package-summary.html">org.biojava.bio.dp</A></B></TD>
<TD>HMM and Dynamic Programming Algorithms.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/dp/onehead/package-summary.html">org.biojava.bio.dp.onehead</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/dp/twohead/package-summary.html">org.biojava.bio.dp.twohead</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Chromatogram and trace file support</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/chromatogram/package-summary.html">org.biojava.bio.chromatogram</A></B></TD>
<TD>
Interfaces and classes for chromatogram data, as produced by DNA sequencing equipment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/chromatogram/graphic/package-summary.html">org.biojava.bio.chromatogram.graphic</A></B></TD>
<TD>
Tools for displaying chromatograms.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/abi/package-summary.html">org.biojava.bio.program.abi</A></B></TD>
<TD>ABI Trace Handling.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/scf/package-summary.html">org.biojava.bio.program.scf</A></B></TD>
<TD>
Support for the SCF chromatogram format.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Macromolecular structure (PDB)</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/das/dasstructure/package-summary.html">org.biojava.bio.program.das.dasstructure</A></B></TD>
<TD>
request a protein structure via DAS</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/package-summary.html">org.biojava.bio.structure</A></B></TD>
<TD>
Interfaces and classes for protein structure (PDB).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/align/package-summary.html">org.biojava.bio.structure.align</A></B></TD>
<TD>Classes for the super-imposition of structures.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/align/helper/package-summary.html">org.biojava.bio.structure.align.helper</A></B></TD>
<TD>Helper classes for structural alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/align/pairwise/package-summary.html">org.biojava.bio.structure.align.pairwise</A></B></TD>
<TD>Classes for the pairwise alignment of structures.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/gui/package-summary.html">org.biojava.bio.structure.gui</A></B></TD>
<TD>
A few convenience classes to view protein structures with Jmol (if it is on the classpath),
and to investigate the internals of the protein structure alignment algorithm</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/io/package-summary.html">org.biojava.bio.structure.io</A></B></TD>
<TD>Input and Output of Structures</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/jama/package-summary.html">org.biojava.bio.structure.jama</A></B></TD>
<TD>Matrix package for from <a href="http://math.nist.gov/javanumerics/jama/">JAMA</a></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/structure/server/package-summary.html">org.biojava.bio.structure.server</A></B></TD>
<TD>
classes for easier management of PDB installations</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Utilities and developers' packages</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/tagvalue/package-summary.html">org.biojava.bio.program.tagvalue</A></B></TD>
<TD> Process files as streams of records, each with tags with values.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/impl/package-summary.html">org.biojava.bio.seq.impl</A></B></TD>
<TD>Standard in-memory implementations of <code>Sequence</code> and
<code>Feature</code>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/io/filterxml/package-summary.html">org.biojava.bio.seq.io.filterxml</A></B></TD>
<TD>Tools for reading and writing an XML representation of BioJava's FeatureFilter language.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/projection/package-summary.html">org.biojava.bio.seq.projection</A></B></TD>
<TD>Code for projecting Feature objects and systematically altering their
properties.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/naming/package-summary.html">org.biojava.naming</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/ontology/package-summary.html">org.biojava.ontology</A></B></TD>
<TD>
A general-purpose API for ontologies.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/ontology/io/package-summary.html">org.biojava.ontology.io</A></B></TD>
<TD>
Tools for loading and saving ontologies.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/package-summary.html">org.biojava.utils</A></B></TD>
<TD>
Miscellaneous utility classes used by other BioJava components.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/automata/package-summary.html">org.biojava.utils.automata</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/cache/package-summary.html">org.biojava.utils.cache</A></B></TD>
<TD>A simple cache system with pluggable caching behaviours.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/candy/package-summary.html">org.biojava.utils.candy</A></B></TD>
<TD>The data structures and interfaces included in this package define an
interoperable framework for simple controlled vocabularies.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/io/package-summary.html">org.biojava.utils.io</A></B></TD>
<TD>I/O utility classes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/math/package-summary.html">org.biojava.utils.math</A></B></TD>
<TD>
Mathematical utility classes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/net/package-summary.html">org.biojava.utils.net</A></B></TD>
<TD>
Network programming utility classes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/process/package-summary.html">org.biojava.utils.process</A></B></TD>
<TD>Classes and interfaces to execute external processes and to handle its input and output streams in multiple threads.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/regex/package-summary.html">org.biojava.utils.regex</A></B></TD>
<TD>This package is used to perform regular expression searches of
SymbolLists defined in arbitrary Alphabets.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/stax/package-summary.html">org.biojava.utils.stax</A></B></TD>
<TD>The Stack API for XML (StAX).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/walker/package-summary.html">org.biojava.utils.walker</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/xml/package-summary.html">org.biojava.utils.xml</A></B></TD>
<TD>
Utility classes for handling and generating XML documents.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Molecular biology packages</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/molbio/package-summary.html">org.biojava.bio.molbio</A></B></TD>
<TD>The classes and interfaces in this package cover common molecular
biological techniques such as restriction digests and PCR.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/proteomics/package-summary.html">org.biojava.bio.proteomics</A></B></TD>
<TD>Utilities to aid in performing various physical analysis of proteins.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/proteomics/aaindex/package-summary.html">org.biojava.bio.proteomics.aaindex</A></B></TD>
<TD>Classes and interfaces to load Amino Acid Index database files.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Experimental packages</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/annodb/package-summary.html">org.biojava.bio.annodb</A></B></TD>
<TD>
Databases of generic structured data (annotation) objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/formats/package-summary.html">org.biojava.bio.program.formats</A></B></TD>
<TD>
Experimental parsers using the tagvalue framework.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/ssaha/package-summary.html">org.biojava.bio.program.ssaha</A></B></TD>
<TD>SSAHA sequence searching API.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/program/unigene/package-summary.html">org.biojava.bio.program.unigene</A></B></TD>
<TD>Objects for representing Unigene clusters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/distributed/package-summary.html">org.biojava.bio.seq.distributed</A></B></TD>
<TD>Sequences and SequenceDBs which are composed from data taken
from a number of data sources.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/bio/seq/homol/package-summary.html">org.biojava.bio.seq.homol</A></B></TD>
<TD>The classes and interfaces for defining sequence similarity and
honology.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/stats/svm/package-summary.html">org.biojava.stats.svm</A></B></TD>
<TD>
Support Vector Machine classification and regression.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/stats/svm/tools/package-summary.html">org.biojava.stats.svm.tools</A></B></TD>
<TD>Tools for use of the SVM package.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojava/utils/lsid/package-summary.html">org.biojava.utils.lsid</A></B></TD>
<TD>Life Science Identifier (LSID) package.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Biojava extension (biojavax) packages</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/package-summary.html">org.biojavax</A></B></TD>
<TD>The Biojava extensions packages, classes that extend the core biojava 
     functionality</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/package-summary.html">org.biojavax.bio</A></B></TD>
<TD>Classes to represent biological entities and their relationships.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/db/package-summary.html">org.biojavax.bio.db</A></B></TD>
<TD>Interactions between biojavax objects and a DB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/db/biosql/package-summary.html">org.biojavax.bio.db.biosql</A></B></TD>
<TD>Interface between biojava and biosql databases</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/db/ncbi/package-summary.html">org.biojavax.bio.db.ncbi</A></B></TD>
<TD>Interfaces to NCBI data.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/seq/package-summary.html">org.biojavax.bio.seq</A></B></TD>
<TD>Rich implementations of Sequences, Locations and Features.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/seq/io/package-summary.html">org.biojavax.bio.seq.io</A></B></TD>
<TD>Classes to support the I/O of RichSequence and 
     Bioentry objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/taxa/package-summary.html">org.biojavax.bio.taxa</A></B></TD>
<TD>Objects that model the NCBI taxonomy schema as described in the BioSQL
  schema.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/taxa/io/package-summary.html">org.biojavax.bio.taxa.io</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/ontology/package-summary.html">org.biojavax.ontology</A></B></TD>
<TD>Extensions to the biojava ontology model that represent BioSQL ontology.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/utils/package-summary.html">org.biojavax.utils</A></B></TD>
<TD>Utility classes that are used by biojavax objects but could have wider utility
  to other applications.</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Genetic Algorithm Framework</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/ga/package-summary.html">org.biojavax.ga</A></B></TD>
<TD>Classes to provide a genetic algorithm framework</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/ga/exception/package-summary.html">org.biojavax.ga.exception</A></B></TD>
<TD>Exceptions related to the GA framework</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/ga/functions/package-summary.html">org.biojavax.ga.functions</A></B></TD>
<TD>GA functions</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/ga/impl/package-summary.html">org.biojavax.ga.impl</A></B></TD>
<TD>Default implementations and abstract classes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/ga/util/package-summary.html">org.biojavax.ga.util</A></B></TD>
<TD>Utility functions and helper classes</TD>
</TR>
</TABLE>

<P>
&nbsp;
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Experimental Phylogeny packages</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/phylo/package-summary.html">org.biojavax.bio.phylo</A></B></TD>
<TD>Classes to support phlylogeny objects.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/phylo/io/nexus/package-summary.html">org.biojavax.bio.phylo.io.nexus</A></B></TD>
<TD>Classes to support the I/O of Nexus files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="20%"><B><A HREF="org/biojavax/bio/phylo/io/phylip/package-summary.html">org.biojavax.bio.phylo.io.phylip</A></B></TD>
<TD>Classes to support the reading and writing of PHYLIP format.</TD>
</TR>
</TABLE>

<P>
&nbsp;<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Overview</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Package</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Use</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="index.html?overview-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="overview-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
